{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#  DAISY- the DAta-mIning SYnthetic-lethality-identification pipeline\n",
    "\n",
    "Please cite: \n",
    "For Implementation: \n",
    "\n",
    "Our paper,\n",
    "\n",
    "For DAISY algorithm: \n",
    "\n",
    "Jerby-Arnon, L., Pfetzer, N., Waldman, Y. Y., McGarry, L., James, D., Shanks, E., ... & Gottlieb, E. (2014). Predicting cancer-specific vulnerability via data-driven detection of synthetic lethality. Cell, 158(5), 1199-1209.\n",
    "\n",
    "For CCLE Omics data:\n",
    "\n",
    "Ghandi, M., Huang, F.W., Jané-Valbuena, J. et al. Next-generation characterization of the Cancer Cell Line Encyclopedia. Nature 569, 503–508 (2019). https://doi.org/10.1038/s41586-019-1186-3\n",
    "\n",
    "For CRISPR Data: \n",
    "\n",
    "Robin M. Meyers, Jordan G. Bryan, James M. McFarland, Barbara A. Weir, ... David E. Root, William C. Hahn, Aviad Tsherniak. Computational correction of copy number effect improves specificity of CRISPR-Cas9 essentiality screens in cancer cells. Nature Genetics 2017 October 49:1779–1784. doi:10.1038/ng.3984\n",
    "\n",
    "Dempster, J. M., Rossen, J., Kazachkova, M., Pan, J., Kugener, G., Root, D. E., & Tsherniak, A. (2019). Extracting Biological Insights from the Project Achilles Genome-Scale CRISPR Screens in Cancer Cell Lines. BioRxiv, 720243.\n",
    "\n",
    "For shRNA Data:\n",
    "\n",
    "\n",
    "This notebook is a reimplementation of DAISY Synthetic Lethal Pair Prediction Algorithm\n",
    "\n",
    "It consists 3 modules: \n",
    "\n",
    "1. SL candidate determination using gene co-expression\n",
    "2. SL candidate determination using survival of fittest\n",
    "3. SL candidate determination using CRISPR and ShRNA experiment\n",
    "\n",
    "* The results from the three modules were then aggregated into one ranked list of candidate SL pairs\n",
    "\n",
    "Input Parameters\n",
    "* Cancer type \n",
    "* The genes whose SL partners are seeked\n",
    "\n",
    "Input Data (available in bigquery tables)\n",
    "* Gene expression data \n",
    "* Gene mutation data\n",
    "* Copy number variation data\n",
    "* Gene effect data (CRISPR)\n",
    "* Gene Dependency scores data (shRNA)\n",
    "\n",
    "Output\n",
    "* Ranked list of candidate SL pairs\n",
    "\n",
    "Please contact Bahar Tercan btercan@systemsbiology.org for your questions and detailed information. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 79,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Once deleted, variables cannot be recovered. Proceed (y/[n])? y\n"
     ]
    }
   ],
   "source": [
    "reset "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 80,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'/Users/bahar/Desktop/Boris__revised_daisy/DAISY_pipeline'"
      ]
     },
     "execution_count": 80,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pwd"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 1. Import python libraries required\n",
    "The required libraries are imported. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 81,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "from datetime import datetime\n",
    "import sys\n",
    "sys.path.append('../scripts/') #need to add \"scripts\" directory in a parent directory \n",
    "from google.cloud import bigquery\n",
    "import importlib\n",
    "import pandas as pd\n",
    "import DAISY_operations\n",
    "importlib.reload(DAISY_operations)\n",
    "from DAISY_operations import *\n",
    "from helper_functions import *\n",
    "from BIGQUERY_operations import *\n",
    "import ipywidgets as widgets"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 82,
   "metadata": {},
   "outputs": [],
   "source": [
    "if not sys.warnoptions:\n",
    "    import warnings\n",
    "    warnings.simplefilter(\"ignore\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 2. Sign in Google Bigquery with the project id\n",
    "\n",
    "Bigquery connection\n",
    "Please replace syntheticlethality with your project name"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 83,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Traceback (most recent call last):\r\n",
      "  File \"/Users/bahar/Downloads/google-cloud-sdk/lib/gcloud.py\", line 104, in <module>\r\n",
      "    main()\r\n",
      "  File \"/Users/bahar/Downloads/google-cloud-sdk/lib/gcloud.py\", line 100, in main\r\n",
      "    sys.exit(gcloud_main.main())\r\n",
      "  File \"/Users/bahar/Downloads/google-cloud-sdk/lib/googlecloudsdk/gcloud_main.py\", line 171, in main\r\n",
      "    gcloud_cli = CreateCLI([])\r\n",
      "  File \"/Users/bahar/Downloads/google-cloud-sdk/lib/googlecloudsdk/gcloud_main.py\", line 151, in CreateCLI\r\n",
      "    generated_cli = loader.Generate()\r\n",
      "  File \"/Users/bahar/Downloads/google-cloud-sdk/lib/googlecloudsdk/calliope/cli.py\", line 504, in Generate\r\n",
      "    cli = self.__MakeCLI(top_group)\r\n",
      "  File \"/Users/bahar/Downloads/google-cloud-sdk/lib/googlecloudsdk/calliope/cli.py\", line 674, in __MakeCLI\r\n",
      "    log.AddFileLogging(self.__logs_dir)\r\n",
      "  File \"/Users/bahar/Downloads/google-cloud-sdk/lib/googlecloudsdk/core/log.py\", line 1039, in AddFileLogging\r\n",
      "    _log_manager.AddLogsDir(logs_dir=logs_dir)\r\n",
      "  File \"/Users/bahar/Downloads/google-cloud-sdk/lib/googlecloudsdk/core/log.py\", line 683, in AddLogsDir\r\n",
      "    self._CleanUpLogs(logs_dir)\r\n",
      "  File \"/Users/bahar/Downloads/google-cloud-sdk/lib/googlecloudsdk/core/log.py\", line 711, in _CleanUpLogs\r\n",
      "    self._CleanLogsDir(logs_dir)\r\n",
      "  File \"/Users/bahar/Downloads/google-cloud-sdk/lib/googlecloudsdk/core/log.py\", line 737, in _CleanLogsDir\r\n",
      "    os.remove(log_file_path)\r\n",
      "PermissionError: [Errno 13] Permission denied: '/Users/bahar/.config/gcloud/logs/2020.08.04/12.16.05.276251.log'\r\n"
     ]
    }
   ],
   "source": [
    "project_id='syntheticlethality'\n",
    "client = bigquery.Client(project_id)\n",
    "#client = bigquery.Client(credentials=credentials, project=credentials.project_id)\n",
    "\n",
    "!gcloud auth login"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 4. Prediction of synthetic lethal partners using different modules on DAISY\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "There are three modules for synthetic lethal pair inferences on DAISY : 1. Pairwise gene coexpression, 2. Genomic survival of the fittest. 3. shRNA or CRISPR based functional examination. You can get more information in the original paper : https://www.sciencedirect.com/science/article/pii/S0092867414009775.\n",
    "\n",
    "In pairwise gene coexpression module and genomic survial of the fittest module, we will use PancancerAtlas and CCLE data.<br>\n",
    "In functional examination module, we will use CRISPR and shRNA data. <br>\n",
    "\n",
    "Python codes for each module are built in our internal library (../scripts/SL_library.py) which was already imported at the beginning. \n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### 4.0. Default parameters for DAISY, you can edit them"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 84,
   "metadata": {},
   "outputs": [],
   "source": [
    "input_mutations = ['Nonsense_Mutation', 'Frame_Shift_Ins', 'Frame_Shift_Del'] # DAISY default parameters\n",
    "percentile_threshold = 10\n",
    "cn_threshold = -0.3 \n",
    "cor_threshold = 0.4\n",
    "p_threshold = 0.05\n",
    "pval_correction = 'Bonferroni'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 85,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "430ca17195eb413c87728dee81b1296b",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "SelectMultiple(description='Tumor type', options=('pancancer', 'CHOL', 'BLCA', 'GBM', 'BRCA', 'CESC', 'COAD', …"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "\n",
    "TCGA_list=GetTCGASubtypes(client)\n",
    "TCGA_list = [i for i in TCGA_list if i]\n",
    "\n",
    "tumor_type = widgets.SelectMultiple(\n",
    "    options=['pancancer'] + TCGA_list  ,\n",
    "    value=[],\n",
    "    description='Tumor type',\n",
    "    disabled=False\n",
    ")\n",
    "display(tumor_type)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### 4.1. Pairwise gene coexpression module"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "4.1.1. Pairwise gene coexpression module on PancancerAtlas."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 86,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "coexp_pancancer = CoexpressionAnalysis(client, 'SL', \"PanCancerAtlas\", ['BRCA1'] , pval_correction, list(tumor_type.value))\n",
    "try:\n",
    "    report=coexp_pancancer.loc[(coexp_pancancer['FDR'] < p_threshold)&(coexp_pancancer['Correlation'] > cor_threshold)]\n",
    "    if report.shape[0]>0:\n",
    "        coexp_pancancer_report=report.groupby('Inactive').apply(lambda x: x.sort_values('FDR'))\n",
    "except:\n",
    "    print(\"No results returned.\")\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 87,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Inactive</th>\n",
       "      <th>InactiveDB</th>\n",
       "      <th>SL_Candidate</th>\n",
       "      <th>#Samples</th>\n",
       "      <th>Correlation</th>\n",
       "      <th>PValue</th>\n",
       "      <th>FDR</th>\n",
       "      <th>Tissue</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>BRCA1</td>\n",
       "      <td>BRCA1</td>\n",
       "      <td>ATAD5</td>\n",
       "      <td>9953</td>\n",
       "      <td>0.827015</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>['pancancer']</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>BRCA1</td>\n",
       "      <td>BRCA1</td>\n",
       "      <td>TOP2A</td>\n",
       "      <td>9953</td>\n",
       "      <td>0.819636</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>['pancancer']</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>BRCA1</td>\n",
       "      <td>BRCA1</td>\n",
       "      <td>CDC6</td>\n",
       "      <td>9953</td>\n",
       "      <td>0.807944</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>['pancancer']</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>BRCA1</td>\n",
       "      <td>BRCA1</td>\n",
       "      <td>DTL</td>\n",
       "      <td>9953</td>\n",
       "      <td>0.802882</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>['pancancer']</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>BRCA1</td>\n",
       "      <td>BRCA1</td>\n",
       "      <td>BRIP1</td>\n",
       "      <td>9953</td>\n",
       "      <td>0.802565</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>['pancancer']</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20277</th>\n",
       "      <td>BRCA1</td>\n",
       "      <td>BRCA1</td>\n",
       "      <td>UBXN6</td>\n",
       "      <td>9953</td>\n",
       "      <td>-0.548051</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>['pancancer']</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20278</th>\n",
       "      <td>BRCA1</td>\n",
       "      <td>BRCA1</td>\n",
       "      <td>TENC1</td>\n",
       "      <td>9953</td>\n",
       "      <td>-0.559599</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>['pancancer']</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20279</th>\n",
       "      <td>BRCA1</td>\n",
       "      <td>BRCA1</td>\n",
       "      <td>MXD4</td>\n",
       "      <td>9953</td>\n",
       "      <td>-0.562911</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>['pancancer']</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20280</th>\n",
       "      <td>BRCA1</td>\n",
       "      <td>BRCA1</td>\n",
       "      <td>CRY2</td>\n",
       "      <td>9953</td>\n",
       "      <td>-0.571943</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>['pancancer']</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20281</th>\n",
       "      <td>BRCA1</td>\n",
       "      <td>BRCA1</td>\n",
       "      <td>GNPTG</td>\n",
       "      <td>9953</td>\n",
       "      <td>-0.584254</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>['pancancer']</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>20282 rows × 8 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      Inactive InactiveDB SL_Candidate  #Samples  Correlation  PValue  FDR  \\\n",
       "0        BRCA1      BRCA1        ATAD5      9953     0.827015     0.0  0.0   \n",
       "1        BRCA1      BRCA1        TOP2A      9953     0.819636     0.0  0.0   \n",
       "2        BRCA1      BRCA1         CDC6      9953     0.807944     0.0  0.0   \n",
       "3        BRCA1      BRCA1          DTL      9953     0.802882     0.0  0.0   \n",
       "4        BRCA1      BRCA1        BRIP1      9953     0.802565     0.0  0.0   \n",
       "...        ...        ...          ...       ...          ...     ...  ...   \n",
       "20277    BRCA1      BRCA1        UBXN6      9953    -0.548051     0.0  0.0   \n",
       "20278    BRCA1      BRCA1        TENC1      9953    -0.559599     0.0  0.0   \n",
       "20279    BRCA1      BRCA1         MXD4      9953    -0.562911     0.0  0.0   \n",
       "20280    BRCA1      BRCA1         CRY2      9953    -0.571943     0.0  0.0   \n",
       "20281    BRCA1      BRCA1        GNPTG      9953    -0.584254     0.0  0.0   \n",
       "\n",
       "              Tissue  \n",
       "0      ['pancancer']  \n",
       "1      ['pancancer']  \n",
       "2      ['pancancer']  \n",
       "3      ['pancancer']  \n",
       "4      ['pancancer']  \n",
       "...              ...  \n",
       "20277  ['pancancer']  \n",
       "20278  ['pancancer']  \n",
       "20279  ['pancancer']  \n",
       "20280  ['pancancer']  \n",
       "20281  ['pancancer']  \n",
       "\n",
       "[20282 rows x 8 columns]"
      ]
     },
     "execution_count": 87,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "coexp_pancancer"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 88,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Inactive</th>\n",
       "      <th>InactiveDB</th>\n",
       "      <th>SL_Candidate</th>\n",
       "      <th>#Samples</th>\n",
       "      <th>Correlation</th>\n",
       "      <th>PValue</th>\n",
       "      <th>FDR</th>\n",
       "      <th>Tissue</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>587</th>\n",
       "      <td>BRCA1</td>\n",
       "      <td>BRCA1</td>\n",
       "      <td>PARP1</td>\n",
       "      <td>9953</td>\n",
       "      <td>0.452299</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>['pancancer']</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    Inactive InactiveDB SL_Candidate  #Samples  Correlation  PValue  FDR  \\\n",
       "587    BRCA1      BRCA1        PARP1      9953     0.452299     0.0  0.0   \n",
       "\n",
       "            Tissue  \n",
       "587  ['pancancer']  "
      ]
     },
     "execution_count": 88,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "coexp_pancancer.loc[coexp_pancancer['SL_Candidate']==\"PARP1\",]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 89,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Inactive</th>\n",
       "      <th>InactiveDB</th>\n",
       "      <th>SL_Candidate</th>\n",
       "      <th>#Samples</th>\n",
       "      <th>Correlation</th>\n",
       "      <th>PValue</th>\n",
       "      <th>FDR</th>\n",
       "      <th>Tissue</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>2092</th>\n",
       "      <td>BRCA1</td>\n",
       "      <td>BRCA1</td>\n",
       "      <td>PARP2</td>\n",
       "      <td>9953</td>\n",
       "      <td>0.266241</td>\n",
       "      <td>4.067601e-161</td>\n",
       "      <td>8.249908e-157</td>\n",
       "      <td>['pancancer']</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     Inactive InactiveDB SL_Candidate  #Samples  Correlation         PValue  \\\n",
       "2092    BRCA1      BRCA1        PARP2      9953     0.266241  4.067601e-161   \n",
       "\n",
       "                FDR         Tissue  \n",
       "2092  8.249908e-157  ['pancancer']  "
      ]
     },
     "execution_count": 89,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "coexp_pancancer.loc[coexp_pancancer['SL_Candidate']==\"PARP2\",]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 90,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Inactive</th>\n",
       "      <th>InactiveDB</th>\n",
       "      <th>SL_Candidate</th>\n",
       "      <th>#Samples</th>\n",
       "      <th>Correlation</th>\n",
       "      <th>PValue</th>\n",
       "      <th>FDR</th>\n",
       "      <th>Tissue</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>BRCA1</td>\n",
       "      <td>BRCA1</td>\n",
       "      <td>ATAD5</td>\n",
       "      <td>9953</td>\n",
       "      <td>0.827015</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>['pancancer']</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>BRCA1</td>\n",
       "      <td>BRCA1</td>\n",
       "      <td>TOP2A</td>\n",
       "      <td>9953</td>\n",
       "      <td>0.819636</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>['pancancer']</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>BRCA1</td>\n",
       "      <td>BRCA1</td>\n",
       "      <td>CDC6</td>\n",
       "      <td>9953</td>\n",
       "      <td>0.807944</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>['pancancer']</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>BRCA1</td>\n",
       "      <td>BRCA1</td>\n",
       "      <td>DTL</td>\n",
       "      <td>9953</td>\n",
       "      <td>0.802882</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>['pancancer']</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>BRCA1</td>\n",
       "      <td>BRCA1</td>\n",
       "      <td>BRIP1</td>\n",
       "      <td>9953</td>\n",
       "      <td>0.802565</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>['pancancer']</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20277</th>\n",
       "      <td>BRCA1</td>\n",
       "      <td>BRCA1</td>\n",
       "      <td>UBXN6</td>\n",
       "      <td>9953</td>\n",
       "      <td>-0.548051</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>['pancancer']</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20278</th>\n",
       "      <td>BRCA1</td>\n",
       "      <td>BRCA1</td>\n",
       "      <td>TENC1</td>\n",
       "      <td>9953</td>\n",
       "      <td>-0.559599</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>['pancancer']</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20279</th>\n",
       "      <td>BRCA1</td>\n",
       "      <td>BRCA1</td>\n",
       "      <td>MXD4</td>\n",
       "      <td>9953</td>\n",
       "      <td>-0.562911</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>['pancancer']</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20280</th>\n",
       "      <td>BRCA1</td>\n",
       "      <td>BRCA1</td>\n",
       "      <td>CRY2</td>\n",
       "      <td>9953</td>\n",
       "      <td>-0.571943</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>['pancancer']</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20281</th>\n",
       "      <td>BRCA1</td>\n",
       "      <td>BRCA1</td>\n",
       "      <td>GNPTG</td>\n",
       "      <td>9953</td>\n",
       "      <td>-0.584254</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>['pancancer']</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>20282 rows × 8 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      Inactive InactiveDB SL_Candidate  #Samples  Correlation  PValue  FDR  \\\n",
       "0        BRCA1      BRCA1        ATAD5      9953     0.827015     0.0  0.0   \n",
       "1        BRCA1      BRCA1        TOP2A      9953     0.819636     0.0  0.0   \n",
       "2        BRCA1      BRCA1         CDC6      9953     0.807944     0.0  0.0   \n",
       "3        BRCA1      BRCA1          DTL      9953     0.802882     0.0  0.0   \n",
       "4        BRCA1      BRCA1        BRIP1      9953     0.802565     0.0  0.0   \n",
       "...        ...        ...          ...       ...          ...     ...  ...   \n",
       "20277    BRCA1      BRCA1        UBXN6      9953    -0.548051     0.0  0.0   \n",
       "20278    BRCA1      BRCA1        TENC1      9953    -0.559599     0.0  0.0   \n",
       "20279    BRCA1      BRCA1         MXD4      9953    -0.562911     0.0  0.0   \n",
       "20280    BRCA1      BRCA1         CRY2      9953    -0.571943     0.0  0.0   \n",
       "20281    BRCA1      BRCA1        GNPTG      9953    -0.584254     0.0  0.0   \n",
       "\n",
       "              Tissue  \n",
       "0      ['pancancer']  \n",
       "1      ['pancancer']  \n",
       "2      ['pancancer']  \n",
       "3      ['pancancer']  \n",
       "4      ['pancancer']  \n",
       "...              ...  \n",
       "20277  ['pancancer']  \n",
       "20278  ['pancancer']  \n",
       "20279  ['pancancer']  \n",
       "20280  ['pancancer']  \n",
       "20281  ['pancancer']  \n",
       "\n",
       "[20282 rows x 8 columns]"
      ]
     },
     "execution_count": 90,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "coexp_pancancer"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<br>\n",
    "4.1.2. Pairwise gene coexpression module on CCLE data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 91,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "coexp_CCLE=CoexpressionAnalysis(client, 'SL', 'CCLE', ['BRCA1'], pval_correction, list(tumor_type.value))\n",
    "try: \n",
    "    report=coexp_CCLE.loc[(coexp_CCLE['FDR'] < p_threshold)&(coexp_CCLE['Correlation'] > cor_threshold)]\n",
    "    if report.shape[0]>0:\n",
    "        coexp_CCLE_report=report.groupby('Inactive').apply(lambda x: x.sort_values('FDR'))\n",
    "except:\n",
    "    print(\"No results returned.\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 92,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Inactive</th>\n",
       "      <th>InactiveDB</th>\n",
       "      <th>SL_Candidate</th>\n",
       "      <th>#Samples</th>\n",
       "      <th>Correlation</th>\n",
       "      <th>PValue</th>\n",
       "      <th>FDR</th>\n",
       "      <th>Tissue</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>135</th>\n",
       "      <td>BRCA1</td>\n",
       "      <td>BRCA1</td>\n",
       "      <td>PARP1</td>\n",
       "      <td>1297</td>\n",
       "      <td>0.591197</td>\n",
       "      <td>4.599640e-123</td>\n",
       "      <td>8.794052e-119</td>\n",
       "      <td>['pancancer']</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    Inactive InactiveDB SL_Candidate  #Samples  Correlation         PValue  \\\n",
       "135    BRCA1      BRCA1        PARP1      1297     0.591197  4.599640e-123   \n",
       "\n",
       "               FDR         Tissue  \n",
       "135  8.794052e-119  ['pancancer']  "
      ]
     },
     "execution_count": 92,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "coexp_CCLE.loc[coexp_CCLE['SL_Candidate']==\"PARP1\",]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 93,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Inactive</th>\n",
       "      <th>InactiveDB</th>\n",
       "      <th>SL_Candidate</th>\n",
       "      <th>#Samples</th>\n",
       "      <th>Correlation</th>\n",
       "      <th>PValue</th>\n",
       "      <th>FDR</th>\n",
       "      <th>Tissue</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1123</th>\n",
       "      <td>BRCA1</td>\n",
       "      <td>BRCA1</td>\n",
       "      <td>PARP2</td>\n",
       "      <td>1297</td>\n",
       "      <td>0.434211</td>\n",
       "      <td>9.154777e-61</td>\n",
       "      <td>1.750302e-56</td>\n",
       "      <td>['pancancer']</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     Inactive InactiveDB SL_Candidate  #Samples  Correlation        PValue  \\\n",
       "1123    BRCA1      BRCA1        PARP2      1297     0.434211  9.154777e-61   \n",
       "\n",
       "               FDR         Tissue  \n",
       "1123  1.750302e-56  ['pancancer']  "
      ]
     },
     "execution_count": 93,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "coexp_CCLE.loc[coexp_CCLE['SL_Candidate']==\"PARP2\",]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 94,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Inactive</th>\n",
       "      <th>InactiveDB</th>\n",
       "      <th>SL_Candidate</th>\n",
       "      <th>#Samples</th>\n",
       "      <th>Correlation</th>\n",
       "      <th>PValue</th>\n",
       "      <th>FDR</th>\n",
       "      <th>Tissue</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>BRCA1</td>\n",
       "      <td>BRCA1</td>\n",
       "      <td>ATAD5</td>\n",
       "      <td>1297</td>\n",
       "      <td>0.746083</td>\n",
       "      <td>6.359384e-231</td>\n",
       "      <td>1.215851e-226</td>\n",
       "      <td>['pancancer']</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>BRCA1</td>\n",
       "      <td>BRCA1</td>\n",
       "      <td>FANCI</td>\n",
       "      <td>1297</td>\n",
       "      <td>0.718720</td>\n",
       "      <td>1.404211e-206</td>\n",
       "      <td>2.684711e-202</td>\n",
       "      <td>['pancancer']</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>BRCA1</td>\n",
       "      <td>BRCA1</td>\n",
       "      <td>C17orf53</td>\n",
       "      <td>1297</td>\n",
       "      <td>0.711617</td>\n",
       "      <td>9.940688e-201</td>\n",
       "      <td>1.900560e-196</td>\n",
       "      <td>['pancancer']</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>BRCA1</td>\n",
       "      <td>BRCA1</td>\n",
       "      <td>CHAF1A</td>\n",
       "      <td>1297</td>\n",
       "      <td>0.704579</td>\n",
       "      <td>4.185098e-195</td>\n",
       "      <td>8.001489e-191</td>\n",
       "      <td>['pancancer']</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>BRCA1</td>\n",
       "      <td>BRCA1</td>\n",
       "      <td>TOPBP1</td>\n",
       "      <td>1297</td>\n",
       "      <td>0.702716</td>\n",
       "      <td>1.209211e-193</td>\n",
       "      <td>2.311890e-189</td>\n",
       "      <td>['pancancer']</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19114</th>\n",
       "      <td>BRCA1</td>\n",
       "      <td>BRCA1</td>\n",
       "      <td>LAMB2</td>\n",
       "      <td>1297</td>\n",
       "      <td>-0.319869</td>\n",
       "      <td>3.065237e-32</td>\n",
       "      <td>5.860427e-28</td>\n",
       "      <td>['pancancer']</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19115</th>\n",
       "      <td>BRCA1</td>\n",
       "      <td>BRCA1</td>\n",
       "      <td>IL1R1</td>\n",
       "      <td>1297</td>\n",
       "      <td>-0.323175</td>\n",
       "      <td>6.536468e-33</td>\n",
       "      <td>1.249707e-28</td>\n",
       "      <td>['pancancer']</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19116</th>\n",
       "      <td>BRCA1</td>\n",
       "      <td>BRCA1</td>\n",
       "      <td>KDELR3</td>\n",
       "      <td>1297</td>\n",
       "      <td>-0.325537</td>\n",
       "      <td>2.141650e-33</td>\n",
       "      <td>4.094620e-29</td>\n",
       "      <td>['pancancer']</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19117</th>\n",
       "      <td>BRCA1</td>\n",
       "      <td>BRCA1</td>\n",
       "      <td>S100A11</td>\n",
       "      <td>1297</td>\n",
       "      <td>-0.361802</td>\n",
       "      <td>2.160366e-41</td>\n",
       "      <td>4.130403e-37</td>\n",
       "      <td>['pancancer']</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19118</th>\n",
       "      <td>BRCA1</td>\n",
       "      <td>BRCA1</td>\n",
       "      <td>RRAS</td>\n",
       "      <td>1297</td>\n",
       "      <td>-0.363530</td>\n",
       "      <td>8.445151e-42</td>\n",
       "      <td>1.614629e-37</td>\n",
       "      <td>['pancancer']</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>19119 rows × 8 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      Inactive InactiveDB SL_Candidate  #Samples  Correlation         PValue  \\\n",
       "0        BRCA1      BRCA1        ATAD5      1297     0.746083  6.359384e-231   \n",
       "1        BRCA1      BRCA1        FANCI      1297     0.718720  1.404211e-206   \n",
       "2        BRCA1      BRCA1     C17orf53      1297     0.711617  9.940688e-201   \n",
       "3        BRCA1      BRCA1       CHAF1A      1297     0.704579  4.185098e-195   \n",
       "4        BRCA1      BRCA1       TOPBP1      1297     0.702716  1.209211e-193   \n",
       "...        ...        ...          ...       ...          ...            ...   \n",
       "19114    BRCA1      BRCA1        LAMB2      1297    -0.319869   3.065237e-32   \n",
       "19115    BRCA1      BRCA1        IL1R1      1297    -0.323175   6.536468e-33   \n",
       "19116    BRCA1      BRCA1       KDELR3      1297    -0.325537   2.141650e-33   \n",
       "19117    BRCA1      BRCA1      S100A11      1297    -0.361802   2.160366e-41   \n",
       "19118    BRCA1      BRCA1         RRAS      1297    -0.363530   8.445151e-42   \n",
       "\n",
       "                 FDR         Tissue  \n",
       "0      1.215851e-226  ['pancancer']  \n",
       "1      2.684711e-202  ['pancancer']  \n",
       "2      1.900560e-196  ['pancancer']  \n",
       "3      8.001489e-191  ['pancancer']  \n",
       "4      2.311890e-189  ['pancancer']  \n",
       "...              ...            ...  \n",
       "19114   5.860427e-28  ['pancancer']  \n",
       "19115   1.249707e-28  ['pancancer']  \n",
       "19116   4.094620e-29  ['pancancer']  \n",
       "19117   4.130403e-37  ['pancancer']  \n",
       "19118   1.614629e-37  ['pancancer']  \n",
       "\n",
       "[19119 rows x 8 columns]"
      ]
     },
     "execution_count": 94,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "coexp_CCLE"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### 4.2. Genomic survival of fittest module"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "4.2.1. Genomic survival of fittest module on CCLE data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 95,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "sof_CCLE = SurvivalOfFittest(client, 'SL', \"CCLE\", ['BRCA1'],  percentile_threshold, cn_threshold, pval_correction, list(tumor_type.value), input_mutations)\n",
    "try: \n",
    "    report=sof_CCLE.loc[(sof_CCLE['PValue'] < p_threshold),]\n",
    "    sof_ccle_report=report.groupby('Inactive').apply(lambda x: x.sort_values('FDR'))\n",
    "except:\n",
    "    print(\"No results returned.\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 96,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Inactive</th>\n",
       "      <th>InactiveDB</th>\n",
       "      <th>SL_Candidate</th>\n",
       "      <th>#InactiveSamples</th>\n",
       "      <th>#Samples</th>\n",
       "      <th>U1</th>\n",
       "      <th>PValue</th>\n",
       "      <th>FDR</th>\n",
       "      <th>Tissue</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1400</th>\n",
       "      <td>BRCA1</td>\n",
       "      <td>BRCA1</td>\n",
       "      <td>PARP1</td>\n",
       "      <td>38</td>\n",
       "      <td>1284</td>\n",
       "      <td>29766.0</td>\n",
       "      <td>0.00341</td>\n",
       "      <td>1.0</td>\n",
       "      <td>['pancancer']</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     Inactive InactiveDB SL_Candidate  #InactiveSamples  #Samples       U1  \\\n",
       "1400    BRCA1      BRCA1        PARP1                38      1284  29766.0   \n",
       "\n",
       "       PValue  FDR         Tissue  \n",
       "1400  0.00341  1.0  ['pancancer']  "
      ]
     },
     "execution_count": 96,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sof_CCLE.loc[sof_CCLE['SL_Candidate']==\"PARP1\",]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 97,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Inactive</th>\n",
       "      <th>InactiveDB</th>\n",
       "      <th>SL_Candidate</th>\n",
       "      <th>#InactiveSamples</th>\n",
       "      <th>#Samples</th>\n",
       "      <th>U1</th>\n",
       "      <th>PValue</th>\n",
       "      <th>FDR</th>\n",
       "      <th>Tissue</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>5332</th>\n",
       "      <td>BRCA1</td>\n",
       "      <td>BRCA1</td>\n",
       "      <td>PARP2</td>\n",
       "      <td>38</td>\n",
       "      <td>1284</td>\n",
       "      <td>27416.0</td>\n",
       "      <td>0.048271</td>\n",
       "      <td>1.0</td>\n",
       "      <td>['pancancer']</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     Inactive InactiveDB SL_Candidate  #InactiveSamples  #Samples       U1  \\\n",
       "5332    BRCA1      BRCA1        PARP2                38      1284  27416.0   \n",
       "\n",
       "        PValue  FDR         Tissue  \n",
       "5332  0.048271  1.0  ['pancancer']  "
      ]
     },
     "execution_count": 97,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sof_CCLE.loc[sof_CCLE['SL_Candidate']==\"PARP2\",]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 98,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "sof_pancancer = SurvivalOfFittest(client, 'SL', \"PanCancerAtlas\", ['BRCA1'], percentile_threshold, cn_threshold, pval_correction, list(tumor_type.value), input_mutations)\n",
    "try:\n",
    "    report=sof_pancancer.loc[(sof_pancancer['FDR'] < p_threshold),]                \n",
    "    if report.shape[0]>0:\n",
    "        sof_pancancer_report=report.groupby('Inactive').apply(lambda x: x.sort_values('FDR'))\n",
    "        sof_pancancer_report  \n",
    "except:\n",
    "    print(\"No results returned.\") "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 99,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Inactive</th>\n",
       "      <th>InactiveDB</th>\n",
       "      <th>SL_Candidate</th>\n",
       "      <th>#InactiveSamples</th>\n",
       "      <th>#Samples</th>\n",
       "      <th>U1</th>\n",
       "      <th>PValue</th>\n",
       "      <th>FDR</th>\n",
       "      <th>Tissue</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>BRCA1</td>\n",
       "      <td>BRCA1</td>\n",
       "      <td>NWD1</td>\n",
       "      <td>234</td>\n",
       "      <td>8930</td>\n",
       "      <td>1305850.5</td>\n",
       "      <td>6.250556e-14</td>\n",
       "      <td>1.570640e-09</td>\n",
       "      <td>['pancancer']</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>BRCA1</td>\n",
       "      <td>BRCA1</td>\n",
       "      <td>KIAA1683</td>\n",
       "      <td>234</td>\n",
       "      <td>8930</td>\n",
       "      <td>1305687.5</td>\n",
       "      <td>6.450396e-14</td>\n",
       "      <td>1.620855e-09</td>\n",
       "      <td>['pancancer']</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>BRCA1</td>\n",
       "      <td>BRCA1</td>\n",
       "      <td>MIR3188</td>\n",
       "      <td>234</td>\n",
       "      <td>8930</td>\n",
       "      <td>1305525.5</td>\n",
       "      <td>6.655787e-14</td>\n",
       "      <td>1.672466e-09</td>\n",
       "      <td>['pancancer']</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>BRCA1</td>\n",
       "      <td>BRCA1</td>\n",
       "      <td>JUND</td>\n",
       "      <td>234</td>\n",
       "      <td>8930</td>\n",
       "      <td>1305525.5</td>\n",
       "      <td>6.655787e-14</td>\n",
       "      <td>1.672466e-09</td>\n",
       "      <td>['pancancer']</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>BRCA1</td>\n",
       "      <td>BRCA1</td>\n",
       "      <td>F2RL3</td>\n",
       "      <td>234</td>\n",
       "      <td>8930</td>\n",
       "      <td>1305301.0</td>\n",
       "      <td>6.955547e-14</td>\n",
       "      <td>1.747790e-09</td>\n",
       "      <td>['pancancer']</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25123</th>\n",
       "      <td>BRCA1</td>\n",
       "      <td>BRCA1</td>\n",
       "      <td>SNORD10</td>\n",
       "      <td>234</td>\n",
       "      <td>8930</td>\n",
       "      <td>525158.0</td>\n",
       "      <td>1.000000e+00</td>\n",
       "      <td>1.000000e+00</td>\n",
       "      <td>['pancancer']</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25124</th>\n",
       "      <td>BRCA1</td>\n",
       "      <td>BRCA1</td>\n",
       "      <td>TBX21</td>\n",
       "      <td>234</td>\n",
       "      <td>8930</td>\n",
       "      <td>474835.0</td>\n",
       "      <td>1.000000e+00</td>\n",
       "      <td>1.000000e+00</td>\n",
       "      <td>['pancancer']</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25125</th>\n",
       "      <td>BRCA1</td>\n",
       "      <td>BRCA1</td>\n",
       "      <td>YWHAEP7</td>\n",
       "      <td>234</td>\n",
       "      <td>8930</td>\n",
       "      <td>367733.0</td>\n",
       "      <td>1.000000e+00</td>\n",
       "      <td>1.000000e+00</td>\n",
       "      <td>['pancancer']</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25126</th>\n",
       "      <td>BRCA1</td>\n",
       "      <td>BRCA1</td>\n",
       "      <td>C1QBP</td>\n",
       "      <td>234</td>\n",
       "      <td>8930</td>\n",
       "      <td>548335.0</td>\n",
       "      <td>1.000000e+00</td>\n",
       "      <td>1.000000e+00</td>\n",
       "      <td>['pancancer']</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25127</th>\n",
       "      <td>BRCA1</td>\n",
       "      <td>BRCA1</td>\n",
       "      <td>TMEM31</td>\n",
       "      <td>234</td>\n",
       "      <td>8930</td>\n",
       "      <td>644197.0</td>\n",
       "      <td>1.000000e+00</td>\n",
       "      <td>1.000000e+00</td>\n",
       "      <td>['pancancer']</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>25128 rows × 9 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      Inactive InactiveDB SL_Candidate  #InactiveSamples  #Samples         U1  \\\n",
       "0        BRCA1      BRCA1         NWD1               234      8930  1305850.5   \n",
       "1        BRCA1      BRCA1     KIAA1683               234      8930  1305687.5   \n",
       "2        BRCA1      BRCA1      MIR3188               234      8930  1305525.5   \n",
       "3        BRCA1      BRCA1         JUND               234      8930  1305525.5   \n",
       "4        BRCA1      BRCA1        F2RL3               234      8930  1305301.0   \n",
       "...        ...        ...          ...               ...       ...        ...   \n",
       "25123    BRCA1      BRCA1      SNORD10               234      8930   525158.0   \n",
       "25124    BRCA1      BRCA1        TBX21               234      8930   474835.0   \n",
       "25125    BRCA1      BRCA1      YWHAEP7               234      8930   367733.0   \n",
       "25126    BRCA1      BRCA1        C1QBP               234      8930   548335.0   \n",
       "25127    BRCA1      BRCA1       TMEM31               234      8930   644197.0   \n",
       "\n",
       "             PValue           FDR         Tissue  \n",
       "0      6.250556e-14  1.570640e-09  ['pancancer']  \n",
       "1      6.450396e-14  1.620855e-09  ['pancancer']  \n",
       "2      6.655787e-14  1.672466e-09  ['pancancer']  \n",
       "3      6.655787e-14  1.672466e-09  ['pancancer']  \n",
       "4      6.955547e-14  1.747790e-09  ['pancancer']  \n",
       "...             ...           ...            ...  \n",
       "25123  1.000000e+00  1.000000e+00  ['pancancer']  \n",
       "25124  1.000000e+00  1.000000e+00  ['pancancer']  \n",
       "25125  1.000000e+00  1.000000e+00  ['pancancer']  \n",
       "25126  1.000000e+00  1.000000e+00  ['pancancer']  \n",
       "25127  1.000000e+00  1.000000e+00  ['pancancer']  \n",
       "\n",
       "[25128 rows x 9 columns]"
      ]
     },
     "execution_count": 99,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sof_pancancer"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 101,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Inactive</th>\n",
       "      <th>InactiveDB</th>\n",
       "      <th>SL_Candidate</th>\n",
       "      <th>#InactiveSamples</th>\n",
       "      <th>#Samples</th>\n",
       "      <th>U1</th>\n",
       "      <th>PValue</th>\n",
       "      <th>FDR</th>\n",
       "      <th>Tissue</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>749</th>\n",
       "      <td>BRCA1</td>\n",
       "      <td>BRCA1</td>\n",
       "      <td>PARP2</td>\n",
       "      <td>234</td>\n",
       "      <td>8930</td>\n",
       "      <td>1234410.5</td>\n",
       "      <td>1.233575e-08</td>\n",
       "      <td>0.00031</td>\n",
       "      <td>['pancancer']</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    Inactive InactiveDB SL_Candidate  #InactiveSamples  #Samples         U1  \\\n",
       "749    BRCA1      BRCA1        PARP2               234      8930  1234410.5   \n",
       "\n",
       "           PValue      FDR         Tissue  \n",
       "749  1.233575e-08  0.00031  ['pancancer']  "
      ]
     },
     "execution_count": 101,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sof_pancancer.loc[sof_pancancer['SL_Candidate']==\"PARP2\",]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 103,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Inactive</th>\n",
       "      <th>InactiveDB</th>\n",
       "      <th>SL_Candidate</th>\n",
       "      <th>#InactiveSamples</th>\n",
       "      <th>#Samples</th>\n",
       "      <th>U1</th>\n",
       "      <th>PValue</th>\n",
       "      <th>FDR</th>\n",
       "      <th>Tissue</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>18913</th>\n",
       "      <td>BRCA1</td>\n",
       "      <td>BRCA1</td>\n",
       "      <td>PARP1</td>\n",
       "      <td>234</td>\n",
       "      <td>8930</td>\n",
       "      <td>831956.0</td>\n",
       "      <td>0.999999</td>\n",
       "      <td>1.0</td>\n",
       "      <td>['pancancer']</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      Inactive InactiveDB SL_Candidate  #InactiveSamples  #Samples        U1  \\\n",
       "18913    BRCA1      BRCA1        PARP1               234      8930  831956.0   \n",
       "\n",
       "         PValue  FDR         Tissue  \n",
       "18913  0.999999  1.0  ['pancancer']  "
      ]
     },
     "execution_count": 103,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sof_pancancer.loc[sof_pancancer['SL_Candidate']==\"PARP1\",]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### 4.3. Functional examination inference module"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "4.3.1. CRISPR based functional examination inference module"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 104,
   "metadata": {},
   "outputs": [],
   "source": [
    "crispr_result = FunctionalExamination(client,'SL', \"CRISPR\", ['BRCA1'], percentile_threshold, cn_threshold, pval_correction,list(tumor_type.value), input_mutations )\n",
    "try:\n",
    "    report=crispr_result.loc[(crispr_result['PValue'] < p_threshold),]\n",
    "                      \n",
    "    if report.shape[0]>0:\n",
    "        crispr_report=report.groupby('Inactive').apply(lambda x: x.sort_values('PValue'))\n",
    "        crispr_report\n",
    "except:\n",
    "    print(\"No results returned.\")\n",
    "    \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 105,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Inactive</th>\n",
       "      <th>InactiveDB</th>\n",
       "      <th>SL_Candidate</th>\n",
       "      <th>#InactiveSamples</th>\n",
       "      <th>#Samples</th>\n",
       "      <th>U1</th>\n",
       "      <th>PValue</th>\n",
       "      <th>FDR</th>\n",
       "      <th>Tissue</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>BRCA1</td>\n",
       "      <td>BRCA1</td>\n",
       "      <td>PARP1</td>\n",
       "      <td>26</td>\n",
       "      <td>771</td>\n",
       "      <td>5418.0</td>\n",
       "      <td>0.000066</td>\n",
       "      <td>1.0</td>\n",
       "      <td>['pancancer']</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  Inactive InactiveDB SL_Candidate  #InactiveSamples  #Samples      U1  \\\n",
       "2    BRCA1      BRCA1        PARP1                26       771  5418.0   \n",
       "\n",
       "     PValue  FDR         Tissue  \n",
       "2  0.000066  1.0  ['pancancer']  "
      ]
     },
     "execution_count": 105,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "crispr_result.loc[crispr_result['SL_Candidate']==\"PARP1\",]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 106,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Inactive</th>\n",
       "      <th>InactiveDB</th>\n",
       "      <th>SL_Candidate</th>\n",
       "      <th>#InactiveSamples</th>\n",
       "      <th>#Samples</th>\n",
       "      <th>U1</th>\n",
       "      <th>PValue</th>\n",
       "      <th>FDR</th>\n",
       "      <th>Tissue</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>5430</th>\n",
       "      <td>BRCA1</td>\n",
       "      <td>BRCA1</td>\n",
       "      <td>PARP2</td>\n",
       "      <td>26</td>\n",
       "      <td>771</td>\n",
       "      <td>9080.0</td>\n",
       "      <td>0.293921</td>\n",
       "      <td>1.0</td>\n",
       "      <td>['pancancer']</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     Inactive InactiveDB SL_Candidate  #InactiveSamples  #Samples      U1  \\\n",
       "5430    BRCA1      BRCA1        PARP2                26       771  9080.0   \n",
       "\n",
       "        PValue  FDR         Tissue  \n",
       "5430  0.293921  1.0  ['pancancer']  "
      ]
     },
     "execution_count": 106,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "crispr_result.loc[crispr_result['SL_Candidate']==\"PARP2\",]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<br>\n",
    "4.3.2. shRNA based functional examination inference module"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 107,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "shRNA_result = FunctionalExamination(client, 'SL', \"shRNA\", ['BRCA1'] , percentile_threshold, \\\n",
    "                                     cn_threshold, pval_correction, list(tumor_type.value),input_mutations)\n",
    "try:\n",
    "    report=shRNA_result.loc[(shRNA_result['PValue'] < p_threshold),]\n",
    "                      \n",
    "    if report.shape[0]>0:\n",
    "        shRNA_report=report.groupby('Inactive').apply(lambda x: x.sort_values('PValue'))\n",
    "    shRNA_report\n",
    "except:\n",
    "    print(\"No results returned.\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 108,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Inactive</th>\n",
       "      <th>InactiveDB</th>\n",
       "      <th>SL_Candidate</th>\n",
       "      <th>#InactiveSamples</th>\n",
       "      <th>#Samples</th>\n",
       "      <th>U1</th>\n",
       "      <th>PValue</th>\n",
       "      <th>FDR</th>\n",
       "      <th>Tissue</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>7573</th>\n",
       "      <td>BRCA1</td>\n",
       "      <td>BRCA1</td>\n",
       "      <td>PARP1</td>\n",
       "      <td>27</td>\n",
       "      <td>651</td>\n",
       "      <td>8429.0</td>\n",
       "      <td>0.502085</td>\n",
       "      <td>1.0</td>\n",
       "      <td>['pancancer']</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     Inactive InactiveDB SL_Candidate  #InactiveSamples  #Samples      U1  \\\n",
       "7573    BRCA1      BRCA1        PARP1                27       651  8429.0   \n",
       "\n",
       "        PValue  FDR         Tissue  \n",
       "7573  0.502085  1.0  ['pancancer']  "
      ]
     },
     "execution_count": 108,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "shRNA_result.loc[shRNA_result['SL_Candidate']==\"PARP1\",]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 109,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Inactive</th>\n",
       "      <th>InactiveDB</th>\n",
       "      <th>SL_Candidate</th>\n",
       "      <th>#InactiveSamples</th>\n",
       "      <th>#Samples</th>\n",
       "      <th>U1</th>\n",
       "      <th>PValue</th>\n",
       "      <th>FDR</th>\n",
       "      <th>Tissue</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>5317</th>\n",
       "      <td>BRCA1</td>\n",
       "      <td>BRCA1</td>\n",
       "      <td>PARP2</td>\n",
       "      <td>15</td>\n",
       "      <td>394</td>\n",
       "      <td>2687.0</td>\n",
       "      <td>0.359623</td>\n",
       "      <td>1.0</td>\n",
       "      <td>['pancancer']</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     Inactive InactiveDB SL_Candidate  #InactiveSamples  #Samples      U1  \\\n",
       "5317    BRCA1      BRCA1        PARP2                15       394  2687.0   \n",
       "\n",
       "        PValue  FDR         Tissue  \n",
       "5317  0.359623  1.0  ['pancancer']  "
      ]
     },
     "execution_count": 109,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "shRNA_result.loc[shRNA_result['SL_Candidate']==\"PARP2\",]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 5. Integration of results"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "5.1. Integration of the pairwise Co-expression gene co-expression results on Pancancer and CCLE"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "coexpression_result = UnionResults([coexp_pancancer_report, coexp_CCLE_report],'SL', ['FDR', 'FDR'],  list(tumor_type.value))\n",
    "coexpression_result=coexpression_result.groupby('Inactive').apply(lambda x: x.sort_values('AggregatedP'))\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<br>\n",
    "5.2. Integration of Survival of Fittest results on Pancancer and CCLE"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 303,
   "metadata": {},
   "outputs": [],
   "source": [
    "sof_result = UnionResults([sof_ccle_report, sof_pancancer_report],  'SL', ['FDR', 'FDR'], list(tumor_type.value))\n",
    "sof_result=sof_result.groupby('Inactive').apply(lambda x: x.sort_values('AggregatedP'))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 302,
   "metadata": {},
   "outputs": [],
   "source": [
    "sof_result=sof_pancancer_report"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<br>\n",
    "5.3. Integration of shRNA and CRISPR based functional examination inference module."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "functional_screening_result = UnionResults([crispr_report, shRNA_report],'SL', ['PValue', 'PValue'], list(tumor_type.value))\n",
    "functional_screening_result=functional_screening_result.groupby('Inactive').apply(lambda x: x.sort_values('AggregatedP'))\n",
    "functional_screening_result"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<br>\n",
    "5.4. Merging the results from all three inference procedures"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "all_merged_results = MergeResults([coexpression_result, sof_result], 'SL', list(tumor_type.value))\n",
    "all_merged_results=all_merged_results.groupby('Inactive').apply(lambda x: x.sort_values('FinalP'))\n",
    "all_merged_results"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Results are saved in excel file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 241,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "WriteToExcel(\"DAISY_SL_results.xlsx\", [ all_merged_results],[ \"SL_results\"])\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
