{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Conservation-based Synthetic Lethal Search\n",
    "\n",
    "## Introduction\n",
    "\n",
    "### Rationale\n",
    "\n",
    "### Use-cases:\n",
    "* Prioritize human candidate synthetic lethal interactions based on prior evidence of interaction in yeast SL screens\n",
    "* _de novo_ discovery of SL interactions\n",
    "\n",
    "### Approach\n",
    "This notebook re-implements the approach outlined in Srivas et al. (2016)\n",
    "\n",
    "### Usage:\n",
    "Add genes of interest to \"inputGenes\" value, then run the next step.\n",
    "Example: inputGenes = \"'DDX3X','DICER1','DROSHA','TNFRSF14','TRAF7','TSC1','POLG','FBXO11','PRDM1','RFWD3','AMER1','LZTR1','ATP2B3'\"\n",
    "\n",
    "### Workflow Overview\n",
    "\n",
    "### Datasets\n",
    "#### Yeast Synthetic Lethal Interactions\n",
    "Constanzo et al. (2016)\n",
    "#### Human to Yeast Ortholog Mapping\n",
    "detailed treatement in the accompanying notebook (Mapping human to yeast orthologs)\n",
    "#### Human Tumor Suppressor Genes\n",
    "\n",
    "\n",
    "### References\n",
    "* Costanzo M, VanderSluis B, Koch EN, Baryshnikova A, Pons C, Tan G, Wang W, Usaj M, Hanchard J, Lee SD, Pelechano V, Styles EB, Billmann M, van Leeuwen J, van Dyk N, Lin ZY, Kuzmin E, Nelson J, Piotrowski JS, Srikumar T, Bahr S, Chen Y, Deshpande R, Kurat CF, Li SC, Li Z, Usaj MM, Okada H, Pascoe N, San Luis BJ, Sharifpoor S, Shuteriqi E, Simpkins SW, Snider J, Suresh HG, Tan Y, Zhu H, Malod-Dognin N, Janjic V, Przulj N, Troyanskaya OG, Stagljar I, Xia T, Ohya Y, Gingras AC, Raught B, Boutros M, Steinmetz LM, Moore CL, Rosebrock AP, Caudy AA, Myers CL, Andrews B, Boone C. **A global genetic interaction network maps a wiring diagram of cellular function.** Science. 2016 Sep 23;353(6306). pii: aaf1420. PubMed PMID: 27708008; PubMed Central PMCID: PMC5661885.\n",
    "* Srivas R, Shen JP, Yang CC, Sun SM, Li J, Gross AM, Jensen J, Licon K, Bojorquez-Gomez A, Klepper K, Huang J, Pekin D, Xu JL, Yeerna H, Sivaganesh V, Kollenstart L, van Attikum H, Aza-Blanc P, Sobol RW, Ideker T. **A Network of Conserved Synthetic Lethal Interactions for Exploration of Precision Cancer Therapy**. Mol Cell. 2016 Aug 4;63(3):514-25. doi:10.1016/j.molcel.2016.06.022.Epub 2016 Jul 21. PubMed PMID: 27453043; PubMed Central PMCID: PMC5209245. "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Preamble\n",
    "This section describes how to setup the analysis environment appropriately, including google cloud platform authentication and importing all the relevant python libraries."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Setup Analysis Environment"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Requirement already satisfied: google-cloud-bigquery in c:\\users\\salta\\miniconda3\\lib\\site-packages (2.4.0)\n",
      "Requirement already satisfied: google-resumable-media<2.0dev,>=0.6.0 in c:\\users\\salta\\miniconda3\\lib\\site-packages (from google-cloud-bigquery) (1.1.0)\n",
      "Requirement already satisfied: six<2.0.0dev,>=1.13.0 in c:\\users\\salta\\miniconda3\\lib\\site-packages (from google-cloud-bigquery) (1.15.0)\n",
      "Requirement already satisfied: proto-plus>=1.10.0 in c:\\users\\salta\\miniconda3\\lib\\site-packages (from google-cloud-bigquery) (1.11.0)\n",
      "Requirement already satisfied: protobuf>=3.12.0 in c:\\users\\salta\\miniconda3\\lib\\site-packages (from google-cloud-bigquery) (3.14.0)\n",
      "Requirement already satisfied: google-api-core[grpc]<2.0.0dev,>=1.23.0 in c:\\users\\salta\\miniconda3\\lib\\site-packages (from google-cloud-bigquery) (1.23.0)\n",
      "Requirement already satisfied: google-cloud-core<2.0dev,>=1.4.1 in c:\\users\\salta\\miniconda3\\lib\\site-packages (from google-cloud-bigquery) (1.4.3)\n",
      "Requirement already satisfied: requests<3.0.0dev,>=2.18.0 in c:\\users\\salta\\miniconda3\\lib\\site-packages (from google-api-core[grpc]<2.0.0dev,>=1.23.0->google-cloud-bigquery) (2.25.1)\n",
      "Requirement already satisfied: googleapis-common-protos<2.0dev,>=1.6.0 in c:\\users\\salta\\miniconda3\\lib\\site-packages (from google-api-core[grpc]<2.0.0dev,>=1.23.0->google-cloud-bigquery) (1.52.0)\n",
      "Requirement already satisfied: pytz in c:\\users\\salta\\miniconda3\\lib\\site-packages (from google-api-core[grpc]<2.0.0dev,>=1.23.0->google-cloud-bigquery) (2021.1)\n",
      "Requirement already satisfied: setuptools>=34.0.0 in c:\\users\\salta\\miniconda3\\lib\\site-packages (from google-api-core[grpc]<2.0.0dev,>=1.23.0->google-cloud-bigquery) (49.6.0.post20210108)\n",
      "Requirement already satisfied: google-auth<2.0dev,>=1.21.1 in c:\\users\\salta\\miniconda3\\lib\\site-packages (from google-api-core[grpc]<2.0.0dev,>=1.23.0->google-cloud-bigquery) (1.23.0)\n",
      "Requirement already satisfied: grpcio<2.0dev,>=1.29.0 in c:\\users\\salta\\miniconda3\\lib\\site-packages (from google-api-core[grpc]<2.0.0dev,>=1.23.0->google-cloud-bigquery) (1.33.2)\n",
      "Requirement already satisfied: rsa<5,>=3.1.4 in c:\\users\\salta\\miniconda3\\lib\\site-packages (from google-auth<2.0dev,>=1.21.1->google-api-core[grpc]<2.0.0dev,>=1.23.0->google-cloud-bigquery) (4.6)\n",
      "Requirement already satisfied: pyasn1-modules>=0.2.1 in c:\\users\\salta\\miniconda3\\lib\\site-packages (from google-auth<2.0dev,>=1.21.1->google-api-core[grpc]<2.0.0dev,>=1.23.0->google-cloud-bigquery) (0.2.8)\n",
      "Requirement already satisfied: cachetools<5.0,>=2.0.0 in c:\\users\\salta\\miniconda3\\lib\\site-packages (from google-auth<2.0dev,>=1.21.1->google-api-core[grpc]<2.0.0dev,>=1.23.0->google-cloud-bigquery) (4.1.1)\n",
      "Requirement already satisfied: google-crc32c<2.0dev,>=1.0 in c:\\users\\salta\\miniconda3\\lib\\site-packages (from google-resumable-media<2.0dev,>=0.6.0->google-cloud-bigquery) (1.0.0)\n",
      "Requirement already satisfied: pyasn1<0.5.0,>=0.4.6 in c:\\users\\salta\\miniconda3\\lib\\site-packages (from pyasn1-modules>=0.2.1->google-auth<2.0dev,>=1.21.1->google-api-core[grpc]<2.0.0dev,>=1.23.0->google-cloud-bigquery) (0.4.8)\n",
      "Requirement already satisfied: urllib3<1.27,>=1.21.1 in c:\\users\\salta\\miniconda3\\lib\\site-packages (from requests<3.0.0dev,>=2.18.0->google-api-core[grpc]<2.0.0dev,>=1.23.0->google-cloud-bigquery) (1.26.3)\n",
      "Requirement already satisfied: chardet<5,>=3.0.2 in c:\\users\\salta\\miniconda3\\lib\\site-packages (from requests<3.0.0dev,>=2.18.0->google-api-core[grpc]<2.0.0dev,>=1.23.0->google-cloud-bigquery) (4.0.0)\n",
      "Requirement already satisfied: certifi>=2017.4.17 in c:\\users\\salta\\miniconda3\\lib\\site-packages (from requests<3.0.0dev,>=2.18.0->google-api-core[grpc]<2.0.0dev,>=1.23.0->google-cloud-bigquery) (2020.12.5)\n",
      "Requirement already satisfied: idna<3,>=2.5 in c:\\users\\salta\\miniconda3\\lib\\site-packages (from requests<3.0.0dev,>=2.18.0->google-api-core[grpc]<2.0.0dev,>=1.23.0->google-cloud-bigquery) (2.10)\n"
     ]
    }
   ],
   "source": [
    "! pip install google-cloud-bigquery"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "# google cloud authentication\n",
    "from google.cloud import bigquery\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "# import modules\n",
    "import sys\n",
    "import matplotlib.pyplot as plt\n",
    "import pandas as pd\n",
    "import scipy\n",
    "from scipy import stats \n",
    "import numpy as np\n",
    "import json\n",
    "import statsmodels.stats.multitest as multi\n",
    "import matplotlib.pyplot as plt\n",
    "import math\n",
    "import ipywidgets as widgets\n",
    "import plotly\n",
    "import plotly.express as px\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Your browser has been opened to visit:\n",
      "\n",
      "    https://accounts.google.com/o/oauth2/auth?response_type=code&client_id=764086051850-6qr4p6gpi6hn506pt8ejuq83di341hur.apps.googleusercontent.com&redirect_uri=http%3A%2F%2Flocalhost%3A8085%2F&scope=openid+https%3A%2F%2Fwww.googleapis.com%2Fauth%2Fuserinfo.email+https%3A%2F%2Fwww.googleapis.com%2Fauth%2Fcloud-platform+https%3A%2F%2Fwww.googleapis.com%2Fauth%2Faccounts.reauth&state=kUr9qkiXSWEoWl4HVcjBcmcpX0dKjw&access_type=offline&code_challenge=lb58RLC2m0r6YYScV7YrmzIsQwNrLp9ay1ZOxWarNYU&code_challenge_method=S256\n",
      "\n",
      "\n",
      "Credentials saved to file: [C:\\Users\\salta\\AppData\\Roaming\\gcloud\\application_default_credentials.json]\n",
      "\n",
      "These credentials will be used by any library that requests Application Default Credentials (ADC).\n",
      "C:\\Program Files (x86)\\Google\\Cloud SDK\\google-cloud-sdk\\bin\\..\\lib\\third_party\\google\\auth\\_default.py:69: UserWarning: Your application has authenticated using end user credentials from Google Cloud SDK without a quota project. You might receive a \"quota exceeded\" or \"API not enabled\" error. We recommend you rerun `gcloud auth application-default login` and make sure a quota project is added. Or you can use service accounts instead. For more information about service accounts, see https://cloud.google.com/docs/authentication/\n",
      "  warnings.warn(_CLOUD_SDK_CREDENTIALS_WARNING)\n",
      "\n",
      "Quota project \"isb-cgc-04-0010\" was added to ADC which can be used by Google client libraries for billing and quota. Note that some services may still bill the project owning the resource.\n"
     ]
    }
   ],
   "source": [
    "# !gcloud auth login\n",
    "!gcloud auth application-default login"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Choose the project to be used for bigquery\n",
    "project_id='syntheticlethality'\n",
    "client = bigquery.Client(project_id) # Replace XXXXXXXX with your project ID"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "%load_ext google.cloud.bigquery"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Define a set of cancer-relevant tumor suppressor genes (TSGs)\n",
    "\n",
    "In this workflow, the search for relevant synthetic lethal interactions is seeded by defining a set of tumor suppressor genes (TSGs) of interest. There are various strategies for obtaining such a list, here we give an example of mining the [COSMIC Cancer Gene Census](https://cancer.sanger.ac.uk/census) for TSG annotations and then prioritizing the list based on driver status or frequency of alteration in a cancer type of interest.\n",
    "\n",
    "If you want to get the SL interactions for genes of interest, please add the genes to \"inputGenes\"."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bigquery tsg\n",
    "SELECT *\n",
    "FROM `isb-cgc.COSMIC_v90_grch38.Cancer_Gene_Census` \n",
    "WHERE Role_in_Cancer = \"TSG\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Gene_Symbol</th>\n",
       "      <th>Name</th>\n",
       "      <th>Entrez_GeneId</th>\n",
       "      <th>Genome_Location</th>\n",
       "      <th>Tier</th>\n",
       "      <th>Hallmark</th>\n",
       "      <th>Chr_Band</th>\n",
       "      <th>Somatic</th>\n",
       "      <th>Germline</th>\n",
       "      <th>Tumour_Types_Somatic</th>\n",
       "      <th>Tumour_Types_Germline</th>\n",
       "      <th>Cancer_Syndrome</th>\n",
       "      <th>Tissue_Type</th>\n",
       "      <th>Molecular_Genetics</th>\n",
       "      <th>Role_in_Cancer</th>\n",
       "      <th>Mutation_Types</th>\n",
       "      <th>Translocation_Partner</th>\n",
       "      <th>Other_Germline_Mut</th>\n",
       "      <th>Other_Syndrome</th>\n",
       "      <th>Synonyms</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ACVR2A</td>\n",
       "      <td>activin A receptor type 2A</td>\n",
       "      <td>92</td>\n",
       "      <td>2:147844517-147930824</td>\n",
       "      <td>1</td>\n",
       "      <td>None</td>\n",
       "      <td>23.10</td>\n",
       "      <td>yes</td>\n",
       "      <td>None</td>\n",
       "      <td>large intestine carcinoma, stomach carcinoma, ...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>E</td>\n",
       "      <td>Rec</td>\n",
       "      <td>TSG</td>\n",
       "      <td>Mis, N, F</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>92,ACTRII,ACVR2,ACVR2A,ENSG00000121989.14,P27037</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>APC</td>\n",
       "      <td>adenomatous polyposis of the colon gene</td>\n",
       "      <td>324</td>\n",
       "      <td>5:112737888-112846239</td>\n",
       "      <td>1</td>\n",
       "      <td>Yes</td>\n",
       "      <td>22.20</td>\n",
       "      <td>yes</td>\n",
       "      <td>yes</td>\n",
       "      <td>colorectal, pancreatic, desmoid, hepatoblastom...</td>\n",
       "      <td>colorectal, pancreatic, desmoid, hepatoblastom...</td>\n",
       "      <td>adenomatous polyposis coli; Turcot syndrome</td>\n",
       "      <td>E, M, O</td>\n",
       "      <td>Rec</td>\n",
       "      <td>TSG</td>\n",
       "      <td>D, Mis, N, F, S</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>324,APC,DP2,DP2.5,DP3,ENSG00000134982.16,P2505...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ARHGEF10</td>\n",
       "      <td>Rho guanine nucleotide exchange factor 10</td>\n",
       "      <td>9639</td>\n",
       "      <td>8:1824015-1958641</td>\n",
       "      <td>2</td>\n",
       "      <td>None</td>\n",
       "      <td>23.30</td>\n",
       "      <td>yes</td>\n",
       "      <td>None</td>\n",
       "      <td>colon cancer</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>E</td>\n",
       "      <td>None</td>\n",
       "      <td>TSG</td>\n",
       "      <td>D</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>9639,ARHGEF10,ENSG00000104728.15,Gef10,KIAA029...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ARHGEF10L</td>\n",
       "      <td>Rho guanine nucleotide exchange factor 10 like</td>\n",
       "      <td>55160</td>\n",
       "      <td>1:17539835-17697869</td>\n",
       "      <td>2</td>\n",
       "      <td>None</td>\n",
       "      <td>36.13</td>\n",
       "      <td>yes</td>\n",
       "      <td>None</td>\n",
       "      <td>lymphoma</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>L</td>\n",
       "      <td>None</td>\n",
       "      <td>TSG</td>\n",
       "      <td>D</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>55160,ARHGEF10L,ENSG00000074964.16,FLJ10521,KI...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ARID1B</td>\n",
       "      <td>AT rich interactive domain 1B</td>\n",
       "      <td>57492</td>\n",
       "      <td>6:156777374-157210779</td>\n",
       "      <td>1</td>\n",
       "      <td>None</td>\n",
       "      <td>25.30</td>\n",
       "      <td>yes</td>\n",
       "      <td>None</td>\n",
       "      <td>breast, hepatocellular carcinoma, clear cell o...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>E</td>\n",
       "      <td>Rec</td>\n",
       "      <td>TSG</td>\n",
       "      <td>Mis, F, N, O</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>57492,6A3-5,ARID1B,BAF250b,DAN15,ELD/OSA1,ENSG...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  Gene_Symbol                                            Name  Entrez_GeneId  \\\n",
       "0      ACVR2A                      activin A receptor type 2A             92   \n",
       "1         APC         adenomatous polyposis of the colon gene            324   \n",
       "2    ARHGEF10       Rho guanine nucleotide exchange factor 10           9639   \n",
       "3   ARHGEF10L  Rho guanine nucleotide exchange factor 10 like          55160   \n",
       "4      ARID1B                   AT rich interactive domain 1B          57492   \n",
       "\n",
       "         Genome_Location  Tier Hallmark  Chr_Band Somatic Germline  \\\n",
       "0  2:147844517-147930824     1     None     23.10     yes     None   \n",
       "1  5:112737888-112846239     1      Yes     22.20     yes      yes   \n",
       "2      8:1824015-1958641     2     None     23.30     yes     None   \n",
       "3    1:17539835-17697869     2     None     36.13     yes     None   \n",
       "4  6:156777374-157210779     1     None     25.30     yes     None   \n",
       "\n",
       "                                Tumour_Types_Somatic  \\\n",
       "0  large intestine carcinoma, stomach carcinoma, ...   \n",
       "1  colorectal, pancreatic, desmoid, hepatoblastom...   \n",
       "2                                       colon cancer   \n",
       "3                                           lymphoma   \n",
       "4  breast, hepatocellular carcinoma, clear cell o...   \n",
       "\n",
       "                               Tumour_Types_Germline  \\\n",
       "0                                               None   \n",
       "1  colorectal, pancreatic, desmoid, hepatoblastom...   \n",
       "2                                               None   \n",
       "3                                               None   \n",
       "4                                               None   \n",
       "\n",
       "                               Cancer_Syndrome Tissue_Type Molecular_Genetics  \\\n",
       "0                                         None           E                Rec   \n",
       "1  adenomatous polyposis coli; Turcot syndrome     E, M, O                Rec   \n",
       "2                                         None           E               None   \n",
       "3                                         None           L               None   \n",
       "4                                         None           E                Rec   \n",
       "\n",
       "  Role_in_Cancer   Mutation_Types Translocation_Partner Other_Germline_Mut  \\\n",
       "0            TSG        Mis, N, F                  None               None   \n",
       "1            TSG  D, Mis, N, F, S                  None               None   \n",
       "2            TSG                D                  None               None   \n",
       "3            TSG                D                  None               None   \n",
       "4            TSG     Mis, F, N, O                  None               None   \n",
       "\n",
       "  Other_Syndrome                                           Synonyms  \n",
       "0           None   92,ACTRII,ACVR2,ACVR2A,ENSG00000121989.14,P27037  \n",
       "1           None  324,APC,DP2,DP2.5,DP3,ENSG00000134982.16,P2505...  \n",
       "2           None  9639,ARHGEF10,ENSG00000104728.15,Gef10,KIAA029...  \n",
       "3           None  55160,ARHGEF10L,ENSG00000074964.16,FLJ10521,KI...  \n",
       "4           None  57492,6A3-5,ARID1B,BAF250b,DAN15,ELD/OSA1,ENSG...  "
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "tsg.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "\"'ACVR2A','APC','ARHGEF10','ARHGEF10L','ARID1B','ARID2','ASXL1','ASXL2','ATM','ATR','ATRX','AXIN1','AXIN2','B2M','BAP1','BARD1','BAX','BAZ1A','BRCA1','BRCA2','CASP3','CASP8','CDC73','CDH1','CDK12','CDKN1B','CDKN2A','CDKN2C','CHD2','CNTNAP2','CSMD3','CTCF','CYLD','DDX3X','DICER1','DROSHA','EED','ELF3','ETNK1','FAS','FAT1','FAT4','FBLN2','FBXW7','GPC5','GRIN2A','HNF1A','ID3','IGF2BP2','KDM5C','KEAP1','KLF6','KMT2C','LARP4B','LATS1','LATS2','LEPROTL1','LRP1B','MAX','MED12','MEN1','MLH1','MSH2','NF2','NFKBIE','PBRM1','PHF6','PHOX2B','PIK3R1','POLE','PPP2R1A','PRDM2','PTCH1','PTEN','PTPN13','PTPRB','PTPRD','PTPRT','RAD17','RB1','RBM10','RNF43','ROBO2','SDHA','SETD1B','SETD2','SFRP4','SH2B3','SIRPA','SMAD2','SMAD3','SMAD4','SMARCA4','SMARCB1','SOCS1','SOX21','SPEN','SPOP','STAG1','STAG2','STK11','SUFU','TET2','TGFBR2','TNFAIP3','TNFRSF14','TRAF7','TSC1','TSC2','USP44','VHL','WNK2','ZFHX3','ZMYM3','ZNRF3','ZRSR2','BLM','BRIP1','BUB1B','CHEK2','ERCC2','ERCC3','ERCC4','ERCC5','EXT2','FANCA','FANCC','FANCD2','FANCE','FANCF','FANCG','FEN1','FH','FLCN','MUTYH','NBN','PALB2','PMS2','PRF1','SBDS','SDHAF2','SDHB','SDHC','SDHD','SMARCE1','TMEM127','WRN','XPA','XPC','TENT5C','CBLB','DNMT3A','SMC1A','POT1','NTHL1','POLD1','POLG','FBXO11','PRDM1','PTPN6','CCNC','CNOT3','DNM2','PTPRC','RPL10','RPL5','NCOR1','SMARCD1','CASP9','MGMT','CEBPA','CDH10','PPP6C','NCOR2','MSH6','CPEB3','CUL3','N4BP2','RFWD3','AMER1','LZTR1','ATP2B3'\""
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# generate a list for inputGenes (Please go to the next block if you want to use your genes instead of tumor suppressor genes)\n",
    "tumor_suppressor_genes = tsg[\"Gene_Symbol\"].tolist() \n",
    "inputGenes = [\"'\"+x+\"'\" for x in tumor_suppressor_genes]\n",
    "inputGenes = ','.join(inputGenes)\n",
    "inputGenes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "# please skip this block if you want to keep using tumor suppressor genes as an input\n",
    "#inputGenes = \"\""
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Map Yeast Orthologs & Get SL insteractions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "sql = '''\n",
    "WITH\n",
    "--- Retreive YeastSymbols mapped to HumanSymbols for the input genes\n",
    "INPUT_H2Y AS (\n",
    "  SELECT YeastSymbol\n",
    "    FROM `syntheticlethality.gene_information.human2Yeast`\n",
    "   WHERE HumanSymbol IN (__INPUTGENES__) AND\n",
    "         AlgorithmsMatch >= __ALGORITHMCUTOFF__\n",
    "),\n",
    "\n",
    "--- Identify protein-protein interactions using the YeastSymbols (left match)\n",
    "Yeast_ITX1 AS (\n",
    "  SELECT UPPER(Query_allele_name)       AS Interactor1, \n",
    "         UPPER(Array_allele_name)       AS Interactor2,\n",
    "         Genetic_interaction_score_____ AS Interaction_score,\n",
    "         P_value\n",
    "    FROM `syntheticlethality.CellMap.CellMap`\n",
    "   WHERE (Genetic_interaction_score_____ < __SCORECUTOFF__ AND P_value < __PvalueCUTOFF__) AND\n",
    "         (UPPER(Query_allele_name) IN (SELECT YeastSymbol FROM INPUT_H2Y))\n",
    "   \n",
    "),\n",
    "\n",
    "--- Identify protein-protein interactions using the YeastSymbols (right match)\n",
    "Yeast_ITX2 AS (\n",
    "  SELECT UPPER(Array_allele_name)       AS Interactor1, \n",
    "         UPPER(Query_allele_name)       AS Interactor2,\n",
    "         Genetic_interaction_score_____ AS Interaction_score,\n",
    "         P_value\n",
    "    FROM `syntheticlethality.CellMap.CellMap`\n",
    "   WHERE (Genetic_interaction_score_____ < __SCORECUTOFF__ AND P_value < __PvalueCUTOFF__) AND\n",
    "         (UPPER(Array_allele_name) IN (SELECT YeastSymbol FROM INPUT_H2Y))\n",
    "   \n",
    "),\n",
    "\n",
    "--- Union interaction tables\n",
    "Union_ITX AS (\n",
    "  SELECT * FROM Yeast_ITX1\n",
    "   UNION ALL\n",
    "  SELECT * FROM Yeast_ITX2\n",
    ")\n",
    "\n",
    "--- Convert YeastSymbols to HumanSymbols in the protein-protein interations\n",
    "SELECT DISTINCT \n",
    "       GINFO1.EntrezID        AS EntrezID_Input,\n",
    "       H2Y1.HumanSymbol       AS Gene_Input,\n",
    "---       Add if you want to know what yeast genes are involved\n",
    "---       YITX.Interactor1       AS Gene_Input_Yeast,\n",
    "       GINFO2.EntrezID        AS EntrezID_SL_Candidate,\n",
    "       H2Y2.HumanSymbol       AS Gene_SL_Candidate,\n",
    "---       Add if you want to know what yeast genes are involved\n",
    "---       YITX.Interactor2       AS Gene_SL_Candidate_Yeast,\n",
    "       YITX.Interaction_score AS Interaction_score,\n",
    "       YITX.P_value           AS P_value\n",
    "       \n",
    "  FROM Union_ITX AS YITX\n",
    "       LEFT JOIN `syntheticlethality.gene_information.human2Yeast`                       AS H2Y1   ON YITX.Interactor1 = H2Y1.YeastSymbol\n",
    "       LEFT JOIN `syntheticlethality.gene_information.human2Yeast`                       AS H2Y2   ON YITX.Interactor2 = H2Y2.YeastSymbol\n",
    "       LEFT JOIN `syntheticlethality.gene_information.gene_info_human_HGNC` AS GINFO1 ON H2Y1.HumanID = GINFO1.HGNCID\n",
    "       LEFT JOIN `syntheticlethality.gene_information.gene_info_human_HGNC` AS GINFO2 ON H2Y2.HumanID = GINFO2.HGNCID\n",
    "       \n",
    " WHERE (H2Y1.HumanSymbol IS NOT NULL AND YITX.Interactor1 IS NOT NULL) AND\n",
    "       (H2Y2.HumanSymbol IS NOT NULL AND YITX.Interactor2 IS NOT NULL)\n",
    "\n",
    "'''\n",
    "# select the thresholds to be used\n",
    "cutoff_algorithmMatchNo = \"3\"\n",
    "cutoff_score = \"-0.35\"\n",
    "cutoff_p = \"0.01\"\n",
    "\n",
    "sql = sql.replace(\"__INPUTGENES__\", inputGenes)\n",
    "sql = sql.replace(\"__ALGORITHMCUTOFF__\", cutoff_algorithmMatchNo)\n",
    "sql = sql.replace(\"__SCORECUTOFF__\", cutoff_score)\n",
    "sql = sql.replace(\"__PvalueCUTOFF__\", cutoff_p)\n",
    "\n",
    "res = client.query(sql).to_dataframe()\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Get Yeast SL Interactions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>EntrezID_Input</th>\n",
       "      <th>Gene_Input</th>\n",
       "      <th>EntrezID_SL_Candidate</th>\n",
       "      <th>Gene_SL_Candidate</th>\n",
       "      <th>Interaction_score</th>\n",
       "      <th>P_value</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>4436.0</td>\n",
       "      <td>MSH2</td>\n",
       "      <td>271</td>\n",
       "      <td>AMPD2</td>\n",
       "      <td>-0.4561</td>\n",
       "      <td>5.070000e-11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>4436.0</td>\n",
       "      <td>MSH2</td>\n",
       "      <td>272</td>\n",
       "      <td>AMPD3</td>\n",
       "      <td>-0.4561</td>\n",
       "      <td>5.070000e-11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>4436.0</td>\n",
       "      <td>MSH2</td>\n",
       "      <td>270</td>\n",
       "      <td>AMPD1</td>\n",
       "      <td>-0.4561</td>\n",
       "      <td>5.070000e-11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>9739.0</td>\n",
       "      <td>SETD1A</td>\n",
       "      <td>55666</td>\n",
       "      <td>NPLOC4</td>\n",
       "      <td>-0.3786</td>\n",
       "      <td>1.475000e-16</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>23067.0</td>\n",
       "      <td>SETD1B</td>\n",
       "      <td>55666</td>\n",
       "      <td>NPLOC4</td>\n",
       "      <td>-0.3786</td>\n",
       "      <td>1.475000e-16</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>452</th>\n",
       "      <td>2237.0</td>\n",
       "      <td>FEN1</td>\n",
       "      <td>1195</td>\n",
       "      <td>CLK1</td>\n",
       "      <td>-0.4042</td>\n",
       "      <td>3.388000e-11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>453</th>\n",
       "      <td>2237.0</td>\n",
       "      <td>FEN1</td>\n",
       "      <td>1196</td>\n",
       "      <td>CLK2</td>\n",
       "      <td>-0.4042</td>\n",
       "      <td>3.388000e-11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>454</th>\n",
       "      <td>2237.0</td>\n",
       "      <td>FEN1</td>\n",
       "      <td>1198</td>\n",
       "      <td>CLK3</td>\n",
       "      <td>-0.4042</td>\n",
       "      <td>3.388000e-11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>455</th>\n",
       "      <td>2237.0</td>\n",
       "      <td>FEN1</td>\n",
       "      <td>57396</td>\n",
       "      <td>CLK4</td>\n",
       "      <td>-0.4042</td>\n",
       "      <td>3.388000e-11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>456</th>\n",
       "      <td>79577.0</td>\n",
       "      <td>CDC73</td>\n",
       "      <td>129563</td>\n",
       "      <td>DIS3L2</td>\n",
       "      <td>-0.4564</td>\n",
       "      <td>6.903000e-11</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>457 rows × 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     EntrezID_Input Gene_Input  EntrezID_SL_Candidate Gene_SL_Candidate  \\\n",
       "0            4436.0       MSH2                    271             AMPD2   \n",
       "1            4436.0       MSH2                    272             AMPD3   \n",
       "2            4436.0       MSH2                    270             AMPD1   \n",
       "3            9739.0     SETD1A                  55666            NPLOC4   \n",
       "4           23067.0     SETD1B                  55666            NPLOC4   \n",
       "..              ...        ...                    ...               ...   \n",
       "452          2237.0       FEN1                   1195              CLK1   \n",
       "453          2237.0       FEN1                   1196              CLK2   \n",
       "454          2237.0       FEN1                   1198              CLK3   \n",
       "455          2237.0       FEN1                  57396              CLK4   \n",
       "456         79577.0      CDC73                 129563            DIS3L2   \n",
       "\n",
       "     Interaction_score       P_value  \n",
       "0              -0.4561  5.070000e-11  \n",
       "1              -0.4561  5.070000e-11  \n",
       "2              -0.4561  5.070000e-11  \n",
       "3              -0.3786  1.475000e-16  \n",
       "4              -0.3786  1.475000e-16  \n",
       "..                 ...           ...  \n",
       "452            -0.4042  3.388000e-11  \n",
       "453            -0.4042  3.388000e-11  \n",
       "454            -0.4042  3.388000e-11  \n",
       "455            -0.4042  3.388000e-11  \n",
       "456            -0.4564  6.903000e-11  \n",
       "\n",
       "[457 rows x 6 columns]"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# shore the SL partner genes for the input genes\n",
    "res\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Write to file & bigQuery Table"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "res.to_csv(path_or_buf='conserved_SL_output.csv', index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Assess Gene Druggability"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Available gene categories in current release:\n",
    "!curl http://dgidb.org/api/v2/gene_categories.json"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Get all genes that are in the druggable genome\n",
    "# parameterization\n",
    "# category_of_interest=['kinase', 'phospolipase']\n",
    "!curl http://dgidb.org/api/v2/genes_in_category.json?category=druggable%20genome"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.8"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
