{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Setup for Conservation-Based Synthetic Lethal Discovery Pipeline\n",
    "\n",
    "This notebook outlines our approach to pre-processing publicly-available datasets to facillitate the discovery of synthetic lethal interactions conserved between human and yeast speicies. The wrangled data are uploaded to bigquery to facillitate .sql-based querying (see pipline notebook conservedSL.ipynb)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 89,
   "metadata": {},
   "outputs": [],
   "source": [
    "# load required modules\n",
    "import pandas as pd"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Reference Dataset for Ortholog Mapping\n",
    "\n",
    "Our source for human to yeast ortholog mapping is the [Alliance of Genome Resources (AGR)](https://www.alliancegenome.org) - Release 3.0.1. \n",
    "Full download is found by navigating from the home page to Data -> Downloads -> Orthology. We are using the 'Alliance combined orthology data' found on the [downloads](https://www.alliancegenome.org/downloads#orthology) page. "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Option 1: pre-download combined orthology data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 90,
   "metadata": {},
   "outputs": [],
   "source": [
    "# local file path to download location\n",
    "# AGR_ORTHOLOGS = 'data/sgd_alliance/ORTHOLOGY-ALLIANCE_COMBINED_4.tsv' "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Option 2: web-based download of combined orthology data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 91,
   "metadata": {},
   "outputs": [],
   "source": [
    "# url to file download\n",
    "AGR_ORTHOLOGS = 'http://download.alliancegenome.org/3.0.1/ORTHOLOGY-ALLIANCE/COMBINED/ORTHOLOGY-ALLIANCE_COMBINED_4.tsv'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 92,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Gene1ID</th>\n",
       "      <th>Gene1Symbol</th>\n",
       "      <th>Gene1SpeciesTaxonID</th>\n",
       "      <th>Gene1SpeciesName</th>\n",
       "      <th>Gene2ID</th>\n",
       "      <th>Gene2Symbol</th>\n",
       "      <th>Gene2SpeciesTaxonID</th>\n",
       "      <th>Gene2SpeciesName</th>\n",
       "      <th>Algorithms</th>\n",
       "      <th>AlgorithmsMatch</th>\n",
       "      <th>OutOfAlgorithms</th>\n",
       "      <th>IsBestScore</th>\n",
       "      <th>IsBestRevScore</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>HGNC:28697</td>\n",
       "      <td>ZADH2</td>\n",
       "      <td>NCBITaxon:9606</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>SGD:S000004937</td>\n",
       "      <td>ADH6</td>\n",
       "      <td>NCBITaxon:559292</td>\n",
       "      <td>Saccharomyces cerevisiae</td>\n",
       "      <td>OMA|Ensembl Compara</td>\n",
       "      <td>2</td>\n",
       "      <td>10</td>\n",
       "      <td>Yes</td>\n",
       "      <td>Yes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>FB:FBgn0033188</td>\n",
       "      <td>Drat</td>\n",
       "      <td>NCBITaxon:7227</td>\n",
       "      <td>Drosophila melanogaster</td>\n",
       "      <td>SGD:S000004937</td>\n",
       "      <td>ADH6</td>\n",
       "      <td>NCBITaxon:559292</td>\n",
       "      <td>Saccharomyces cerevisiae</td>\n",
       "      <td>PANTHER|Ensembl Compara</td>\n",
       "      <td>2</td>\n",
       "      <td>10</td>\n",
       "      <td>Yes</td>\n",
       "      <td>Yes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>MGI:2448516</td>\n",
       "      <td>Cryzl2</td>\n",
       "      <td>NCBITaxon:10090</td>\n",
       "      <td>Mus musculus</td>\n",
       "      <td>SGD:S000004937</td>\n",
       "      <td>ADH6</td>\n",
       "      <td>NCBITaxon:559292</td>\n",
       "      <td>Saccharomyces cerevisiae</td>\n",
       "      <td>OMA|Ensembl Compara</td>\n",
       "      <td>2</td>\n",
       "      <td>10</td>\n",
       "      <td>Yes</td>\n",
       "      <td>Yes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>WB:WBGene00014096</td>\n",
       "      <td>ZK829.7</td>\n",
       "      <td>NCBITaxon:6239</td>\n",
       "      <td>Caenorhabditis elegans</td>\n",
       "      <td>SGD:S000004937</td>\n",
       "      <td>ADH6</td>\n",
       "      <td>NCBITaxon:559292</td>\n",
       "      <td>Saccharomyces cerevisiae</td>\n",
       "      <td>PANTHER|Ensembl Compara</td>\n",
       "      <td>2</td>\n",
       "      <td>10</td>\n",
       "      <td>Yes</td>\n",
       "      <td>Yes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>RGD:1304982</td>\n",
       "      <td>Tp53i3</td>\n",
       "      <td>NCBITaxon:10116</td>\n",
       "      <td>Rattus norvegicus</td>\n",
       "      <td>SGD:S000004937</td>\n",
       "      <td>ADH6</td>\n",
       "      <td>NCBITaxon:559292</td>\n",
       "      <td>Saccharomyces cerevisiae</td>\n",
       "      <td>OMA|Ensembl Compara</td>\n",
       "      <td>2</td>\n",
       "      <td>9</td>\n",
       "      <td>Yes</td>\n",
       "      <td>Yes</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "             Gene1ID Gene1Symbol Gene1SpeciesTaxonID         Gene1SpeciesName  \\\n",
       "0         HGNC:28697       ZADH2      NCBITaxon:9606             Homo sapiens   \n",
       "1     FB:FBgn0033188        Drat      NCBITaxon:7227  Drosophila melanogaster   \n",
       "2        MGI:2448516      Cryzl2     NCBITaxon:10090             Mus musculus   \n",
       "3  WB:WBGene00014096     ZK829.7      NCBITaxon:6239   Caenorhabditis elegans   \n",
       "4        RGD:1304982      Tp53i3     NCBITaxon:10116        Rattus norvegicus   \n",
       "\n",
       "          Gene2ID Gene2Symbol Gene2SpeciesTaxonID          Gene2SpeciesName  \\\n",
       "0  SGD:S000004937        ADH6    NCBITaxon:559292  Saccharomyces cerevisiae   \n",
       "1  SGD:S000004937        ADH6    NCBITaxon:559292  Saccharomyces cerevisiae   \n",
       "2  SGD:S000004937        ADH6    NCBITaxon:559292  Saccharomyces cerevisiae   \n",
       "3  SGD:S000004937        ADH6    NCBITaxon:559292  Saccharomyces cerevisiae   \n",
       "4  SGD:S000004937        ADH6    NCBITaxon:559292  Saccharomyces cerevisiae   \n",
       "\n",
       "                Algorithms  AlgorithmsMatch  OutOfAlgorithms IsBestScore  \\\n",
       "0      OMA|Ensembl Compara                2               10         Yes   \n",
       "1  PANTHER|Ensembl Compara                2               10         Yes   \n",
       "2      OMA|Ensembl Compara                2               10         Yes   \n",
       "3  PANTHER|Ensembl Compara                2               10         Yes   \n",
       "4      OMA|Ensembl Compara                2                9         Yes   \n",
       "\n",
       "  IsBestRevScore  \n",
       "0            Yes  \n",
       "1            Yes  \n",
       "2            Yes  \n",
       "3            Yes  \n",
       "4            Yes  "
      ]
     },
     "execution_count": 92,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ortholog_table = pd.read_csv(AGR_ORTHOLOGS, sep='\\t', comment='#')\n",
    "ortholog_table.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Data Cleaning"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 93,
   "metadata": {},
   "outputs": [],
   "source": [
    "# data cleaning - remove identifier names within each field\n",
    "# e..g HGNC:28697 --> 28697\n",
    "headers = ortholog_table.columns.values\n",
    "cols_to_clean = headers[['ID' in s for s in headers]]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 94,
   "metadata": {},
   "outputs": [],
   "source": [
    "def remove_column_annotation(gene_info):\n",
    "    \"\"\" Clean up identifier columns\n",
    "    Remove the exact source info and keep true identifiers \n",
    "    e.g. remove 'MGI:' and keep mouse gene identifier\n",
    "    \"\"\"\n",
    "    return(gene_info.str.split(':').str[-1])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 95,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Gene1ID</th>\n",
       "      <th>Gene1Symbol</th>\n",
       "      <th>Gene1SpeciesTaxonID</th>\n",
       "      <th>Gene1SpeciesName</th>\n",
       "      <th>Gene2ID</th>\n",
       "      <th>Gene2Symbol</th>\n",
       "      <th>Gene2SpeciesTaxonID</th>\n",
       "      <th>Gene2SpeciesName</th>\n",
       "      <th>Algorithms</th>\n",
       "      <th>AlgorithmsMatch</th>\n",
       "      <th>OutOfAlgorithms</th>\n",
       "      <th>IsBestScore</th>\n",
       "      <th>IsBestRevScore</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>28697</td>\n",
       "      <td>ZADH2</td>\n",
       "      <td>9606</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>S000004937</td>\n",
       "      <td>ADH6</td>\n",
       "      <td>559292</td>\n",
       "      <td>Saccharomyces cerevisiae</td>\n",
       "      <td>OMA|Ensembl Compara</td>\n",
       "      <td>2</td>\n",
       "      <td>10</td>\n",
       "      <td>Yes</td>\n",
       "      <td>Yes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>FBgn0033188</td>\n",
       "      <td>Drat</td>\n",
       "      <td>7227</td>\n",
       "      <td>Drosophila melanogaster</td>\n",
       "      <td>S000004937</td>\n",
       "      <td>ADH6</td>\n",
       "      <td>559292</td>\n",
       "      <td>Saccharomyces cerevisiae</td>\n",
       "      <td>PANTHER|Ensembl Compara</td>\n",
       "      <td>2</td>\n",
       "      <td>10</td>\n",
       "      <td>Yes</td>\n",
       "      <td>Yes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2448516</td>\n",
       "      <td>Cryzl2</td>\n",
       "      <td>10090</td>\n",
       "      <td>Mus musculus</td>\n",
       "      <td>S000004937</td>\n",
       "      <td>ADH6</td>\n",
       "      <td>559292</td>\n",
       "      <td>Saccharomyces cerevisiae</td>\n",
       "      <td>OMA|Ensembl Compara</td>\n",
       "      <td>2</td>\n",
       "      <td>10</td>\n",
       "      <td>Yes</td>\n",
       "      <td>Yes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>WBGene00014096</td>\n",
       "      <td>ZK829.7</td>\n",
       "      <td>6239</td>\n",
       "      <td>Caenorhabditis elegans</td>\n",
       "      <td>S000004937</td>\n",
       "      <td>ADH6</td>\n",
       "      <td>559292</td>\n",
       "      <td>Saccharomyces cerevisiae</td>\n",
       "      <td>PANTHER|Ensembl Compara</td>\n",
       "      <td>2</td>\n",
       "      <td>10</td>\n",
       "      <td>Yes</td>\n",
       "      <td>Yes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1304982</td>\n",
       "      <td>Tp53i3</td>\n",
       "      <td>10116</td>\n",
       "      <td>Rattus norvegicus</td>\n",
       "      <td>S000004937</td>\n",
       "      <td>ADH6</td>\n",
       "      <td>559292</td>\n",
       "      <td>Saccharomyces cerevisiae</td>\n",
       "      <td>OMA|Ensembl Compara</td>\n",
       "      <td>2</td>\n",
       "      <td>9</td>\n",
       "      <td>Yes</td>\n",
       "      <td>Yes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>560951</th>\n",
       "      <td>12340</td>\n",
       "      <td>TRPS1</td>\n",
       "      <td>9606</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>1927616</td>\n",
       "      <td>Trps1</td>\n",
       "      <td>10090</td>\n",
       "      <td>Mus musculus</td>\n",
       "      <td>PANTHER|HGNC|PhylomeDB|OrthoFinder|InParanoid|...</td>\n",
       "      <td>9</td>\n",
       "      <td>11</td>\n",
       "      <td>Yes</td>\n",
       "      <td>Yes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>560952</th>\n",
       "      <td>ZDB-GENE-030131-8404</td>\n",
       "      <td>trps1</td>\n",
       "      <td>7955</td>\n",
       "      <td>Danio rerio</td>\n",
       "      <td>1927616</td>\n",
       "      <td>Trps1</td>\n",
       "      <td>10090</td>\n",
       "      <td>Mus musculus</td>\n",
       "      <td>ZFIN|PANTHER|PhylomeDB|OrthoFinder|InParanoid|...</td>\n",
       "      <td>7</td>\n",
       "      <td>11</td>\n",
       "      <td>Yes</td>\n",
       "      <td>Yes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>560953</th>\n",
       "      <td>WBGene00017033</td>\n",
       "      <td>nekl-4</td>\n",
       "      <td>6239</td>\n",
       "      <td>Caenorhabditis elegans</td>\n",
       "      <td>2685128</td>\n",
       "      <td>Nek10</td>\n",
       "      <td>10090</td>\n",
       "      <td>Mus musculus</td>\n",
       "      <td>PANTHER|PhylomeDB|OrthoFinder|InParanoid|Ortho...</td>\n",
       "      <td>8</td>\n",
       "      <td>10</td>\n",
       "      <td>Yes</td>\n",
       "      <td>Yes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>560954</th>\n",
       "      <td>18592</td>\n",
       "      <td>NEK10</td>\n",
       "      <td>9606</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>2685128</td>\n",
       "      <td>Nek10</td>\n",
       "      <td>10090</td>\n",
       "      <td>Mus musculus</td>\n",
       "      <td>PANTHER|HGNC|Ensembl Compara|OMA|PhylomeDB|Ort...</td>\n",
       "      <td>11</td>\n",
       "      <td>11</td>\n",
       "      <td>Yes</td>\n",
       "      <td>Yes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>560955</th>\n",
       "      <td>ZDB-GENE-130530-570</td>\n",
       "      <td>nek10</td>\n",
       "      <td>7955</td>\n",
       "      <td>Danio rerio</td>\n",
       "      <td>2685128</td>\n",
       "      <td>Nek10</td>\n",
       "      <td>10090</td>\n",
       "      <td>Mus musculus</td>\n",
       "      <td>ZFIN|PANTHER|PhylomeDB|OrthoFinder|InParanoid|...</td>\n",
       "      <td>8</td>\n",
       "      <td>11</td>\n",
       "      <td>Yes</td>\n",
       "      <td>Yes</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>560956 rows × 13 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                     Gene1ID Gene1Symbol Gene1SpeciesTaxonID  \\\n",
       "0                      28697       ZADH2                9606   \n",
       "1                FBgn0033188        Drat                7227   \n",
       "2                    2448516      Cryzl2               10090   \n",
       "3             WBGene00014096     ZK829.7                6239   \n",
       "4                    1304982      Tp53i3               10116   \n",
       "...                      ...         ...                 ...   \n",
       "560951                 12340       TRPS1                9606   \n",
       "560952  ZDB-GENE-030131-8404       trps1                7955   \n",
       "560953        WBGene00017033      nekl-4                6239   \n",
       "560954                 18592       NEK10                9606   \n",
       "560955   ZDB-GENE-130530-570       nek10                7955   \n",
       "\n",
       "               Gene1SpeciesName     Gene2ID Gene2Symbol Gene2SpeciesTaxonID  \\\n",
       "0                  Homo sapiens  S000004937        ADH6              559292   \n",
       "1       Drosophila melanogaster  S000004937        ADH6              559292   \n",
       "2                  Mus musculus  S000004937        ADH6              559292   \n",
       "3        Caenorhabditis elegans  S000004937        ADH6              559292   \n",
       "4             Rattus norvegicus  S000004937        ADH6              559292   \n",
       "...                         ...         ...         ...                 ...   \n",
       "560951             Homo sapiens     1927616       Trps1               10090   \n",
       "560952              Danio rerio     1927616       Trps1               10090   \n",
       "560953   Caenorhabditis elegans     2685128       Nek10               10090   \n",
       "560954             Homo sapiens     2685128       Nek10               10090   \n",
       "560955              Danio rerio     2685128       Nek10               10090   \n",
       "\n",
       "                Gene2SpeciesName  \\\n",
       "0       Saccharomyces cerevisiae   \n",
       "1       Saccharomyces cerevisiae   \n",
       "2       Saccharomyces cerevisiae   \n",
       "3       Saccharomyces cerevisiae   \n",
       "4       Saccharomyces cerevisiae   \n",
       "...                          ...   \n",
       "560951              Mus musculus   \n",
       "560952              Mus musculus   \n",
       "560953              Mus musculus   \n",
       "560954              Mus musculus   \n",
       "560955              Mus musculus   \n",
       "\n",
       "                                               Algorithms  AlgorithmsMatch  \\\n",
       "0                                     OMA|Ensembl Compara                2   \n",
       "1                                 PANTHER|Ensembl Compara                2   \n",
       "2                                     OMA|Ensembl Compara                2   \n",
       "3                                 PANTHER|Ensembl Compara                2   \n",
       "4                                     OMA|Ensembl Compara                2   \n",
       "...                                                   ...              ...   \n",
       "560951  PANTHER|HGNC|PhylomeDB|OrthoFinder|InParanoid|...                9   \n",
       "560952  ZFIN|PANTHER|PhylomeDB|OrthoFinder|InParanoid|...                7   \n",
       "560953  PANTHER|PhylomeDB|OrthoFinder|InParanoid|Ortho...                8   \n",
       "560954  PANTHER|HGNC|Ensembl Compara|OMA|PhylomeDB|Ort...               11   \n",
       "560955  ZFIN|PANTHER|PhylomeDB|OrthoFinder|InParanoid|...                8   \n",
       "\n",
       "        OutOfAlgorithms IsBestScore IsBestRevScore  \n",
       "0                    10         Yes            Yes  \n",
       "1                    10         Yes            Yes  \n",
       "2                    10         Yes            Yes  \n",
       "3                    10         Yes            Yes  \n",
       "4                     9         Yes            Yes  \n",
       "...                 ...         ...            ...  \n",
       "560951               11         Yes            Yes  \n",
       "560952               11         Yes            Yes  \n",
       "560953               10         Yes            Yes  \n",
       "560954               11         Yes            Yes  \n",
       "560955               11         Yes            Yes  \n",
       "\n",
       "[560956 rows x 13 columns]"
      ]
     },
     "execution_count": 95,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df = ortholog_table.apply(lambda x: remove_column_annotation(x) if x.name in cols_to_clean else x)\n",
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 96,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Gene1ID</th>\n",
       "      <th>Gene1Symbol</th>\n",
       "      <th>Gene1SpeciesTaxonID</th>\n",
       "      <th>Gene1SpeciesName</th>\n",
       "      <th>Gene2ID</th>\n",
       "      <th>Gene2Symbol</th>\n",
       "      <th>Gene2SpeciesTaxonID</th>\n",
       "      <th>Gene2SpeciesName</th>\n",
       "      <th>Algorithms</th>\n",
       "      <th>AlgorithmsMatch</th>\n",
       "      <th>OutOfAlgorithms</th>\n",
       "      <th>IsBestScore</th>\n",
       "      <th>IsBestRevScore</th>\n",
       "      <th>AlgorithmsMatchPerc</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>28697</td>\n",
       "      <td>ZADH2</td>\n",
       "      <td>9606</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>S000004937</td>\n",
       "      <td>ADH6</td>\n",
       "      <td>559292</td>\n",
       "      <td>Saccharomyces cerevisiae</td>\n",
       "      <td>OMA|Ensembl Compara</td>\n",
       "      <td>2</td>\n",
       "      <td>10</td>\n",
       "      <td>Yes</td>\n",
       "      <td>Yes</td>\n",
       "      <td>0.200000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>FBgn0033188</td>\n",
       "      <td>Drat</td>\n",
       "      <td>7227</td>\n",
       "      <td>Drosophila melanogaster</td>\n",
       "      <td>S000004937</td>\n",
       "      <td>ADH6</td>\n",
       "      <td>559292</td>\n",
       "      <td>Saccharomyces cerevisiae</td>\n",
       "      <td>PANTHER|Ensembl Compara</td>\n",
       "      <td>2</td>\n",
       "      <td>10</td>\n",
       "      <td>Yes</td>\n",
       "      <td>Yes</td>\n",
       "      <td>0.200000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2448516</td>\n",
       "      <td>Cryzl2</td>\n",
       "      <td>10090</td>\n",
       "      <td>Mus musculus</td>\n",
       "      <td>S000004937</td>\n",
       "      <td>ADH6</td>\n",
       "      <td>559292</td>\n",
       "      <td>Saccharomyces cerevisiae</td>\n",
       "      <td>OMA|Ensembl Compara</td>\n",
       "      <td>2</td>\n",
       "      <td>10</td>\n",
       "      <td>Yes</td>\n",
       "      <td>Yes</td>\n",
       "      <td>0.200000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>WBGene00014096</td>\n",
       "      <td>ZK829.7</td>\n",
       "      <td>6239</td>\n",
       "      <td>Caenorhabditis elegans</td>\n",
       "      <td>S000004937</td>\n",
       "      <td>ADH6</td>\n",
       "      <td>559292</td>\n",
       "      <td>Saccharomyces cerevisiae</td>\n",
       "      <td>PANTHER|Ensembl Compara</td>\n",
       "      <td>2</td>\n",
       "      <td>10</td>\n",
       "      <td>Yes</td>\n",
       "      <td>Yes</td>\n",
       "      <td>0.200000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1304982</td>\n",
       "      <td>Tp53i3</td>\n",
       "      <td>10116</td>\n",
       "      <td>Rattus norvegicus</td>\n",
       "      <td>S000004937</td>\n",
       "      <td>ADH6</td>\n",
       "      <td>559292</td>\n",
       "      <td>Saccharomyces cerevisiae</td>\n",
       "      <td>OMA|Ensembl Compara</td>\n",
       "      <td>2</td>\n",
       "      <td>9</td>\n",
       "      <td>Yes</td>\n",
       "      <td>Yes</td>\n",
       "      <td>0.222222</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>560951</th>\n",
       "      <td>12340</td>\n",
       "      <td>TRPS1</td>\n",
       "      <td>9606</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>1927616</td>\n",
       "      <td>Trps1</td>\n",
       "      <td>10090</td>\n",
       "      <td>Mus musculus</td>\n",
       "      <td>PANTHER|HGNC|PhylomeDB|OrthoFinder|InParanoid|...</td>\n",
       "      <td>9</td>\n",
       "      <td>11</td>\n",
       "      <td>Yes</td>\n",
       "      <td>Yes</td>\n",
       "      <td>0.818182</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>560952</th>\n",
       "      <td>ZDB-GENE-030131-8404</td>\n",
       "      <td>trps1</td>\n",
       "      <td>7955</td>\n",
       "      <td>Danio rerio</td>\n",
       "      <td>1927616</td>\n",
       "      <td>Trps1</td>\n",
       "      <td>10090</td>\n",
       "      <td>Mus musculus</td>\n",
       "      <td>ZFIN|PANTHER|PhylomeDB|OrthoFinder|InParanoid|...</td>\n",
       "      <td>7</td>\n",
       "      <td>11</td>\n",
       "      <td>Yes</td>\n",
       "      <td>Yes</td>\n",
       "      <td>0.636364</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>560953</th>\n",
       "      <td>WBGene00017033</td>\n",
       "      <td>nekl-4</td>\n",
       "      <td>6239</td>\n",
       "      <td>Caenorhabditis elegans</td>\n",
       "      <td>2685128</td>\n",
       "      <td>Nek10</td>\n",
       "      <td>10090</td>\n",
       "      <td>Mus musculus</td>\n",
       "      <td>PANTHER|PhylomeDB|OrthoFinder|InParanoid|Ortho...</td>\n",
       "      <td>8</td>\n",
       "      <td>10</td>\n",
       "      <td>Yes</td>\n",
       "      <td>Yes</td>\n",
       "      <td>0.800000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>560954</th>\n",
       "      <td>18592</td>\n",
       "      <td>NEK10</td>\n",
       "      <td>9606</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>2685128</td>\n",
       "      <td>Nek10</td>\n",
       "      <td>10090</td>\n",
       "      <td>Mus musculus</td>\n",
       "      <td>PANTHER|HGNC|Ensembl Compara|OMA|PhylomeDB|Ort...</td>\n",
       "      <td>11</td>\n",
       "      <td>11</td>\n",
       "      <td>Yes</td>\n",
       "      <td>Yes</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>560955</th>\n",
       "      <td>ZDB-GENE-130530-570</td>\n",
       "      <td>nek10</td>\n",
       "      <td>7955</td>\n",
       "      <td>Danio rerio</td>\n",
       "      <td>2685128</td>\n",
       "      <td>Nek10</td>\n",
       "      <td>10090</td>\n",
       "      <td>Mus musculus</td>\n",
       "      <td>ZFIN|PANTHER|PhylomeDB|OrthoFinder|InParanoid|...</td>\n",
       "      <td>8</td>\n",
       "      <td>11</td>\n",
       "      <td>Yes</td>\n",
       "      <td>Yes</td>\n",
       "      <td>0.727273</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>560956 rows × 14 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                     Gene1ID Gene1Symbol Gene1SpeciesTaxonID  \\\n",
       "0                      28697       ZADH2                9606   \n",
       "1                FBgn0033188        Drat                7227   \n",
       "2                    2448516      Cryzl2               10090   \n",
       "3             WBGene00014096     ZK829.7                6239   \n",
       "4                    1304982      Tp53i3               10116   \n",
       "...                      ...         ...                 ...   \n",
       "560951                 12340       TRPS1                9606   \n",
       "560952  ZDB-GENE-030131-8404       trps1                7955   \n",
       "560953        WBGene00017033      nekl-4                6239   \n",
       "560954                 18592       NEK10                9606   \n",
       "560955   ZDB-GENE-130530-570       nek10                7955   \n",
       "\n",
       "               Gene1SpeciesName     Gene2ID Gene2Symbol Gene2SpeciesTaxonID  \\\n",
       "0                  Homo sapiens  S000004937        ADH6              559292   \n",
       "1       Drosophila melanogaster  S000004937        ADH6              559292   \n",
       "2                  Mus musculus  S000004937        ADH6              559292   \n",
       "3        Caenorhabditis elegans  S000004937        ADH6              559292   \n",
       "4             Rattus norvegicus  S000004937        ADH6              559292   \n",
       "...                         ...         ...         ...                 ...   \n",
       "560951             Homo sapiens     1927616       Trps1               10090   \n",
       "560952              Danio rerio     1927616       Trps1               10090   \n",
       "560953   Caenorhabditis elegans     2685128       Nek10               10090   \n",
       "560954             Homo sapiens     2685128       Nek10               10090   \n",
       "560955              Danio rerio     2685128       Nek10               10090   \n",
       "\n",
       "                Gene2SpeciesName  \\\n",
       "0       Saccharomyces cerevisiae   \n",
       "1       Saccharomyces cerevisiae   \n",
       "2       Saccharomyces cerevisiae   \n",
       "3       Saccharomyces cerevisiae   \n",
       "4       Saccharomyces cerevisiae   \n",
       "...                          ...   \n",
       "560951              Mus musculus   \n",
       "560952              Mus musculus   \n",
       "560953              Mus musculus   \n",
       "560954              Mus musculus   \n",
       "560955              Mus musculus   \n",
       "\n",
       "                                               Algorithms  AlgorithmsMatch  \\\n",
       "0                                     OMA|Ensembl Compara                2   \n",
       "1                                 PANTHER|Ensembl Compara                2   \n",
       "2                                     OMA|Ensembl Compara                2   \n",
       "3                                 PANTHER|Ensembl Compara                2   \n",
       "4                                     OMA|Ensembl Compara                2   \n",
       "...                                                   ...              ...   \n",
       "560951  PANTHER|HGNC|PhylomeDB|OrthoFinder|InParanoid|...                9   \n",
       "560952  ZFIN|PANTHER|PhylomeDB|OrthoFinder|InParanoid|...                7   \n",
       "560953  PANTHER|PhylomeDB|OrthoFinder|InParanoid|Ortho...                8   \n",
       "560954  PANTHER|HGNC|Ensembl Compara|OMA|PhylomeDB|Ort...               11   \n",
       "560955  ZFIN|PANTHER|PhylomeDB|OrthoFinder|InParanoid|...                8   \n",
       "\n",
       "        OutOfAlgorithms IsBestScore IsBestRevScore  AlgorithmsMatchPerc  \n",
       "0                    10         Yes            Yes             0.200000  \n",
       "1                    10         Yes            Yes             0.200000  \n",
       "2                    10         Yes            Yes             0.200000  \n",
       "3                    10         Yes            Yes             0.200000  \n",
       "4                     9         Yes            Yes             0.222222  \n",
       "...                 ...         ...            ...                  ...  \n",
       "560951               11         Yes            Yes             0.818182  \n",
       "560952               11         Yes            Yes             0.636364  \n",
       "560953               10         Yes            Yes             0.800000  \n",
       "560954               11         Yes            Yes             1.000000  \n",
       "560955               11         Yes            Yes             0.727273  \n",
       "\n",
       "[560956 rows x 14 columns]"
      ]
     },
     "execution_count": 96,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# add column: Algorithm Match Percentage - the % of algorithms that agree on the ortholog mapping \n",
    "df[\"AlgorithmsMatchPerc\"] = df[\"AlgorithmsMatch\"]/df[\"OutOfAlgorithms\"] \n",
    "df"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Data Filtering\n",
    "Keep ortholog mapping information for relevant species for our purposes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 97,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>HumanID</th>\n",
       "      <th>HumanSymbol</th>\n",
       "      <th>YeastID</th>\n",
       "      <th>YeastSymbol</th>\n",
       "      <th>Algorithms</th>\n",
       "      <th>AlgorithmsMatch</th>\n",
       "      <th>OutOfAlgorithms</th>\n",
       "      <th>AlgorithmsMatchPerc</th>\n",
       "      <th>IsBestScore</th>\n",
       "      <th>IsBestRevScore</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>28697</td>\n",
       "      <td>ZADH2</td>\n",
       "      <td>S000004937</td>\n",
       "      <td>ADH6</td>\n",
       "      <td>OMA|Ensembl Compara</td>\n",
       "      <td>2</td>\n",
       "      <td>10</td>\n",
       "      <td>0.2</td>\n",
       "      <td>Yes</td>\n",
       "      <td>Yes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>33848</td>\n",
       "      <td>COA5</td>\n",
       "      <td>S000003795</td>\n",
       "      <td>PET191</td>\n",
       "      <td>PANTHER|PhylomeDB|OrthoFinder|OrthoInspector|H...</td>\n",
       "      <td>7</td>\n",
       "      <td>10</td>\n",
       "      <td>0.7</td>\n",
       "      <td>Yes</td>\n",
       "      <td>Yes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>20154</td>\n",
       "      <td>ABHD4</td>\n",
       "      <td>S000003342</td>\n",
       "      <td>CLD1</td>\n",
       "      <td>PANTHER|Ensembl Compara|PhylomeDB|Hieranoid|Tr...</td>\n",
       "      <td>6</td>\n",
       "      <td>10</td>\n",
       "      <td>0.6</td>\n",
       "      <td>No</td>\n",
       "      <td>Yes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>21396</td>\n",
       "      <td>ABHD5</td>\n",
       "      <td>S000003342</td>\n",
       "      <td>CLD1</td>\n",
       "      <td>PANTHER|Ensembl Compara|PhylomeDB|Hieranoid|Tr...</td>\n",
       "      <td>6</td>\n",
       "      <td>10</td>\n",
       "      <td>0.6</td>\n",
       "      <td>No</td>\n",
       "      <td>Yes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>10524</td>\n",
       "      <td>SALL1</td>\n",
       "      <td>S000006217</td>\n",
       "      <td>CMR3</td>\n",
       "      <td>PANTHER|Roundup</td>\n",
       "      <td>2</td>\n",
       "      <td>10</td>\n",
       "      <td>0.2</td>\n",
       "      <td>Yes</td>\n",
       "      <td>Yes</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   HumanID HumanSymbol     YeastID YeastSymbol  \\\n",
       "0    28697       ZADH2  S000004937        ADH6   \n",
       "5    33848        COA5  S000003795      PET191   \n",
       "11   20154       ABHD4  S000003342        CLD1   \n",
       "12   21396       ABHD5  S000003342        CLD1   \n",
       "19   10524       SALL1  S000006217        CMR3   \n",
       "\n",
       "                                           Algorithms  AlgorithmsMatch  \\\n",
       "0                                 OMA|Ensembl Compara                2   \n",
       "5   PANTHER|PhylomeDB|OrthoFinder|OrthoInspector|H...                7   \n",
       "11  PANTHER|Ensembl Compara|PhylomeDB|Hieranoid|Tr...                6   \n",
       "12  PANTHER|Ensembl Compara|PhylomeDB|Hieranoid|Tr...                6   \n",
       "19                                    PANTHER|Roundup                2   \n",
       "\n",
       "    OutOfAlgorithms  AlgorithmsMatchPerc IsBestScore IsBestRevScore  \n",
       "0                10                  0.2         Yes            Yes  \n",
       "5                10                  0.7         Yes            Yes  \n",
       "11               10                  0.6          No            Yes  \n",
       "12               10                  0.6          No            Yes  \n",
       "19               10                  0.2         Yes            Yes  "
      ]
     },
     "execution_count": 97,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# create human to yeast mapping table\n",
    "human2yeast = df.loc[(df['Gene1SpeciesName'] == 'Homo sapiens') & (df['Gene2SpeciesName'] == 'Saccharomyces cerevisiae')]\n",
    "human2yeast = human2yeast.rename(columns={'Gene1ID': 'HumanID', 'Gene1Symbol': 'HumanSymbol',\n",
    "                                'Gene2ID': 'YeastID', 'Gene2Symbol': 'YeastSymbol',})\n",
    "\n",
    "human2yeast = human2yeast.filter(items=['HumanID', 'HumanSymbol', 'YeastID', 'YeastSymbol',\n",
    "                'Algorithms', 'AlgorithmsMatch', 'OutOfAlgorithms', 'AlgorithmsMatchPerc',\n",
    "                          'IsBestScore', 'IsBestRevScore',])\n",
    "human2yeast.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 98,
   "metadata": {},
   "outputs": [],
   "source": [
    "human2yeast.infer_objects()\n",
    "human2yeast['HumanID']=human2yeast['HumanID'].astype(str).astype(int)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 99,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "HumanID                  int64\n",
       "HumanSymbol             object\n",
       "YeastID                 object\n",
       "YeastSymbol             object\n",
       "Algorithms              object\n",
       "AlgorithmsMatch          int64\n",
       "OutOfAlgorithms          int64\n",
       "AlgorithmsMatchPerc    float64\n",
       "IsBestScore             object\n",
       "IsBestRevScore          object\n",
       "dtype: object"
      ]
     },
     "execution_count": 99,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "human2yeast.dtypes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 100,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>YeastID</th>\n",
       "      <th>YeastSymbol</th>\n",
       "      <th>HumanID</th>\n",
       "      <th>HumanSymbol</th>\n",
       "      <th>Algorithms</th>\n",
       "      <th>AlgorithmsMatch</th>\n",
       "      <th>OutOfAlgorithms</th>\n",
       "      <th>AlgorithmsMatchPerc</th>\n",
       "      <th>IsBestScore</th>\n",
       "      <th>IsBestRevScore</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>286719</th>\n",
       "      <td>S000002326</td>\n",
       "      <td>TEX13A</td>\n",
       "      <td>11735</td>\n",
       "      <td>NRP1</td>\n",
       "      <td>OrthoFinder|Ensembl Compara</td>\n",
       "      <td>2</td>\n",
       "      <td>10</td>\n",
       "      <td>0.2</td>\n",
       "      <td>Yes</td>\n",
       "      <td>Yes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>286733</th>\n",
       "      <td>S000000321</td>\n",
       "      <td>TKT</td>\n",
       "      <td>11834</td>\n",
       "      <td>TKL2</td>\n",
       "      <td>Ensembl Compara|OMA|PhylomeDB|OrthoFinder|InPa...</td>\n",
       "      <td>9</td>\n",
       "      <td>10</td>\n",
       "      <td>0.9</td>\n",
       "      <td>Yes</td>\n",
       "      <td>Yes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>286734</th>\n",
       "      <td>S000006278</td>\n",
       "      <td>TKT</td>\n",
       "      <td>11834</td>\n",
       "      <td>TKL1</td>\n",
       "      <td>Ensembl Compara|OMA|PhylomeDB|OrthoFinder|InPa...</td>\n",
       "      <td>9</td>\n",
       "      <td>10</td>\n",
       "      <td>0.9</td>\n",
       "      <td>Yes</td>\n",
       "      <td>Yes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>286750</th>\n",
       "      <td>S000001295</td>\n",
       "      <td>PRKAR1B</td>\n",
       "      <td>9390</td>\n",
       "      <td>BCY1</td>\n",
       "      <td>PANTHER|OMA|OrthoFinder|OrthoInspector|TreeFam...</td>\n",
       "      <td>6</td>\n",
       "      <td>10</td>\n",
       "      <td>0.6</td>\n",
       "      <td>No</td>\n",
       "      <td>Yes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>286756</th>\n",
       "      <td>S000005041</td>\n",
       "      <td>ING2</td>\n",
       "      <td>6063</td>\n",
       "      <td>PHO23</td>\n",
       "      <td>PANTHER|Ensembl Compara|OrthoInspector|Hierano...</td>\n",
       "      <td>6</td>\n",
       "      <td>10</td>\n",
       "      <td>0.6</td>\n",
       "      <td>Yes</td>\n",
       "      <td>Yes</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           YeastID YeastSymbol HumanID HumanSymbol  \\\n",
       "286719  S000002326      TEX13A   11735        NRP1   \n",
       "286733  S000000321         TKT   11834        TKL2   \n",
       "286734  S000006278         TKT   11834        TKL1   \n",
       "286750  S000001295     PRKAR1B    9390        BCY1   \n",
       "286756  S000005041        ING2    6063       PHO23   \n",
       "\n",
       "                                               Algorithms  AlgorithmsMatch  \\\n",
       "286719                        OrthoFinder|Ensembl Compara                2   \n",
       "286733  Ensembl Compara|OMA|PhylomeDB|OrthoFinder|InPa...                9   \n",
       "286734  Ensembl Compara|OMA|PhylomeDB|OrthoFinder|InPa...                9   \n",
       "286750  PANTHER|OMA|OrthoFinder|OrthoInspector|TreeFam...                6   \n",
       "286756  PANTHER|Ensembl Compara|OrthoInspector|Hierano...                6   \n",
       "\n",
       "        OutOfAlgorithms  AlgorithmsMatchPerc IsBestScore IsBestRevScore  \n",
       "286719               10                  0.2         Yes            Yes  \n",
       "286733               10                  0.9         Yes            Yes  \n",
       "286734               10                  0.9         Yes            Yes  \n",
       "286750               10                  0.6          No            Yes  \n",
       "286756               10                  0.6         Yes            Yes  "
      ]
     },
     "execution_count": 100,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# create yeast to human mapping table\n",
    "yeast2human = df.loc[(df['Gene1SpeciesName'] == 'Saccharomyces cerevisiae') & (df['Gene2SpeciesName'] == 'Homo sapiens')]\n",
    "yeast2human = yeast2human.rename(columns={'Gene1ID': 'YeastID', 'Gene1Symbol': 'HumanSymbol',\n",
    "                                'Gene2ID': 'HumanID', 'Gene2Symbol': 'YeastSymbol',})\n",
    "yeast2human = yeast2human.filter(items= ['YeastID', 'YeastSymbol','HumanID', 'HumanSymbol',\n",
    "                         'Algorithms', 'AlgorithmsMatch', 'OutOfAlgorithms', 'AlgorithmsMatchPerc',\n",
    "                            'IsBestScore', 'IsBestRevScore'])\n",
    "yeast2human.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 101,
   "metadata": {},
   "outputs": [],
   "source": [
    "yeast2human['HumanID']=yeast2human['HumanID'].astype(str).astype(int)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 102,
   "metadata": {},
   "outputs": [],
   "source": [
    "# write ortholog mapping tables to file\n",
    "yeast2human.to_csv(path_or_buf='data/sgd_alliance/yeast2human_alliance_v4.csv', index=False)\n",
    "human2yeast.to_csv(path_or_buf='data/sgd_alliance/human2yeast_alliance_v4.csv', index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Create BigQuery Dataset and Upload Data\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 103,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Your browser has been opened to visit:\n",
      "\n",
      "    https://accounts.google.com/o/oauth2/auth?code_challenge=uNVhind8gJVqxcyEgUv8OKO9u4ooK2XlqrKVqbO2hfw&prompt=select_account&code_challenge_method=S256&access_type=offline&redirect_uri=http%3A%2F%2Flocalhost%3A8085%2F&response_type=code&client_id=764086051850-6qr4p6gpi6hn506pt8ejuq83di341hur.apps.googleusercontent.com&scope=openid+https%3A%2F%2Fwww.googleapis.com%2Fauth%2Fuserinfo.email+https%3A%2F%2Fwww.googleapis.com%2Fauth%2Fcloud-platform+https%3A%2F%2Fwww.googleapis.com%2Fauth%2Faccounts.reauth\n",
      "\n",
      "\n",
      "\n",
      "Credentials saved to file: [/Users/nchamb/.config/gcloud/application_default_credentials.json]\n",
      "\n",
      "These credentials will be used by any library that requests Application Default Credentials (ADC).\n",
      "\n",
      "Quota project \"syntheticlethality\" was added to ADC which can be used by Google client libraries for billing and quota. Note that some services may still bill the project owning the resource.\n"
     ]
    }
   ],
   "source": [
    "# google cloud authentication\n",
    "!gcloud auth application-default login"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 104,
   "metadata": {},
   "outputs": [],
   "source": [
    "# pip install pyarrow\n",
    "from google.cloud import bigquery"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 105,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/opt/anaconda3/lib/python3.7/site-packages/google/auth/_default.py:69: UserWarning: Your application has authenticated using end user credentials from Google Cloud SDK. We recommend that most server applications use service accounts instead. If your application continues to use end user credentials from Cloud SDK, you might receive a \"quota exceeded\" or \"API not enabled\" error. For more information about service accounts, see https://cloud.google.com/docs/authentication/\n",
      "  warnings.warn(_CLOUD_SDK_CREDENTIALS_WARNING)\n"
     ]
    }
   ],
   "source": [
    "# configure project info and bigquery client\n",
    "project='syntheticlethality'\n",
    "\n",
    "# construct a BigQuery client object.\n",
    "client = bigquery.Client(project) # Replace XXXXXXXX with your project ID"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Create a bigquery dataset within your project \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 106,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Dataset: syntheticlethality already exists\n"
     ]
    }
   ],
   "source": [
    "# if it does not already exist\n",
    "dataset_name='Orthology'\n",
    "dataset_id = bigquery.Dataset(project +'.'+ dataset_name)\n",
    "try:\n",
    "    client.get_dataset(dataset_id)\n",
    "    print(\"Dataset: {} already exists\".format(project, dataset_name))\n",
    "except:\n",
    "    # construct a full Dataset object to send to the API.\n",
    "    dataset = bigquery.Dataset(dataset_id)\n",
    "    \n",
    "    # send the dataset to the API for creation.\n",
    "    dataset = client.create_dataset(dataset_id)  # Make an API request.\n",
    "    print(\"Created dataset: {}.{}\".format(project_id,  dataset_id))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Configure upload job - Yeast2Human Table\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 107,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['YeastID', 'YeastSymbol', 'HumanID', 'HumanSymbol', 'Algorithms',\n",
       "       'AlgorithmsMatch', 'OutOfAlgorithms', 'AlgorithmsMatchPerc',\n",
       "       'IsBestScore', 'IsBestRevScore'], dtype=object)"
      ]
     },
     "execution_count": 107,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "yeast2human.columns.values"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 108,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "YeastID                 object\n",
      "YeastSymbol             object\n",
      "HumanID                  int64\n",
      "HumanSymbol             object\n",
      "Algorithms              object\n",
      "AlgorithmsMatch          int64\n",
      "OutOfAlgorithms          int64\n",
      "AlgorithmsMatchPerc    float64\n",
      "IsBestScore             object\n",
      "IsBestRevScore          object\n",
      "dtype: object\n"
     ]
    }
   ],
   "source": [
    "print(yeast2human.dtypes) "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 109,
   "metadata": {},
   "outputs": [],
   "source": [
    "table_description = '''\n",
    "Mapping conserved genes from yeast (Saccharomyces cerevisiae) to human (Homo sapiens) derived from Integrated orthology inferences created using the Drosophila RNAi Screening Center(DRSC) Integrative Ortholog Prediction Tool (DIOPT) \n",
    "at Harvard Medical School see more information at: https://www.flyrnai.org/diopt. \n",
    "\n",
    "Table downloaded from the Alliance for Genome Resources. Release 3.0.1\"\n",
    "https://www.alliancegenome.org/downloads#orthology\n",
    "'''"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 110,
   "metadata": {},
   "outputs": [],
   "source": [
    "job_config = bigquery.LoadJobConfig(\n",
    "   destination_table_description=table_description,\n",
    "    \n",
    "    # specify table schema\n",
    "    schema=[\n",
    "        bigquery.SchemaField(name=\"YeastID\", field_type=bigquery.enums.SqlTypeNames.STRING, \n",
    "                             description='Saccharomyces Genome Database (SGD) gene identifier'),\n",
    "        bigquery.SchemaField(name=\"YeastSymbol\", field_type=bigquery.enums.SqlTypeNames.STRING, \n",
    "                             description=\"Yeast official gene symbol\"),\n",
    "        bigquery.SchemaField(name=\"HumanID\", field_type=bigquery.enums.SqlTypeNames.INTEGER, \n",
    "                             description='HGNC gene identifier'),\n",
    "        bigquery.SchemaField(name=\"HumanSymbol\", field_type=bigquery.enums.SqlTypeNames.STRING, \n",
    "                             description='HGNC gene symbol'),\n",
    "       bigquery.SchemaField(name=\"Algorithms\", field_type=bigquery.enums.SqlTypeNames.STRING, \n",
    "                          description='The orthology methods that make the orthology association for the species'),\n",
    "        bigquery.SchemaField(name=\"AlgorithmsMatch\", field_type=bigquery.enums.SqlTypeNames.INTEGER, \n",
    "                             description='The number of orthology methods that make the orthology association for the species'),\n",
    "        bigquery.SchemaField(name=\"OutOfAlgorithms\", field_type=bigquery.enums.SqlTypeNames.INTEGER, \n",
    "                             description='The toal number of orthology methods that could make the orthology association for the species'),\n",
    "        bigquery.SchemaField(name=\"AlgorithmsMatchPerc\", field_type=bigquery.enums.SqlTypeNames.FLOAT, \n",
    "                             description='The proportion of orthology methods that make the orthology association for the species'),\n",
    "        bigquery.SchemaField(name=\"IsBestScore\", field_type=bigquery.enums.SqlTypeNames.STRING, \n",
    "                             description='Within the species, whether this gene is called the ortholog of the input gene by the highest number of algorithms'),\n",
    "        bigquery.SchemaField(name=\"IsBestRevScore\", field_type=bigquery.enums.SqlTypeNames.STRING, \n",
    "                             description='Within the species of the input gene, whether the input gene is called the ortholog of the gene by the highest number of algorithms'), \n",
    "    ],\n",
    "    \n",
    "    write_disposition=\"WRITE_TRUNCATE\" #replaces the table with the loaded data\n",
    ")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 111,
   "metadata": {},
   "outputs": [],
   "source": [
    "# create table name\n",
    "table_name = 'YEAST2HUMAN'\n",
    "table_id = \"{}.{}.{}\".format(project,  dataset_name, table_name)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Load tables to bigquery"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 112,
   "metadata": {},
   "outputs": [],
   "source": [
    "job = client.load_table_from_dataframe(yeast2human, table_id, job_config=job_config)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 113,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<google.cloud.bigquery.job.LoadJob at 0x1516af490>"
      ]
     },
     "execution_count": 113,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "job.result()  # Wait for the job to complete.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 114,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Loaded 6124 rows and 10 columns to syntheticlethality.Orthology.YEAST2HUMAN\n"
     ]
    }
   ],
   "source": [
    "table = client.get_table(table_id)  # Make an API request.\n",
    "print(\n",
    "    \"Loaded {} rows and {} columns to {}\".format(\n",
    "        table.num_rows, len(table.schema), table_id\n",
    "    )\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Configure upload job - Human2Yeast Table"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 115,
   "metadata": {},
   "outputs": [],
   "source": [
    "table_description = '''\n",
    "Mapping conserved genes from human (Homo sapiens) to yeast (Saccharomyces cerevisiae) to  derived from Integrated orthology inferences created using the Drosophila RNAi Screening Center(DRSC) Integrative Ortholog Prediction Tool (DIOPT) \n",
    "at Harvard Medical School see more information at: https://www.flyrnai.org/diopt. \n",
    "\n",
    "Table downloaded from the Alliance for Genome Resources. Release 3.0.1\"\n",
    "https://www.alliancegenome.org/downloads#orthology\n",
    "'''"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 116,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['HumanID', 'HumanSymbol', 'YeastID', 'YeastSymbol', 'Algorithms',\n",
       "       'AlgorithmsMatch', 'OutOfAlgorithms', 'AlgorithmsMatchPerc',\n",
       "       'IsBestScore', 'IsBestRevScore'], dtype=object)"
      ]
     },
     "execution_count": 116,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "human2yeast.columns.values"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 117,
   "metadata": {},
   "outputs": [],
   "source": [
    "# configure upload\n",
    "job_config = bigquery.LoadJobConfig(\n",
    "   destination_table_description=table_description,\n",
    "    \n",
    "    # specify table schema\n",
    "    schema=[\n",
    "        bigquery.SchemaField(name=\"HumanID\", field_type=bigquery.enums.SqlTypeNames.INTEGER, \n",
    "                             description='HGNC gene identifier'),\n",
    "        bigquery.SchemaField(name=\"HumanSymbol\", field_type=bigquery.enums.SqlTypeNames.STRING, \n",
    "                             description='HGNC gene symbol'),\n",
    "               bigquery.SchemaField(name=\"YeastID\", field_type=bigquery.enums.SqlTypeNames.STRING, \n",
    "                             description='Saccharomyces Genome Database (SGD) gene identifier'),\n",
    "        bigquery.SchemaField(name=\"YeastSymbol\", field_type=bigquery.enums.SqlTypeNames.STRING, \n",
    "                             description=\"Yeast official gene symbol\"),\n",
    "       bigquery.SchemaField(name=\"Algorithms\", field_type=bigquery.enums.SqlTypeNames.STRING, \n",
    "                          description='The orthology methods that make the orthology association for the species'),\n",
    "        bigquery.SchemaField(name=\"AlgorithmsMatch\", field_type=bigquery.enums.SqlTypeNames.INTEGER, \n",
    "                             description='The number of orthology methods that make the orthology association for the species'),\n",
    "        bigquery.SchemaField(name=\"OutOfAlgorithms\", field_type=bigquery.enums.SqlTypeNames.INTEGER, \n",
    "                             description='The toal number of orthology methods that could make the orthology association for the species'),\n",
    "        bigquery.SchemaField(name=\"AlgorithmsMatchPerc\", field_type=bigquery.enums.SqlTypeNames.FLOAT, \n",
    "                             description='The proportion of orthology methods that make the orthology association for the species'),\n",
    "        bigquery.SchemaField(name=\"IsBestScore\", field_type=bigquery.enums.SqlTypeNames.STRING, \n",
    "                             description='Within the species, whether this gene is called the ortholog of the input gene by the highest number of algorithms'),\n",
    "        bigquery.SchemaField(name=\"IsBestRevScore\", field_type=bigquery.enums.SqlTypeNames.STRING, \n",
    "                             description='Within the species of the input gene, whether the input gene is called the ortholog of the gene by the highest number of algorithms'), \n",
    "    ],\n",
    "    \n",
    "    write_disposition=\"WRITE_TRUNCATE\" #replaces the table with the loaded data\n",
    ")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 118,
   "metadata": {},
   "outputs": [],
   "source": [
    "# create table name\n",
    "table_name = 'HUMAN2YEAST'\n",
    "table_id = \"{}.{}.{}\".format(project,  dataset_name, table_name)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 119,
   "metadata": {},
   "outputs": [],
   "source": [
    "job = client.load_table_from_dataframe(human2yeast, table_id, job_config=job_config)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 120,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<google.cloud.bigquery.job.LoadJob at 0x1516d0690>"
      ]
     },
     "execution_count": 120,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "job.result()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 121,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Loaded 6124 rows and 10 columns to syntheticlethality.Orthology.HUMAN2YEAST\n"
     ]
    }
   ],
   "source": [
    "table = client.get_table(table_id)  # Make an API request.\n",
    "print(\n",
    "    \"Loaded {} rows and {} columns to {}\".format(\n",
    "        table.num_rows, len(table.schema), table_id\n",
    "    )\n",
    ")"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.8"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
