{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#  DAISY- the DAta-mIning SYnthetic-lethality-identification pipeline\n",
    "\n",
    "Please cite: \n",
    "\n",
    "For Implementation: \n",
    "\n",
    "Our paper,\n",
    "\n",
    "For DAISY algorithm: \n",
    "\n",
    "Jerby-Arnon, L., Pfetzer, N., Waldman, Y. Y., McGarry, L., James, D., Shanks, E., ... & Gottlieb, E. (2014). Predicting cancer-specific vulnerability via data-driven detection of synthetic lethality. Cell, 158(5), 1199-1209.\n",
    "\n",
    "For CCLE Omics data:\n",
    "\n",
    "Ghandi, M., Huang, F.W., Jané-Valbuena, J. et al. Next-generation characterization of the Cancer Cell Line Encyclopedia. Nature 569, 503–508 (2019). https://doi.org/10.1038/s41586-019-1186-3\n",
    "\n",
    "For CRISPR Data: \n",
    "\n",
    "Robin M. Meyers, Jordan G. Bryan, James M. McFarland, Barbara A. Weir, ... David E. Root, William C. Hahn, Aviad Tsherniak. Computational correction of copy number effect improves specificity of CRISPR-Cas9 essentiality screens in cancer cells. Nature Genetics 2017 October 49:1779–1784. doi:10.1038/ng.3984\n",
    "\n",
    "Dempster, J. M., Rossen, J., Kazachkova, M., Pan, J., Kugener, G., Root, D. E., & Tsherniak, A. (2019). Extracting Biological Insights from the Project Achilles Genome-Scale CRISPR Screens in Cancer Cell Lines. BioRxiv, 720243.\n",
    "\n",
    "\n",
    "This notebook is a reimplementation of DAISY Synthetic Lethal Pair Prediction Algorithm\n",
    "\n",
    "Please first run the table_creation notebook before runnnig the DAISY notebook. \n",
    "\n",
    "It consists 3 modules: \n",
    "\n",
    "1. SL candidate determination using gene co-expression\n",
    "2. SL candidate determination using survival of fittest\n",
    "3. SL candidate determination using CRISPR and ShRNA experiment\n",
    "\n",
    "\n",
    "* The results from the three modules were then aggregated into one ranked list of candidate SL pairs\n",
    "\n",
    "\n",
    "Input Parameters\n",
    "* Cancer type \n",
    "* The genes whose SL partners are seeked\n",
    "\n",
    "\n",
    "Input Data\n",
    "* Gene expression data \n",
    "* Gene mutation data\n",
    "* Copy number variation data\n",
    "* Gene effect data (CRISPR)\n",
    "* Gene Dependency scores data (shRNA)\n",
    "\n",
    "Output\n",
    "* Ranked list of candidate SL pairs\n",
    "![../../figures/daisy_pipeline.png](attachment:dene.png)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Once deleted, variables cannot be recovered. Proceed (y/[n])? y\n"
     ]
    }
   ],
   "source": [
    "reset "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'/Users/bahar/Desktop/DAISY_imp/SL-Cloud-main/DAISY_pipeline'"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pwd"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 1. Import python libraries required\n",
    "The required libraries are imported. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "from datetime import datetime\n",
    "import sys\n",
    "sys.path.append('../scripts/') #need to add \"scripts\" directory in a parent directory \n",
    "from google.cloud import bigquery\n",
    "import importlib\n",
    "import pandas as pd\n",
    "import DAISY_operations\n",
    "importlib.reload(DAISY_operations)\n",
    "from DAISY_operations import *\n",
    "from helper_functions import *\n",
    "from BIGQUERY_operations import *"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "if not sys.warnoptions:\n",
    "    import warnings\n",
    "    warnings.simplefilter(\"ignore\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 2. Sign in Google Bigquery with the project id\n",
    "\n",
    "Bigquery connection\n",
    "Please replace syntheticlethality with your project name"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Traceback (most recent call last):\r\n",
      "  File \"/Users/bahar/Downloads/google-cloud-sdk/lib/gcloud.py\", line 104, in <module>\r\n",
      "    main()\r\n",
      "  File \"/Users/bahar/Downloads/google-cloud-sdk/lib/gcloud.py\", line 100, in main\r\n",
      "    sys.exit(gcloud_main.main())\r\n",
      "  File \"/Users/bahar/Downloads/google-cloud-sdk/lib/googlecloudsdk/gcloud_main.py\", line 171, in main\r\n",
      "    gcloud_cli = CreateCLI([])\r\n",
      "  File \"/Users/bahar/Downloads/google-cloud-sdk/lib/googlecloudsdk/gcloud_main.py\", line 151, in CreateCLI\r\n",
      "    generated_cli = loader.Generate()\r\n",
      "  File \"/Users/bahar/Downloads/google-cloud-sdk/lib/googlecloudsdk/calliope/cli.py\", line 504, in Generate\r\n",
      "    cli = self.__MakeCLI(top_group)\r\n",
      "  File \"/Users/bahar/Downloads/google-cloud-sdk/lib/googlecloudsdk/calliope/cli.py\", line 674, in __MakeCLI\r\n",
      "    log.AddFileLogging(self.__logs_dir)\r\n",
      "  File \"/Users/bahar/Downloads/google-cloud-sdk/lib/googlecloudsdk/core/log.py\", line 1039, in AddFileLogging\r\n",
      "    _log_manager.AddLogsDir(logs_dir=logs_dir)\r\n",
      "  File \"/Users/bahar/Downloads/google-cloud-sdk/lib/googlecloudsdk/core/log.py\", line 683, in AddLogsDir\r\n",
      "    self._CleanUpLogs(logs_dir)\r\n",
      "  File \"/Users/bahar/Downloads/google-cloud-sdk/lib/googlecloudsdk/core/log.py\", line 711, in _CleanUpLogs\r\n",
      "    self._CleanLogsDir(logs_dir)\r\n",
      "  File \"/Users/bahar/Downloads/google-cloud-sdk/lib/googlecloudsdk/core/log.py\", line 737, in _CleanLogsDir\r\n",
      "    os.remove(log_file_path)\r\n",
      "PermissionError: [Errno 13] Permission denied: '/Users/bahar/.config/gcloud/logs/2020.08.04/12.16.05.276251.log'\r\n"
     ]
    }
   ],
   "source": [
    "project_id='syntheticlethality'\n",
    "client = bigquery.Client(project_id)\n",
    "#client = bigquery.Client(credentials=credentials, project=credentials.project_id)\n",
    "\n",
    "!gcloud auth login"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The google.cloud.bigquery extension is already loaded. To reload it, use:\n",
      "  %reload_ext google.cloud.bigquery\n"
     ]
    }
   ],
   "source": [
    "%load_ext google.cloud.bigquery"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "UsageError: %%bigquery is a cell magic, but the cell body is empty.\n"
     ]
    }
   ],
   "source": [
    "%%bigquery tsg"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 3. Input genes of interest for the SL partner prediction\n",
    "\n",
    "We will predict synthetic lethal partner genes for tumor suppressor genes as default.\n",
    "The query will use a permission required big-query table for tumor suppressor genes.\n",
    "To execute this query, you need to register for a new COSMIC account and to get a permission.\n",
    "Please follow this link for the registration : https://isb-cancer-genomics-cloud.readthedocs.io/en/latest/sections/data/COSMIC_about.html.\n",
    "If you want to test the genes of your interest, please skip this and add you own genes to the variable \"input_genes\"."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "query='''\n",
    "SELECT Gene_Symbol\n",
    "  FROM `isb-cgc.COSMIC_v90_grch38.Cancer_Gene_Census` \n",
    " WHERE Role_in_Cancer LIKE '%TSG%'\n",
    "\n",
    "INTERSECT DISTINCT\n",
    "\n",
    "SELECT HGNC_gene_symbol\n",
    "  FROM `syntheticlethality.gene_information.cancer_driver_genes`\n",
    " \n",
    "'''\n",
    "driver_tsg_genes = client.query(query).result().to_dataframe()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<br>\n",
    "Conversion from Hugo Symbols into EntrezIDs "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Gene</th>\n",
       "      <th>EntrezID</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>APC</td>\n",
       "      <td>324</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ATM</td>\n",
       "      <td>472</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ATR</td>\n",
       "      <td>545</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>B2M</td>\n",
       "      <td>567</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>CIC</td>\n",
       "      <td>23152</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>108</th>\n",
       "      <td>PRKAR1A</td>\n",
       "      <td>5573</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>109</th>\n",
       "      <td>SMARCA4</td>\n",
       "      <td>6597</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>110</th>\n",
       "      <td>SMARCB1</td>\n",
       "      <td>6598</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>111</th>\n",
       "      <td>TBL1XR1</td>\n",
       "      <td>79718</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>112</th>\n",
       "      <td>TNFAIP3</td>\n",
       "      <td>7128</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>113 rows × 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        Gene  EntrezID\n",
       "0        APC       324\n",
       "1        ATM       472\n",
       "2        ATR       545\n",
       "3        B2M       567\n",
       "4        CIC     23152\n",
       "..       ...       ...\n",
       "108  PRKAR1A      5573\n",
       "109  SMARCA4      6597\n",
       "110  SMARCB1      6598\n",
       "111  TBL1XR1     79718\n",
       "112  TNFAIP3      7128\n",
       "\n",
       "[113 rows x 2 columns]"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "input_genes = driver_tsg_genes[\"Gene_Symbol\"].to_list()\n",
    "input_entrez_ids = ConvertGene(client, input_genes, 'Gene', ['EntrezID'])\n",
    "input_entrez_ids"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 4. Prediction of synthetic lethal partners using different modules on DAISY\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "There are three modules for synthetic lethal pair inferences on DAISY : 1. Pairwise gene coexpression, 2. Genomic survival of the fittest. 3. shRNA or CRISPR based functional examination. You can get more information in the original paper : https://www.sciencedirect.com/science/article/pii/S0092867414009775.\n",
    "\n",
    "In pairwise gene coexpression module and genomic survial of the fittest module, we will use PancancerAtlas and CCLE data.<br>\n",
    "In functional examination module, we will use CRISPR and shRNA data. <br>\n",
    "\n",
    "Python codes for each module are built in our internal library (../scripts/SL_library.py) which was already imported at the beginning. \n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### 4.0. Default parameters for DAISY, you can edit them"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [],
   "source": [
    "input_mutations = ['Nonsense_Mutation', 'Frame_Shift_Ins', 'Frame_Shift_Del'] # Three mutation types were chosed as default by DAISY.\n",
    "percentile_threshold = 10\n",
    "cn_threshold = -0.3 \n",
    "cor_threshold = 0.5\n",
    "p_threshold = 0.05\n",
    "pval_correction = 'Bonferroni'"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### 4.1. Pairwise gene coexpression module"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "4.1.1. Pairwise gene coexpression module on PancancerAtlas."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "coexp_pancancer = CoexpressionAnalysis(client, \"PanCancerAtlas\", [472], cor_threshold, p_threshold, pval_correction)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th>EntrezID_Inactive</th>\n",
       "      <th>Gene_Inactive</th>\n",
       "      <th>EntrezID_SL_Candidate</th>\n",
       "      <th>Gene_SL_Candidate</th>\n",
       "      <th>Correlation</th>\n",
       "      <th>PValue</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Gene_Inactive</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th rowspan=\"11\" valign=\"top\">ATM</th>\n",
       "      <th>0</th>\n",
       "      <td>472</td>\n",
       "      <td>ATM</td>\n",
       "      <td>4297</td>\n",
       "      <td>KMT2A</td>\n",
       "      <td>0.719314</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>130</th>\n",
       "      <td>472</td>\n",
       "      <td>ATM</td>\n",
       "      <td>1106</td>\n",
       "      <td>CHD2</td>\n",
       "      <td>0.525927</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>131</th>\n",
       "      <td>472</td>\n",
       "      <td>ATM</td>\n",
       "      <td>7621</td>\n",
       "      <td>ZNF70</td>\n",
       "      <td>0.525772</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>132</th>\n",
       "      <td>472</td>\n",
       "      <td>ATM</td>\n",
       "      <td>100101467</td>\n",
       "      <td>ZSCAN30</td>\n",
       "      <td>0.525512</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>133</th>\n",
       "      <td>472</td>\n",
       "      <td>ATM</td>\n",
       "      <td>23506</td>\n",
       "      <td>BICRAL</td>\n",
       "      <td>0.525503</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>71</th>\n",
       "      <td>472</td>\n",
       "      <td>ATM</td>\n",
       "      <td>55619</td>\n",
       "      <td>DOCK10</td>\n",
       "      <td>0.551148</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>72</th>\n",
       "      <td>472</td>\n",
       "      <td>ATM</td>\n",
       "      <td>54891</td>\n",
       "      <td>INO80D</td>\n",
       "      <td>0.550570</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>73</th>\n",
       "      <td>472</td>\n",
       "      <td>ATM</td>\n",
       "      <td>22990</td>\n",
       "      <td>PCNX1</td>\n",
       "      <td>0.550037</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>63</th>\n",
       "      <td>472</td>\n",
       "      <td>ATM</td>\n",
       "      <td>64430</td>\n",
       "      <td>PCNX4</td>\n",
       "      <td>0.554314</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>204</th>\n",
       "      <td>472</td>\n",
       "      <td>ATM</td>\n",
       "      <td>7767</td>\n",
       "      <td>ZNF224</td>\n",
       "      <td>0.500113</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>205 rows × 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                   EntrezID_Inactive Gene_Inactive  EntrezID_SL_Candidate  \\\n",
       "Gene_Inactive                                                               \n",
       "ATM           0                  472           ATM                   4297   \n",
       "              130                472           ATM                   1106   \n",
       "              131                472           ATM                   7621   \n",
       "              132                472           ATM              100101467   \n",
       "              133                472           ATM                  23506   \n",
       "...                              ...           ...                    ...   \n",
       "              71                 472           ATM                  55619   \n",
       "              72                 472           ATM                  54891   \n",
       "              73                 472           ATM                  22990   \n",
       "              63                 472           ATM                  64430   \n",
       "              204                472           ATM                   7767   \n",
       "\n",
       "                  Gene_SL_Candidate  Correlation  PValue  \n",
       "Gene_Inactive                                             \n",
       "ATM           0               KMT2A     0.719314     0.0  \n",
       "              130              CHD2     0.525927     0.0  \n",
       "              131             ZNF70     0.525772     0.0  \n",
       "              132           ZSCAN30     0.525512     0.0  \n",
       "              133            BICRAL     0.525503     0.0  \n",
       "...                             ...          ...     ...  \n",
       "              71             DOCK10     0.551148     0.0  \n",
       "              72             INO80D     0.550570     0.0  \n",
       "              73              PCNX1     0.550037     0.0  \n",
       "              63              PCNX4     0.554314     0.0  \n",
       "              204            ZNF224     0.500113     0.0  \n",
       "\n",
       "[205 rows x 6 columns]"
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "coexp_pancancer"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<br>\n",
    "4.1.2. Pairwise gene coexpression module on CCLE data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "coexp_CCLE=CoexpressionAnalysis(client, \"CCLE\", [472], cor_threshold, p_threshold, pval_correction) "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th>EntrezID_Inactive</th>\n",
       "      <th>Gene_Inactive</th>\n",
       "      <th>EntrezID_SL_Candidate</th>\n",
       "      <th>Gene_SL_Candidate</th>\n",
       "      <th>Correlation</th>\n",
       "      <th>PValue</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Gene_Inactive</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th rowspan=\"11\" valign=\"top\">ATM</th>\n",
       "      <th>0</th>\n",
       "      <td>472</td>\n",
       "      <td>ATM</td>\n",
       "      <td>4863</td>\n",
       "      <td>NPAT</td>\n",
       "      <td>0.671066</td>\n",
       "      <td>3.222075e-167</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>472</td>\n",
       "      <td>ATM</td>\n",
       "      <td>200576</td>\n",
       "      <td>PIKFYVE</td>\n",
       "      <td>0.637336</td>\n",
       "      <td>2.370146e-145</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>472</td>\n",
       "      <td>ATM</td>\n",
       "      <td>2145</td>\n",
       "      <td>EZH1</td>\n",
       "      <td>0.593870</td>\n",
       "      <td>6.509447e-121</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>472</td>\n",
       "      <td>ATM</td>\n",
       "      <td>84437</td>\n",
       "      <td>MSANTD4</td>\n",
       "      <td>0.579161</td>\n",
       "      <td>1.842351e-113</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>472</td>\n",
       "      <td>ATM</td>\n",
       "      <td>143684</td>\n",
       "      <td>FAM76B</td>\n",
       "      <td>0.571153</td>\n",
       "      <td>1.462356e-109</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>59</th>\n",
       "      <td>472</td>\n",
       "      <td>ATM</td>\n",
       "      <td>9044</td>\n",
       "      <td>BTAF1</td>\n",
       "      <td>0.502733</td>\n",
       "      <td>3.157164e-80</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>60</th>\n",
       "      <td>472</td>\n",
       "      <td>ATM</td>\n",
       "      <td>10181</td>\n",
       "      <td>RBM5</td>\n",
       "      <td>0.502723</td>\n",
       "      <td>3.183590e-80</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>61</th>\n",
       "      <td>472</td>\n",
       "      <td>ATM</td>\n",
       "      <td>10180</td>\n",
       "      <td>RBM6</td>\n",
       "      <td>0.502311</td>\n",
       "      <td>4.570113e-80</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>62</th>\n",
       "      <td>472</td>\n",
       "      <td>ATM</td>\n",
       "      <td>55599</td>\n",
       "      <td>RNPC3</td>\n",
       "      <td>0.500605</td>\n",
       "      <td>2.032599e-79</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>63</th>\n",
       "      <td>472</td>\n",
       "      <td>ATM</td>\n",
       "      <td>7753</td>\n",
       "      <td>ZNF202</td>\n",
       "      <td>0.500491</td>\n",
       "      <td>2.245387e-79</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>64 rows × 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                  EntrezID_Inactive Gene_Inactive  EntrezID_SL_Candidate  \\\n",
       "Gene_Inactive                                                              \n",
       "ATM           0                 472           ATM                   4863   \n",
       "              1                 472           ATM                 200576   \n",
       "              2                 472           ATM                   2145   \n",
       "              3                 472           ATM                  84437   \n",
       "              4                 472           ATM                 143684   \n",
       "...                             ...           ...                    ...   \n",
       "              59                472           ATM                   9044   \n",
       "              60                472           ATM                  10181   \n",
       "              61                472           ATM                  10180   \n",
       "              62                472           ATM                  55599   \n",
       "              63                472           ATM                   7753   \n",
       "\n",
       "                 Gene_SL_Candidate  Correlation         PValue  \n",
       "Gene_Inactive                                                   \n",
       "ATM           0               NPAT     0.671066  3.222075e-167  \n",
       "              1            PIKFYVE     0.637336  2.370146e-145  \n",
       "              2               EZH1     0.593870  6.509447e-121  \n",
       "              3            MSANTD4     0.579161  1.842351e-113  \n",
       "              4             FAM76B     0.571153  1.462356e-109  \n",
       "...                            ...          ...            ...  \n",
       "              59             BTAF1     0.502733   3.157164e-80  \n",
       "              60              RBM5     0.502723   3.183590e-80  \n",
       "              61              RBM6     0.502311   4.570113e-80  \n",
       "              62             RNPC3     0.500605   2.032599e-79  \n",
       "              63            ZNF202     0.500491   2.245387e-79  \n",
       "\n",
       "[64 rows x 6 columns]"
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "coexp_CCLE"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### 4.2. Genomic survival of fittest module"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "4.2.1. Genomic survival of fittest module on CCLE data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Empty DataFrame\n",
      "Columns: [Gene_Inactive, Gene_SL_Candidate, PValue]\n",
      "Index: []\n"
     ]
    }
   ],
   "source": [
    "sof_CCLE = SurvivalOfFittest(client, \"CCLE\", p_threshold, [472], input_mutations, percentile_threshold, cn_threshold, pval_correction)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Gene_Inactive</th>\n",
       "      <th>Gene_SL_Candidate</th>\n",
       "      <th>PValue</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [Gene_Inactive, Gene_SL_Candidate, PValue]\n",
       "Index: []"
      ]
     },
     "execution_count": 25,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sof_CCLE"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<br>\n",
    "4.2.2. Genomic survival of fittest module on PancancerAtlas"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "     Gene_Inactive Gene_SL_Candidate    PValue\n",
      "0              ATM      LOC100132111  0.000000\n",
      "1              ATM            OR10J5  0.000000\n",
      "2              ATM              DEDD  0.000000\n",
      "3              ATM           TOMM40L  0.000000\n",
      "4              ATM             UHMK1  0.000000\n",
      "...            ...               ...       ...\n",
      "2446           ATM              ATL1  0.049183\n",
      "2447           ATM      RP11-88I21.2  0.049570\n",
      "2448           ATM              GNG2  0.049795\n",
      "2449           ATM             GATA2  0.049817\n",
      "2450           ATM             FRMD6  0.049853\n",
      "\n",
      "[2451 rows x 3 columns]\n"
     ]
    }
   ],
   "source": [
    "sof_pancancer = SurvivalOfFittest(client, \"PanCancerAtlas\", p_threshold, [472], input_mutations, percentile_threshold, cn_threshold,pval_correction)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th>EntrezID_Inactive</th>\n",
       "      <th>Gene_Inactive</th>\n",
       "      <th>EntrezID_SL_Candidate</th>\n",
       "      <th>Gene_SL_Candidate</th>\n",
       "      <th>PValue</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Gene_Inactive</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th rowspan=\"11\" valign=\"top\">ATM</th>\n",
       "      <th>0</th>\n",
       "      <td>472</td>\n",
       "      <td>ATM</td>\n",
       "      <td>127385</td>\n",
       "      <td>OR10J5</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>58</th>\n",
       "      <td>472</td>\n",
       "      <td>ATM</td>\n",
       "      <td>57823</td>\n",
       "      <td>SLAMF7</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>57</th>\n",
       "      <td>472</td>\n",
       "      <td>ATM</td>\n",
       "      <td>2312</td>\n",
       "      <td>FLG</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>56</th>\n",
       "      <td>472</td>\n",
       "      <td>ATM</td>\n",
       "      <td>84824</td>\n",
       "      <td>FCRLA</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>55</th>\n",
       "      <td>472</td>\n",
       "      <td>ATM</td>\n",
       "      <td>5824</td>\n",
       "      <td>PEX19</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2179</th>\n",
       "      <td>472</td>\n",
       "      <td>ATM</td>\n",
       "      <td>6400</td>\n",
       "      <td>SEL1L</td>\n",
       "      <td>0.048261</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2180</th>\n",
       "      <td>472</td>\n",
       "      <td>ATM</td>\n",
       "      <td>51062</td>\n",
       "      <td>ATL1</td>\n",
       "      <td>0.049183</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2181</th>\n",
       "      <td>472</td>\n",
       "      <td>ATM</td>\n",
       "      <td>54331</td>\n",
       "      <td>GNG2</td>\n",
       "      <td>0.049795</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2182</th>\n",
       "      <td>472</td>\n",
       "      <td>ATM</td>\n",
       "      <td>2624</td>\n",
       "      <td>GATA2</td>\n",
       "      <td>0.049817</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2183</th>\n",
       "      <td>472</td>\n",
       "      <td>ATM</td>\n",
       "      <td>122786</td>\n",
       "      <td>FRMD6</td>\n",
       "      <td>0.049853</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2184 rows × 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                    EntrezID_Inactive Gene_Inactive  EntrezID_SL_Candidate  \\\n",
       "Gene_Inactive                                                                \n",
       "ATM           0                   472           ATM                 127385   \n",
       "              58                  472           ATM                  57823   \n",
       "              57                  472           ATM                   2312   \n",
       "              56                  472           ATM                  84824   \n",
       "              55                  472           ATM                   5824   \n",
       "...                               ...           ...                    ...   \n",
       "              2179                472           ATM                   6400   \n",
       "              2180                472           ATM                  51062   \n",
       "              2181                472           ATM                  54331   \n",
       "              2182                472           ATM                   2624   \n",
       "              2183                472           ATM                 122786   \n",
       "\n",
       "                   Gene_SL_Candidate    PValue  \n",
       "Gene_Inactive                                   \n",
       "ATM           0               OR10J5  0.000000  \n",
       "              58              SLAMF7  0.000000  \n",
       "              57                 FLG  0.000000  \n",
       "              56               FCRLA  0.000000  \n",
       "              55               PEX19  0.000000  \n",
       "...                              ...       ...  \n",
       "              2179             SEL1L  0.048261  \n",
       "              2180              ATL1  0.049183  \n",
       "              2181              GNG2  0.049795  \n",
       "              2182             GATA2  0.049817  \n",
       "              2183             FRMD6  0.049853  \n",
       "\n",
       "[2184 rows x 5 columns]"
      ]
     },
     "execution_count": 27,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sof_pancancer"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### 4.3. Functional examination inference module"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "4.3.1. CRISPR based functional examination inference module"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "crispr_result = FunctionalExamination(client, \"CRISPR\", p_threshold,[472], percentile_threshold, cn_threshold, 'none')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th>EntrezID_Inactive</th>\n",
       "      <th>Gene_Inactive</th>\n",
       "      <th>EntrezID_SL_Candidate</th>\n",
       "      <th>Gene_SL_Candidate</th>\n",
       "      <th>PValue</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Gene_Inactive</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th rowspan=\"11\" valign=\"top\">ATM</th>\n",
       "      <th>0</th>\n",
       "      <td>472</td>\n",
       "      <td>ATM</td>\n",
       "      <td>84811</td>\n",
       "      <td>BUD13</td>\n",
       "      <td>2.512259e-10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>472</td>\n",
       "      <td>ATM</td>\n",
       "      <td>2017</td>\n",
       "      <td>CTTN</td>\n",
       "      <td>3.285199e-10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>472</td>\n",
       "      <td>ATM</td>\n",
       "      <td>156</td>\n",
       "      <td>GRK2</td>\n",
       "      <td>3.697528e-10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>472</td>\n",
       "      <td>ATM</td>\n",
       "      <td>10235</td>\n",
       "      <td>RASGRP2</td>\n",
       "      <td>5.162614e-10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>472</td>\n",
       "      <td>ATM</td>\n",
       "      <td>6734</td>\n",
       "      <td>SRPRA</td>\n",
       "      <td>2.068264e-09</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2147</th>\n",
       "      <td>472</td>\n",
       "      <td>ATM</td>\n",
       "      <td>27248</td>\n",
       "      <td>ERLEC1</td>\n",
       "      <td>4.975831e-02</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2146</th>\n",
       "      <td>472</td>\n",
       "      <td>ATM</td>\n",
       "      <td>10874</td>\n",
       "      <td>NMU</td>\n",
       "      <td>4.975831e-02</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2148</th>\n",
       "      <td>472</td>\n",
       "      <td>ATM</td>\n",
       "      <td>2145</td>\n",
       "      <td>EZH1</td>\n",
       "      <td>4.981428e-02</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2149</th>\n",
       "      <td>472</td>\n",
       "      <td>ATM</td>\n",
       "      <td>55032</td>\n",
       "      <td>SLC35A5</td>\n",
       "      <td>4.982033e-02</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2150</th>\n",
       "      <td>472</td>\n",
       "      <td>ATM</td>\n",
       "      <td>128646</td>\n",
       "      <td>SIRPD</td>\n",
       "      <td>4.988241e-02</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2151 rows × 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                    EntrezID_Inactive Gene_Inactive  EntrezID_SL_Candidate  \\\n",
       "Gene_Inactive                                                                \n",
       "ATM           0                   472           ATM                  84811   \n",
       "              1                   472           ATM                   2017   \n",
       "              2                   472           ATM                    156   \n",
       "              3                   472           ATM                  10235   \n",
       "              4                   472           ATM                   6734   \n",
       "...                               ...           ...                    ...   \n",
       "              2147                472           ATM                  27248   \n",
       "              2146                472           ATM                  10874   \n",
       "              2148                472           ATM                   2145   \n",
       "              2149                472           ATM                  55032   \n",
       "              2150                472           ATM                 128646   \n",
       "\n",
       "                   Gene_SL_Candidate        PValue  \n",
       "Gene_Inactive                                       \n",
       "ATM           0                BUD13  2.512259e-10  \n",
       "              1                 CTTN  3.285199e-10  \n",
       "              2                 GRK2  3.697528e-10  \n",
       "              3              RASGRP2  5.162614e-10  \n",
       "              4                SRPRA  2.068264e-09  \n",
       "...                              ...           ...  \n",
       "              2147            ERLEC1  4.975831e-02  \n",
       "              2146               NMU  4.975831e-02  \n",
       "              2148              EZH1  4.981428e-02  \n",
       "              2149           SLC35A5  4.982033e-02  \n",
       "              2150             SIRPD  4.988241e-02  \n",
       "\n",
       "[2151 rows x 5 columns]"
      ]
     },
     "execution_count": 29,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "crispr_result"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<br>\n",
    "4.3.2. shRNA based functional examination inference module"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [],
   "source": [
    "siRNA_result = FunctionalExamination(client, \"siRNA\", p_threshold, [472], percentile_threshold, cn_threshold, 'none')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th>EntrezID_Inactive</th>\n",
       "      <th>Gene_Inactive</th>\n",
       "      <th>EntrezID_SL_Candidate</th>\n",
       "      <th>Gene_SL_Candidate</th>\n",
       "      <th>PValue</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Gene_Inactive</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th rowspan=\"11\" valign=\"top\">ATM</th>\n",
       "      <th>0</th>\n",
       "      <td>472</td>\n",
       "      <td>ATM</td>\n",
       "      <td>441087</td>\n",
       "      <td>LOC441087</td>\n",
       "      <td>0.000006</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>472</td>\n",
       "      <td>ATM</td>\n",
       "      <td>341640</td>\n",
       "      <td>FREM2</td>\n",
       "      <td>0.000007</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>472</td>\n",
       "      <td>ATM</td>\n",
       "      <td>89910</td>\n",
       "      <td>UBE3B</td>\n",
       "      <td>0.000008</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>472</td>\n",
       "      <td>ATM</td>\n",
       "      <td>6734</td>\n",
       "      <td>SRPRA</td>\n",
       "      <td>0.000016</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>472</td>\n",
       "      <td>ATM</td>\n",
       "      <td>29942</td>\n",
       "      <td>PURG</td>\n",
       "      <td>0.000023</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1206</th>\n",
       "      <td>472</td>\n",
       "      <td>ATM</td>\n",
       "      <td>26145</td>\n",
       "      <td>IRF2BP1</td>\n",
       "      <td>0.049769</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1207</th>\n",
       "      <td>472</td>\n",
       "      <td>ATM</td>\n",
       "      <td>85437</td>\n",
       "      <td>ZCRB1</td>\n",
       "      <td>0.049814</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1208</th>\n",
       "      <td>472</td>\n",
       "      <td>ATM</td>\n",
       "      <td>3123</td>\n",
       "      <td>HLA-DRB1</td>\n",
       "      <td>0.049859</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1209</th>\n",
       "      <td>472</td>\n",
       "      <td>ATM</td>\n",
       "      <td>115548</td>\n",
       "      <td>FCHO2</td>\n",
       "      <td>0.049892</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1210</th>\n",
       "      <td>472</td>\n",
       "      <td>ATM</td>\n",
       "      <td>7592</td>\n",
       "      <td>ZNF41</td>\n",
       "      <td>0.049979</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1211 rows × 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                    EntrezID_Inactive Gene_Inactive  EntrezID_SL_Candidate  \\\n",
       "Gene_Inactive                                                                \n",
       "ATM           0                   472           ATM                 441087   \n",
       "              1                   472           ATM                 341640   \n",
       "              2                   472           ATM                  89910   \n",
       "              3                   472           ATM                   6734   \n",
       "              4                   472           ATM                  29942   \n",
       "...                               ...           ...                    ...   \n",
       "              1206                472           ATM                  26145   \n",
       "              1207                472           ATM                  85437   \n",
       "              1208                472           ATM                   3123   \n",
       "              1209                472           ATM                 115548   \n",
       "              1210                472           ATM                   7592   \n",
       "\n",
       "                   Gene_SL_Candidate    PValue  \n",
       "Gene_Inactive                                   \n",
       "ATM           0            LOC441087  0.000006  \n",
       "              1                FREM2  0.000007  \n",
       "              2                UBE3B  0.000008  \n",
       "              3                SRPRA  0.000016  \n",
       "              4                 PURG  0.000023  \n",
       "...                              ...       ...  \n",
       "              1206           IRF2BP1  0.049769  \n",
       "              1207             ZCRB1  0.049814  \n",
       "              1208          HLA-DRB1  0.049859  \n",
       "              1209             FCHO2  0.049892  \n",
       "              1210             ZNF41  0.049979  \n",
       "\n",
       "[1211 rows x 5 columns]"
      ]
     },
     "execution_count": 31,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "siRNA_result"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 5. Integration of results"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "5.1. Integration of the pairwise Co-expression gene co-expression results on Pancancer and CCLE"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "coexpression_result = UnionResults([coexp_pancancer, coexp_CCLE])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>EntrezID_Inactive</th>\n",
       "      <th>Gene_Inactive</th>\n",
       "      <th>EntrezID_SL_Candidate</th>\n",
       "      <th>Gene_SL_Candidate</th>\n",
       "      <th>Correlation_0</th>\n",
       "      <th>PValue_0</th>\n",
       "      <th>Correlation_1</th>\n",
       "      <th>PValue_1</th>\n",
       "      <th>PValue</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>472</td>\n",
       "      <td>ATM</td>\n",
       "      <td>4297</td>\n",
       "      <td>KMT2A</td>\n",
       "      <td>0.719314</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.550823</td>\n",
       "      <td>3.966528e-100</td>\n",
       "      <td>0.000000e+00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>472</td>\n",
       "      <td>ATM</td>\n",
       "      <td>1106</td>\n",
       "      <td>CHD2</td>\n",
       "      <td>0.525927</td>\n",
       "      <td>0.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.000000e+00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>472</td>\n",
       "      <td>ATM</td>\n",
       "      <td>7621</td>\n",
       "      <td>ZNF70</td>\n",
       "      <td>0.525772</td>\n",
       "      <td>0.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.000000e+00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>472</td>\n",
       "      <td>ATM</td>\n",
       "      <td>100101467</td>\n",
       "      <td>ZSCAN30</td>\n",
       "      <td>0.525512</td>\n",
       "      <td>0.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.000000e+00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>472</td>\n",
       "      <td>ATM</td>\n",
       "      <td>23506</td>\n",
       "      <td>BICRAL</td>\n",
       "      <td>0.525503</td>\n",
       "      <td>0.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.000000e+00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>241</th>\n",
       "      <td>472</td>\n",
       "      <td>ATM</td>\n",
       "      <td>64766</td>\n",
       "      <td>S100PBP</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.504090</td>\n",
       "      <td>9.554430e-81</td>\n",
       "      <td>9.554430e-81</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>242</th>\n",
       "      <td>472</td>\n",
       "      <td>ATM</td>\n",
       "      <td>10181</td>\n",
       "      <td>RBM5</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.502723</td>\n",
       "      <td>3.183590e-80</td>\n",
       "      <td>3.183590e-80</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>243</th>\n",
       "      <td>472</td>\n",
       "      <td>ATM</td>\n",
       "      <td>10180</td>\n",
       "      <td>RBM6</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.502311</td>\n",
       "      <td>4.570113e-80</td>\n",
       "      <td>4.570113e-80</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>244</th>\n",
       "      <td>472</td>\n",
       "      <td>ATM</td>\n",
       "      <td>55599</td>\n",
       "      <td>RNPC3</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.500605</td>\n",
       "      <td>2.032599e-79</td>\n",
       "      <td>2.032599e-79</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>245</th>\n",
       "      <td>472</td>\n",
       "      <td>ATM</td>\n",
       "      <td>7753</td>\n",
       "      <td>ZNF202</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.500491</td>\n",
       "      <td>2.245387e-79</td>\n",
       "      <td>2.245387e-79</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>246 rows × 9 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     EntrezID_Inactive Gene_Inactive  EntrezID_SL_Candidate Gene_SL_Candidate  \\\n",
       "0                  472           ATM                   4297             KMT2A   \n",
       "1                  472           ATM                   1106              CHD2   \n",
       "2                  472           ATM                   7621             ZNF70   \n",
       "3                  472           ATM              100101467           ZSCAN30   \n",
       "4                  472           ATM                  23506            BICRAL   \n",
       "..                 ...           ...                    ...               ...   \n",
       "241                472           ATM                  64766           S100PBP   \n",
       "242                472           ATM                  10181              RBM5   \n",
       "243                472           ATM                  10180              RBM6   \n",
       "244                472           ATM                  55599             RNPC3   \n",
       "245                472           ATM                   7753            ZNF202   \n",
       "\n",
       "     Correlation_0  PValue_0  Correlation_1       PValue_1        PValue  \n",
       "0         0.719314       0.0       0.550823  3.966528e-100  0.000000e+00  \n",
       "1         0.525927       0.0            NaN            NaN  0.000000e+00  \n",
       "2         0.525772       0.0            NaN            NaN  0.000000e+00  \n",
       "3         0.525512       0.0            NaN            NaN  0.000000e+00  \n",
       "4         0.525503       0.0            NaN            NaN  0.000000e+00  \n",
       "..             ...       ...            ...            ...           ...  \n",
       "241            NaN       NaN       0.504090   9.554430e-81  9.554430e-81  \n",
       "242            NaN       NaN       0.502723   3.183590e-80  3.183590e-80  \n",
       "243            NaN       NaN       0.502311   4.570113e-80  4.570113e-80  \n",
       "244            NaN       NaN       0.500605   2.032599e-79  2.032599e-79  \n",
       "245            NaN       NaN       0.500491   2.245387e-79  2.245387e-79  \n",
       "\n",
       "[246 rows x 9 columns]"
      ]
     },
     "execution_count": 33,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "coexpression_result"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<br>\n",
    "5.2. Integration of Survival of Fittest results on Pancancer and CCLE"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "At least one of the dataframes is empty, please run only with nonempty dataframes\n"
     ]
    }
   ],
   "source": [
    "sof_result = UnionResults([sof_CCLE, sof_pancancer])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {},
   "outputs": [],
   "source": [
    "sof_result=sof_pancancer"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<br>\n",
    "5.3. Integration of shRNA and CRISPR based functional examination inference module."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {},
   "outputs": [],
   "source": [
    "functional_screening_result = UnionResults([crispr_result, siRNA_result])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<br>\n",
    "5.4. Merging the results from all three inference procedures"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {},
   "outputs": [],
   "source": [
    "all_merged_results = MergeResults([coexpression_result, sof_result, functional_screening_result])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>EntrezID_Inactive</th>\n",
       "      <th>Gene_Inactive</th>\n",
       "      <th>EntrezID_SL_Candidate</th>\n",
       "      <th>Gene_SL_Candidate</th>\n",
       "      <th>PValue</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>472</td>\n",
       "      <td>ATM</td>\n",
       "      <td>27185</td>\n",
       "      <td>DISC1</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>472</td>\n",
       "      <td>ATM</td>\n",
       "      <td>9859</td>\n",
       "      <td>CEP170</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   EntrezID_Inactive Gene_Inactive  EntrezID_SL_Candidate Gene_SL_Candidate  \\\n",
       "0                472           ATM                  27185             DISC1   \n",
       "1                472           ATM                   9859            CEP170   \n",
       "\n",
       "   PValue  \n",
       "0     0.0  \n",
       "1     0.0  "
      ]
     },
     "execution_count": 38,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "all_merged_results"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Results are saved in excel file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {},
   "outputs": [],
   "source": [
    "WriteToExcel(\"DAISY_results.xlsx\", [coexp_pancancer,  all_merged_results],[\"Co-exp_Pancancer\", \"All\"])\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
