{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#  DAISY- the DAta-mIning SYnthetic-lethality-identification pipeline\n",
    "\n",
    "Please cite: \n",
    "For Implementation: \n",
    "\n",
    "Our paper,\n",
    "\n",
    "For DAISY algorithm: \n",
    "\n",
    "Jerby-Arnon, L., Pfetzer, N., Waldman, Y. Y., McGarry, L., James, D., Shanks, E., ... & Gottlieb, E. (2014). Predicting cancer-specific vulnerability via data-driven detection of synthetic lethality. Cell, 158(5), 1199-1209.\n",
    "\n",
    "For CCLE Omics data:\n",
    "\n",
    "Ghandi, M., Huang, F.W., Jané-Valbuena, J. et al. Next-generation characterization of the Cancer Cell Line Encyclopedia. Nature 569, 503–508 (2019). https://doi.org/10.1038/s41586-019-1186-3\n",
    "\n",
    "For CRISPR Data: \n",
    "\n",
    "Robin M. Meyers, Jordan G. Bryan, James M. McFarland, Barbara A. Weir, ... David E. Root, William C. Hahn, Aviad Tsherniak. Computational correction of copy number effect improves specificity of CRISPR-Cas9 essentiality screens in cancer cells. Nature Genetics 2017 October 49:1779–1784. doi:10.1038/ng.3984\n",
    "\n",
    "Dempster, J. M., Rossen, J., Kazachkova, M., Pan, J., Kugener, G., Root, D. E., & Tsherniak, A. (2019). Extracting Biological Insights from the Project Achilles Genome-Scale CRISPR Screens in Cancer Cell Lines. BioRxiv, 720243.\n",
    "\n",
    "For shRNA Data:\n",
    "\n",
    "\n",
    "This notebook is a reimplementation of DAISY Synthetic Lethal Pair Prediction Algorithm\n",
    "\n",
    "It consists 3 modules: \n",
    "\n",
    "1. SL candidate determination using gene co-expression\n",
    "2. SL candidate determination using survival of fittest\n",
    "3. SL candidate determination using CRISPR and ShRNA experiment\n",
    "\n",
    "* The results from the three modules were then aggregated into one ranked list of candidate SL pairs\n",
    "\n",
    "Input Parameters\n",
    "* Cancer type \n",
    "* The genes whose SL partners are seeked\n",
    "\n",
    "Input Data (available in bigquery tables)\n",
    "* Gene expression data \n",
    "* Gene mutation data\n",
    "* Copy number variation data\n",
    "* Gene effect data (CRISPR)\n",
    "* Gene Dependency scores data (shRNA)\n",
    "\n",
    "Output\n",
    "* Ranked list of candidate SL pairs\n",
    "\n",
    "Please contact Bahar Tercan btercan@systemsbiology.org for your questions and detailed information. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 192,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Once deleted, variables cannot be recovered. Proceed (y/[n])? y\n"
     ]
    }
   ],
   "source": [
    "reset "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 193,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'/Users/bahar/Desktop/Boris__revised_daisy/DAISY_pipeline'"
      ]
     },
     "execution_count": 193,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pwd"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 1. Import python libraries required\n",
    "The required libraries are imported. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 194,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "from datetime import datetime\n",
    "import sys\n",
    "sys.path.append('../scripts/') #need to add \"scripts\" directory in a parent directory \n",
    "from google.cloud import bigquery\n",
    "import importlib\n",
    "import pandas as pd\n",
    "import DAISY_operations2\n",
    "importlib.reload(DAISY_operations2)\n",
    "from DAISY_operations2 import *\n",
    "from helper_functions import *\n",
    "from BIGQUERY_operations import *\n",
    "import ipywidgets as widgets"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 195,
   "metadata": {},
   "outputs": [],
   "source": [
    "if not sys.warnoptions:\n",
    "    import warnings\n",
    "    warnings.simplefilter(\"ignore\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 2. Sign in Google Bigquery with the project id\n",
    "\n",
    "Bigquery connection\n",
    "Please replace syntheticlethality with your project name"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 196,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Traceback (most recent call last):\r\n",
      "  File \"/Users/bahar/Downloads/google-cloud-sdk/lib/gcloud.py\", line 104, in <module>\r\n",
      "    main()\r\n",
      "  File \"/Users/bahar/Downloads/google-cloud-sdk/lib/gcloud.py\", line 100, in main\r\n",
      "    sys.exit(gcloud_main.main())\r\n",
      "  File \"/Users/bahar/Downloads/google-cloud-sdk/lib/googlecloudsdk/gcloud_main.py\", line 171, in main\r\n",
      "    gcloud_cli = CreateCLI([])\r\n",
      "  File \"/Users/bahar/Downloads/google-cloud-sdk/lib/googlecloudsdk/gcloud_main.py\", line 151, in CreateCLI\r\n",
      "    generated_cli = loader.Generate()\r\n",
      "  File \"/Users/bahar/Downloads/google-cloud-sdk/lib/googlecloudsdk/calliope/cli.py\", line 504, in Generate\r\n",
      "    cli = self.__MakeCLI(top_group)\r\n",
      "  File \"/Users/bahar/Downloads/google-cloud-sdk/lib/googlecloudsdk/calliope/cli.py\", line 674, in __MakeCLI\r\n",
      "    log.AddFileLogging(self.__logs_dir)\r\n",
      "  File \"/Users/bahar/Downloads/google-cloud-sdk/lib/googlecloudsdk/core/log.py\", line 1039, in AddFileLogging\r\n",
      "    _log_manager.AddLogsDir(logs_dir=logs_dir)\r\n",
      "  File \"/Users/bahar/Downloads/google-cloud-sdk/lib/googlecloudsdk/core/log.py\", line 683, in AddLogsDir\r\n",
      "    self._CleanUpLogs(logs_dir)\r\n",
      "  File \"/Users/bahar/Downloads/google-cloud-sdk/lib/googlecloudsdk/core/log.py\", line 711, in _CleanUpLogs\r\n",
      "    self._CleanLogsDir(logs_dir)\r\n",
      "  File \"/Users/bahar/Downloads/google-cloud-sdk/lib/googlecloudsdk/core/log.py\", line 737, in _CleanLogsDir\r\n",
      "    os.remove(log_file_path)\r\n",
      "PermissionError: [Errno 13] Permission denied: '/Users/bahar/.config/gcloud/logs/2020.08.04/12.16.05.276251.log'\r\n"
     ]
    }
   ],
   "source": [
    "project_id='syntheticlethality'\n",
    "client = bigquery.Client(project_id)\n",
    "#client = bigquery.Client(credentials=credentials, project=credentials.project_id)\n",
    "\n",
    "!gcloud auth login"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 4. Prediction of synthetic lethal partners using different modules on DAISY\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "There are three modules for synthetic lethal pair inferences on DAISY : 1. Pairwise gene coexpression, 2. Genomic survival of the fittest. 3. shRNA or CRISPR based functional examination. You can get more information in the original paper : https://www.sciencedirect.com/science/article/pii/S0092867414009775.\n",
    "\n",
    "In pairwise gene coexpression module and genomic survial of the fittest module, we will use PancancerAtlas and CCLE data.<br>\n",
    "In functional examination module, we will use CRISPR and shRNA data. <br>\n",
    "\n",
    "Python codes for each module are built in our internal library (../scripts/SL_library.py) which was already imported at the beginning. \n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### 4.0. Default parameters for DAISY, you can edit them"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 197,
   "metadata": {},
   "outputs": [],
   "source": [
    "input_mutations = ['Nonsense_Mutation', 'Frame_Shift_Ins', 'Frame_Shift_Del'] # DAISY default parameters\n",
    "percentile_threshold = 10\n",
    "cn_threshold = -0.3 \n",
    "cor_threshold = 0.5\n",
    "p_threshold = 0.1\n",
    "pval_correction = 'Bonferroni'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 198,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "e60e962501d641ee8c5269b2f85415a5",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "SelectMultiple(description='Tumor type', options=('pancancer', 'CHOL', 'BLCA', 'GBM', 'BRCA', 'CESC', 'COAD', …"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "\n",
    "TCGA_list=GetTCGASubtypes(client)\n",
    "TCGA_list = [i for i in TCGA_list if i]\n",
    "\n",
    "tumor_type = widgets.SelectMultiple(\n",
    "    options=['pancancer'] + TCGA_list  ,\n",
    "    value=[],\n",
    "    description='Tumor type',\n",
    "    disabled=False\n",
    ")\n",
    "display(tumor_type)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### 4.1. Pairwise gene coexpression module"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "4.1.1. Pairwise gene coexpression module on PancancerAtlas."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 199,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th>Inactive</th>\n",
       "      <th>InactiveDB</th>\n",
       "      <th>SL_Candidate</th>\n",
       "      <th>#Samples</th>\n",
       "      <th>Correlation</th>\n",
       "      <th>PValue</th>\n",
       "      <th>FDR</th>\n",
       "      <th>Tissue</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Inactive</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th rowspan=\"11\" valign=\"top\">BRCA1</th>\n",
       "      <th>0</th>\n",
       "      <td>BRCA1</td>\n",
       "      <td>BRCA1</td>\n",
       "      <td>ATAD5</td>\n",
       "      <td>9953</td>\n",
       "      <td>0.827015</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>['pancancer']</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>288</th>\n",
       "      <td>BRCA1</td>\n",
       "      <td>BRCA1</td>\n",
       "      <td>UTP6</td>\n",
       "      <td>9953</td>\n",
       "      <td>0.580536</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>['pancancer']</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>287</th>\n",
       "      <td>BRCA1</td>\n",
       "      <td>BRCA1</td>\n",
       "      <td>C9orf100</td>\n",
       "      <td>9953</td>\n",
       "      <td>0.580693</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>['pancancer']</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>286</th>\n",
       "      <td>BRCA1</td>\n",
       "      <td>BRCA1</td>\n",
       "      <td>DEK</td>\n",
       "      <td>9953</td>\n",
       "      <td>0.580963</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>['pancancer']</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>285</th>\n",
       "      <td>BRCA1</td>\n",
       "      <td>BRCA1</td>\n",
       "      <td>ZNF207</td>\n",
       "      <td>9953</td>\n",
       "      <td>0.580978</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>['pancancer']</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>136</th>\n",
       "      <td>BRCA1</td>\n",
       "      <td>BRCA1</td>\n",
       "      <td>C13orf34</td>\n",
       "      <td>9953</td>\n",
       "      <td>0.693333</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>['pancancer']</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>135</th>\n",
       "      <td>BRCA1</td>\n",
       "      <td>BRCA1</td>\n",
       "      <td>E2F8</td>\n",
       "      <td>9953</td>\n",
       "      <td>0.693417</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>['pancancer']</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>134</th>\n",
       "      <td>BRCA1</td>\n",
       "      <td>BRCA1</td>\n",
       "      <td>MYBL2</td>\n",
       "      <td>9953</td>\n",
       "      <td>0.694103</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>['pancancer']</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>132</th>\n",
       "      <td>BRCA1</td>\n",
       "      <td>BRCA1</td>\n",
       "      <td>FEN1</td>\n",
       "      <td>9953</td>\n",
       "      <td>0.697051</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>['pancancer']</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>424</th>\n",
       "      <td>BRCA1</td>\n",
       "      <td>BRCA1</td>\n",
       "      <td>LTV1</td>\n",
       "      <td>9953</td>\n",
       "      <td>0.500224</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>['pancancer']</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>425 rows × 8 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "             Inactive InactiveDB SL_Candidate  #Samples  Correlation  PValue  \\\n",
       "Inactive                                                                       \n",
       "BRCA1    0      BRCA1      BRCA1        ATAD5      9953     0.827015     0.0   \n",
       "         288    BRCA1      BRCA1         UTP6      9953     0.580536     0.0   \n",
       "         287    BRCA1      BRCA1     C9orf100      9953     0.580693     0.0   \n",
       "         286    BRCA1      BRCA1          DEK      9953     0.580963     0.0   \n",
       "         285    BRCA1      BRCA1       ZNF207      9953     0.580978     0.0   \n",
       "...               ...        ...          ...       ...          ...     ...   \n",
       "         136    BRCA1      BRCA1     C13orf34      9953     0.693333     0.0   \n",
       "         135    BRCA1      BRCA1         E2F8      9953     0.693417     0.0   \n",
       "         134    BRCA1      BRCA1        MYBL2      9953     0.694103     0.0   \n",
       "         132    BRCA1      BRCA1         FEN1      9953     0.697051     0.0   \n",
       "         424    BRCA1      BRCA1         LTV1      9953     0.500224     0.0   \n",
       "\n",
       "              FDR         Tissue  \n",
       "Inactive                          \n",
       "BRCA1    0    0.0  ['pancancer']  \n",
       "         288  0.0  ['pancancer']  \n",
       "         287  0.0  ['pancancer']  \n",
       "         286  0.0  ['pancancer']  \n",
       "         285  0.0  ['pancancer']  \n",
       "...           ...            ...  \n",
       "         136  0.0  ['pancancer']  \n",
       "         135  0.0  ['pancancer']  \n",
       "         134  0.0  ['pancancer']  \n",
       "         132  0.0  ['pancancer']  \n",
       "         424  0.0  ['pancancer']  \n",
       "\n",
       "[425 rows x 8 columns]"
      ]
     },
     "execution_count": 199,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import DAISY_operations2\n",
    "importlib.reload(DAISY_operations2)\n",
    "from DAISY_operations2 import *\n",
    "\n",
    "coexp_pancancer = CoexpressionAnalysis(client, 'SL', \"PanCancerAtlas\", ['BRCA1'] , pval_correction, list(tumor_type.value))\n",
    "report=coexp_pancancer.loc[(coexp_pancancer['FDR'] < p_threshold)&(coexp_pancancer['Correlation'] > cor_threshold)]\n",
    "if report.shape[0]>1:\n",
    "    coexp_pancancer_report=report.groupby('Inactive').apply(lambda x: x.sort_values('FDR'))\n",
    "    \n",
    "coexp_pancancer_report    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 200,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Inactive</th>\n",
       "      <th>InactiveDB</th>\n",
       "      <th>SL_Candidate</th>\n",
       "      <th>#Samples</th>\n",
       "      <th>Correlation</th>\n",
       "      <th>PValue</th>\n",
       "      <th>FDR</th>\n",
       "      <th>Tissue</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>587</th>\n",
       "      <td>BRCA1</td>\n",
       "      <td>BRCA1</td>\n",
       "      <td>PARP1</td>\n",
       "      <td>9953</td>\n",
       "      <td>0.452299</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>['pancancer']</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    Inactive InactiveDB SL_Candidate  #Samples  Correlation  PValue  FDR  \\\n",
       "587    BRCA1      BRCA1        PARP1      9953     0.452299     0.0  0.0   \n",
       "\n",
       "            Tissue  \n",
       "587  ['pancancer']  "
      ]
     },
     "execution_count": 200,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "coexp_pancancer.loc[(coexp_pancancer['Inactive']=='BRCA1')&(coexp_pancancer['SL_Candidate']=='PARP1'), ]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 201,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Inactive</th>\n",
       "      <th>InactiveDB</th>\n",
       "      <th>SL_Candidate</th>\n",
       "      <th>#Samples</th>\n",
       "      <th>Correlation</th>\n",
       "      <th>PValue</th>\n",
       "      <th>FDR</th>\n",
       "      <th>Tissue</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>2092</th>\n",
       "      <td>BRCA1</td>\n",
       "      <td>BRCA1</td>\n",
       "      <td>PARP2</td>\n",
       "      <td>9953</td>\n",
       "      <td>0.266241</td>\n",
       "      <td>4.067601e-161</td>\n",
       "      <td>8.249908e-157</td>\n",
       "      <td>['pancancer']</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     Inactive InactiveDB SL_Candidate  #Samples  Correlation         PValue  \\\n",
       "2092    BRCA1      BRCA1        PARP2      9953     0.266241  4.067601e-161   \n",
       "\n",
       "                FDR         Tissue  \n",
       "2092  8.249908e-157  ['pancancer']  "
      ]
     },
     "execution_count": 201,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "coexp_pancancer.loc[(coexp_pancancer['Inactive']=='BRCA1')&(coexp_pancancer['SL_Candidate']=='PARP2'), ]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<br>\n",
    "4.1.2. Pairwise gene coexpression module on CCLE data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 202,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th>Inactive</th>\n",
       "      <th>InactiveDB</th>\n",
       "      <th>SL_Candidate</th>\n",
       "      <th>#Samples</th>\n",
       "      <th>Correlation</th>\n",
       "      <th>PValue</th>\n",
       "      <th>FDR</th>\n",
       "      <th>Tissue</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Inactive</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th rowspan=\"11\" valign=\"top\">BRCA1</th>\n",
       "      <th>0</th>\n",
       "      <td>BRCA1</td>\n",
       "      <td>BRCA1</td>\n",
       "      <td>ATAD5</td>\n",
       "      <td>1297</td>\n",
       "      <td>0.746083</td>\n",
       "      <td>6.359384e-231</td>\n",
       "      <td>1.215851e-226</td>\n",
       "      <td>['pancancer']</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>BRCA1</td>\n",
       "      <td>BRCA1</td>\n",
       "      <td>FANCI</td>\n",
       "      <td>1297</td>\n",
       "      <td>0.718720</td>\n",
       "      <td>1.404211e-206</td>\n",
       "      <td>2.684711e-202</td>\n",
       "      <td>['pancancer']</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>BRCA1</td>\n",
       "      <td>BRCA1</td>\n",
       "      <td>C17orf53</td>\n",
       "      <td>1297</td>\n",
       "      <td>0.711617</td>\n",
       "      <td>9.940688e-201</td>\n",
       "      <td>1.900560e-196</td>\n",
       "      <td>['pancancer']</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>BRCA1</td>\n",
       "      <td>BRCA1</td>\n",
       "      <td>CHAF1A</td>\n",
       "      <td>1297</td>\n",
       "      <td>0.704579</td>\n",
       "      <td>4.185098e-195</td>\n",
       "      <td>8.001489e-191</td>\n",
       "      <td>['pancancer']</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>BRCA1</td>\n",
       "      <td>BRCA1</td>\n",
       "      <td>TOPBP1</td>\n",
       "      <td>1297</td>\n",
       "      <td>0.702716</td>\n",
       "      <td>1.209211e-193</td>\n",
       "      <td>2.311890e-189</td>\n",
       "      <td>['pancancer']</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>522</th>\n",
       "      <td>BRCA1</td>\n",
       "      <td>BRCA1</td>\n",
       "      <td>RBM33</td>\n",
       "      <td>1297</td>\n",
       "      <td>0.500611</td>\n",
       "      <td>3.360346e-83</td>\n",
       "      <td>6.424646e-79</td>\n",
       "      <td>['pancancer']</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>523</th>\n",
       "      <td>BRCA1</td>\n",
       "      <td>BRCA1</td>\n",
       "      <td>SLC25A19</td>\n",
       "      <td>1297</td>\n",
       "      <td>0.500536</td>\n",
       "      <td>3.584723e-83</td>\n",
       "      <td>6.853632e-79</td>\n",
       "      <td>['pancancer']</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>524</th>\n",
       "      <td>BRCA1</td>\n",
       "      <td>BRCA1</td>\n",
       "      <td>PRRC2C</td>\n",
       "      <td>1297</td>\n",
       "      <td>0.500421</td>\n",
       "      <td>3.962322e-83</td>\n",
       "      <td>7.575563e-79</td>\n",
       "      <td>['pancancer']</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>525</th>\n",
       "      <td>BRCA1</td>\n",
       "      <td>BRCA1</td>\n",
       "      <td>HNRNPA3</td>\n",
       "      <td>1297</td>\n",
       "      <td>0.500273</td>\n",
       "      <td>4.505149e-83</td>\n",
       "      <td>8.613395e-79</td>\n",
       "      <td>['pancancer']</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>526</th>\n",
       "      <td>BRCA1</td>\n",
       "      <td>BRCA1</td>\n",
       "      <td>SMG1</td>\n",
       "      <td>1297</td>\n",
       "      <td>0.500050</td>\n",
       "      <td>5.464105e-83</td>\n",
       "      <td>1.044682e-78</td>\n",
       "      <td>['pancancer']</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>527 rows × 8 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "             Inactive InactiveDB SL_Candidate  #Samples  Correlation  \\\n",
       "Inactive                                                               \n",
       "BRCA1    0      BRCA1      BRCA1        ATAD5      1297     0.746083   \n",
       "         1      BRCA1      BRCA1        FANCI      1297     0.718720   \n",
       "         2      BRCA1      BRCA1     C17orf53      1297     0.711617   \n",
       "         3      BRCA1      BRCA1       CHAF1A      1297     0.704579   \n",
       "         4      BRCA1      BRCA1       TOPBP1      1297     0.702716   \n",
       "...               ...        ...          ...       ...          ...   \n",
       "         522    BRCA1      BRCA1        RBM33      1297     0.500611   \n",
       "         523    BRCA1      BRCA1     SLC25A19      1297     0.500536   \n",
       "         524    BRCA1      BRCA1       PRRC2C      1297     0.500421   \n",
       "         525    BRCA1      BRCA1      HNRNPA3      1297     0.500273   \n",
       "         526    BRCA1      BRCA1         SMG1      1297     0.500050   \n",
       "\n",
       "                     PValue            FDR         Tissue  \n",
       "Inactive                                                   \n",
       "BRCA1    0    6.359384e-231  1.215851e-226  ['pancancer']  \n",
       "         1    1.404211e-206  2.684711e-202  ['pancancer']  \n",
       "         2    9.940688e-201  1.900560e-196  ['pancancer']  \n",
       "         3    4.185098e-195  8.001489e-191  ['pancancer']  \n",
       "         4    1.209211e-193  2.311890e-189  ['pancancer']  \n",
       "...                     ...            ...            ...  \n",
       "         522   3.360346e-83   6.424646e-79  ['pancancer']  \n",
       "         523   3.584723e-83   6.853632e-79  ['pancancer']  \n",
       "         524   3.962322e-83   7.575563e-79  ['pancancer']  \n",
       "         525   4.505149e-83   8.613395e-79  ['pancancer']  \n",
       "         526   5.464105e-83   1.044682e-78  ['pancancer']  \n",
       "\n",
       "[527 rows x 8 columns]"
      ]
     },
     "execution_count": 202,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "coexp_CCLE=CoexpressionAnalysis(client, 'SL', 'CCLE', ['BRCA1'], pval_correction, list(tumor_type.value))\n",
    "report=coexp_CCLE.loc[(coexp_CCLE['FDR'] < p_threshold)&(coexp_CCLE['Correlation'] > cor_threshold)]\n",
    "if report.shape[0]>1:\n",
    "    coexp_CCLE_report=report.groupby('Inactive').apply(lambda x: x.sort_values('FDR'))\n",
    "coexp_CCLE_report"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 203,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Inactive</th>\n",
       "      <th>InactiveDB</th>\n",
       "      <th>SL_Candidate</th>\n",
       "      <th>#Samples</th>\n",
       "      <th>Correlation</th>\n",
       "      <th>PValue</th>\n",
       "      <th>FDR</th>\n",
       "      <th>Tissue</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>135</th>\n",
       "      <td>BRCA1</td>\n",
       "      <td>BRCA1</td>\n",
       "      <td>PARP1</td>\n",
       "      <td>1297</td>\n",
       "      <td>0.591197</td>\n",
       "      <td>4.599640e-123</td>\n",
       "      <td>8.794052e-119</td>\n",
       "      <td>['pancancer']</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    Inactive InactiveDB SL_Candidate  #Samples  Correlation         PValue  \\\n",
       "135    BRCA1      BRCA1        PARP1      1297     0.591197  4.599640e-123   \n",
       "\n",
       "               FDR         Tissue  \n",
       "135  8.794052e-119  ['pancancer']  "
      ]
     },
     "execution_count": 203,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "coexp_CCLE.loc[(coexp_CCLE['Inactive']=='BRCA1')&(coexp_CCLE['SL_Candidate']=='PARP1'), ]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 204,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Inactive</th>\n",
       "      <th>InactiveDB</th>\n",
       "      <th>SL_Candidate</th>\n",
       "      <th>#Samples</th>\n",
       "      <th>Correlation</th>\n",
       "      <th>PValue</th>\n",
       "      <th>FDR</th>\n",
       "      <th>Tissue</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1123</th>\n",
       "      <td>BRCA1</td>\n",
       "      <td>BRCA1</td>\n",
       "      <td>PARP2</td>\n",
       "      <td>1297</td>\n",
       "      <td>0.434211</td>\n",
       "      <td>9.154777e-61</td>\n",
       "      <td>1.750302e-56</td>\n",
       "      <td>['pancancer']</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     Inactive InactiveDB SL_Candidate  #Samples  Correlation        PValue  \\\n",
       "1123    BRCA1      BRCA1        PARP2      1297     0.434211  9.154777e-61   \n",
       "\n",
       "               FDR         Tissue  \n",
       "1123  1.750302e-56  ['pancancer']  "
      ]
     },
     "execution_count": 204,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "coexp_CCLE.loc[(coexp_CCLE['Inactive']=='BRCA1')&(coexp_CCLE['SL_Candidate']=='PARP2'), ]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### 4.2. Genomic survival of fittest module"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "4.2.1. Genomic survival of fittest module on CCLE data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 205,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "import DAISY_operations2\n",
    "importlib.reload(DAISY_operations2)\n",
    "from DAISY_operations2 import *\n",
    "sof_CCLE = SurvivalOfFittest(client, 'SL', \"CCLE\", ['BRCA1'],  percentile_threshold, cn_threshold, pval_correction, list(tumor_type.value), input_mutations)\n",
    "report=sof_CCLE.loc[(sof_CCLE['FDR'] < p_threshold),]\n",
    "                      \n",
    "if report.shape[0]>1:\n",
    "    sof_ccle_report=report.groupby('Inactive').apply(lambda x: x.sort_values('FDR'))\n",
    "    sof_ccle_report\n",
    "# no result returned"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 207,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Inactive</th>\n",
       "      <th>InactiveDB</th>\n",
       "      <th>SL_Candidate</th>\n",
       "      <th>#InactiveSamples</th>\n",
       "      <th>#Samples</th>\n",
       "      <th>U1</th>\n",
       "      <th>PValue</th>\n",
       "      <th>FDR</th>\n",
       "      <th>Tissue</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1400</th>\n",
       "      <td>BRCA1</td>\n",
       "      <td>BRCA1</td>\n",
       "      <td>PARP1</td>\n",
       "      <td>38</td>\n",
       "      <td>1284</td>\n",
       "      <td>29766.0</td>\n",
       "      <td>0.00341</td>\n",
       "      <td>1.0</td>\n",
       "      <td>['pancancer']</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     Inactive InactiveDB SL_Candidate  #InactiveSamples  #Samples       U1  \\\n",
       "1400    BRCA1      BRCA1        PARP1                38      1284  29766.0   \n",
       "\n",
       "       PValue  FDR         Tissue  \n",
       "1400  0.00341  1.0  ['pancancer']  "
      ]
     },
     "execution_count": 207,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sof_CCLE.loc[(sof_CCLE['Inactive']=='BRCA1')&(sof_CCLE['SL_Candidate']=='PARP1'), ]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 208,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Inactive</th>\n",
       "      <th>InactiveDB</th>\n",
       "      <th>SL_Candidate</th>\n",
       "      <th>#InactiveSamples</th>\n",
       "      <th>#Samples</th>\n",
       "      <th>U1</th>\n",
       "      <th>PValue</th>\n",
       "      <th>FDR</th>\n",
       "      <th>Tissue</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>5329</th>\n",
       "      <td>BRCA1</td>\n",
       "      <td>BRCA1</td>\n",
       "      <td>PARP2</td>\n",
       "      <td>38</td>\n",
       "      <td>1284</td>\n",
       "      <td>27416.0</td>\n",
       "      <td>0.048271</td>\n",
       "      <td>1.0</td>\n",
       "      <td>['pancancer']</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     Inactive InactiveDB SL_Candidate  #InactiveSamples  #Samples       U1  \\\n",
       "5329    BRCA1      BRCA1        PARP2                38      1284  27416.0   \n",
       "\n",
       "        PValue  FDR         Tissue  \n",
       "5329  0.048271  1.0  ['pancancer']  "
      ]
     },
     "execution_count": 208,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sof_CCLE.loc[(sof_CCLE['Inactive']=='BRCA1')&(sof_CCLE['SL_Candidate']=='PARP2'), ]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 209,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th>Inactive</th>\n",
       "      <th>InactiveDB</th>\n",
       "      <th>SL_Candidate</th>\n",
       "      <th>#InactiveSamples</th>\n",
       "      <th>#Samples</th>\n",
       "      <th>U1</th>\n",
       "      <th>PValue</th>\n",
       "      <th>FDR</th>\n",
       "      <th>Tissue</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Inactive</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th rowspan=\"11\" valign=\"top\">BRCA1</th>\n",
       "      <th>0</th>\n",
       "      <td>BRCA1</td>\n",
       "      <td>BRCA1</td>\n",
       "      <td>NWD1</td>\n",
       "      <td>234</td>\n",
       "      <td>8930</td>\n",
       "      <td>1305850.5</td>\n",
       "      <td>6.250556e-14</td>\n",
       "      <td>1.570640e-09</td>\n",
       "      <td>['pancancer']</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>BRCA1</td>\n",
       "      <td>BRCA1</td>\n",
       "      <td>KIAA1683</td>\n",
       "      <td>234</td>\n",
       "      <td>8930</td>\n",
       "      <td>1305687.5</td>\n",
       "      <td>6.450396e-14</td>\n",
       "      <td>1.620855e-09</td>\n",
       "      <td>['pancancer']</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>BRCA1</td>\n",
       "      <td>BRCA1</td>\n",
       "      <td>JUND</td>\n",
       "      <td>234</td>\n",
       "      <td>8930</td>\n",
       "      <td>1305525.5</td>\n",
       "      <td>6.655787e-14</td>\n",
       "      <td>1.672466e-09</td>\n",
       "      <td>['pancancer']</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>BRCA1</td>\n",
       "      <td>BRCA1</td>\n",
       "      <td>MIR3188</td>\n",
       "      <td>234</td>\n",
       "      <td>8930</td>\n",
       "      <td>1305525.5</td>\n",
       "      <td>6.655787e-14</td>\n",
       "      <td>1.672466e-09</td>\n",
       "      <td>['pancancer']</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>BRCA1</td>\n",
       "      <td>BRCA1</td>\n",
       "      <td>F2RL3</td>\n",
       "      <td>234</td>\n",
       "      <td>8930</td>\n",
       "      <td>1305301.0</td>\n",
       "      <td>6.955547e-14</td>\n",
       "      <td>1.747790e-09</td>\n",
       "      <td>['pancancer']</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2142</th>\n",
       "      <td>BRCA1</td>\n",
       "      <td>BRCA1</td>\n",
       "      <td>OTOP1</td>\n",
       "      <td>234</td>\n",
       "      <td>8930</td>\n",
       "      <td>1191358.5</td>\n",
       "      <td>3.923963e-06</td>\n",
       "      <td>9.860135e-02</td>\n",
       "      <td>['pancancer']</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2143</th>\n",
       "      <td>BRCA1</td>\n",
       "      <td>BRCA1</td>\n",
       "      <td>ZNF260</td>\n",
       "      <td>234</td>\n",
       "      <td>8930</td>\n",
       "      <td>1191282.5</td>\n",
       "      <td>3.959944e-06</td>\n",
       "      <td>9.950548e-02</td>\n",
       "      <td>['pancancer']</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2144</th>\n",
       "      <td>BRCA1</td>\n",
       "      <td>BRCA1</td>\n",
       "      <td>GRSF1</td>\n",
       "      <td>234</td>\n",
       "      <td>8930</td>\n",
       "      <td>1191265.5</td>\n",
       "      <td>3.968036e-06</td>\n",
       "      <td>9.970880e-02</td>\n",
       "      <td>['pancancer']</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2145</th>\n",
       "      <td>BRCA1</td>\n",
       "      <td>BRCA1</td>\n",
       "      <td>NRXN3</td>\n",
       "      <td>234</td>\n",
       "      <td>8930</td>\n",
       "      <td>1191248.0</td>\n",
       "      <td>3.976382e-06</td>\n",
       "      <td>9.991852e-02</td>\n",
       "      <td>['pancancer']</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2146</th>\n",
       "      <td>BRCA1</td>\n",
       "      <td>BRCA1</td>\n",
       "      <td>FAM177A1</td>\n",
       "      <td>234</td>\n",
       "      <td>8930</td>\n",
       "      <td>1191243.0</td>\n",
       "      <td>3.978769e-06</td>\n",
       "      <td>9.997851e-02</td>\n",
       "      <td>['pancancer']</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2147 rows × 9 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "              Inactive InactiveDB SL_Candidate  #InactiveSamples  #Samples  \\\n",
       "Inactive                                                                     \n",
       "BRCA1    0       BRCA1      BRCA1         NWD1               234      8930   \n",
       "         1       BRCA1      BRCA1     KIAA1683               234      8930   \n",
       "         2       BRCA1      BRCA1         JUND               234      8930   \n",
       "         3       BRCA1      BRCA1      MIR3188               234      8930   \n",
       "         4       BRCA1      BRCA1        F2RL3               234      8930   \n",
       "...                ...        ...          ...               ...       ...   \n",
       "         2142    BRCA1      BRCA1        OTOP1               234      8930   \n",
       "         2143    BRCA1      BRCA1       ZNF260               234      8930   \n",
       "         2144    BRCA1      BRCA1        GRSF1               234      8930   \n",
       "         2145    BRCA1      BRCA1        NRXN3               234      8930   \n",
       "         2146    BRCA1      BRCA1     FAM177A1               234      8930   \n",
       "\n",
       "                      U1        PValue           FDR         Tissue  \n",
       "Inactive                                                             \n",
       "BRCA1    0     1305850.5  6.250556e-14  1.570640e-09  ['pancancer']  \n",
       "         1     1305687.5  6.450396e-14  1.620855e-09  ['pancancer']  \n",
       "         2     1305525.5  6.655787e-14  1.672466e-09  ['pancancer']  \n",
       "         3     1305525.5  6.655787e-14  1.672466e-09  ['pancancer']  \n",
       "         4     1305301.0  6.955547e-14  1.747790e-09  ['pancancer']  \n",
       "...                  ...           ...           ...            ...  \n",
       "         2142  1191358.5  3.923963e-06  9.860135e-02  ['pancancer']  \n",
       "         2143  1191282.5  3.959944e-06  9.950548e-02  ['pancancer']  \n",
       "         2144  1191265.5  3.968036e-06  9.970880e-02  ['pancancer']  \n",
       "         2145  1191248.0  3.976382e-06  9.991852e-02  ['pancancer']  \n",
       "         2146  1191243.0  3.978769e-06  9.997851e-02  ['pancancer']  \n",
       "\n",
       "[2147 rows x 9 columns]"
      ]
     },
     "execution_count": 209,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import DAISY_operations2\n",
    "importlib.reload(DAISY_operations2)\n",
    "from DAISY_operations2 import *\n",
    "sof_pancancer = SurvivalOfFittest(client, 'SL', \"PanCancerAtlas\", ['BRCA1'], percentile_threshold, cn_threshold, pval_correction, list(tumor_type.value), input_mutations)\n",
    "report=sof_pancancer.loc[(sof_pancancer['FDR'] < p_threshold),]                \n",
    "if report.shape[0]>1:\n",
    "    sof_pancancer_report=report.groupby('Inactive').apply(lambda x: x.sort_values('FDR'))\n",
    "    sof_pancancer_report  \n",
    "sof_pancancer_report    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 210,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Inactive</th>\n",
       "      <th>InactiveDB</th>\n",
       "      <th>SL_Candidate</th>\n",
       "      <th>#InactiveSamples</th>\n",
       "      <th>#Samples</th>\n",
       "      <th>U1</th>\n",
       "      <th>PValue</th>\n",
       "      <th>FDR</th>\n",
       "      <th>Tissue</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>18913</th>\n",
       "      <td>BRCA1</td>\n",
       "      <td>BRCA1</td>\n",
       "      <td>PARP1</td>\n",
       "      <td>234</td>\n",
       "      <td>8930</td>\n",
       "      <td>831956.0</td>\n",
       "      <td>0.999999</td>\n",
       "      <td>1.0</td>\n",
       "      <td>['pancancer']</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      Inactive InactiveDB SL_Candidate  #InactiveSamples  #Samples        U1  \\\n",
       "18913    BRCA1      BRCA1        PARP1               234      8930  831956.0   \n",
       "\n",
       "         PValue  FDR         Tissue  \n",
       "18913  0.999999  1.0  ['pancancer']  "
      ]
     },
     "execution_count": 210,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sof_pancancer.loc[(sof_pancancer['Inactive']=='BRCA1')&(sof_pancancer['SL_Candidate']=='PARP1'), ]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 211,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Inactive</th>\n",
       "      <th>InactiveDB</th>\n",
       "      <th>SL_Candidate</th>\n",
       "      <th>#InactiveSamples</th>\n",
       "      <th>#Samples</th>\n",
       "      <th>U1</th>\n",
       "      <th>PValue</th>\n",
       "      <th>FDR</th>\n",
       "      <th>Tissue</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>749</th>\n",
       "      <td>BRCA1</td>\n",
       "      <td>BRCA1</td>\n",
       "      <td>PARP2</td>\n",
       "      <td>234</td>\n",
       "      <td>8930</td>\n",
       "      <td>1234410.5</td>\n",
       "      <td>1.233575e-08</td>\n",
       "      <td>0.00031</td>\n",
       "      <td>['pancancer']</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    Inactive InactiveDB SL_Candidate  #InactiveSamples  #Samples         U1  \\\n",
       "749    BRCA1      BRCA1        PARP2               234      8930  1234410.5   \n",
       "\n",
       "           PValue      FDR         Tissue  \n",
       "749  1.233575e-08  0.00031  ['pancancer']  "
      ]
     },
     "execution_count": 211,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sof_pancancer.loc[(sof_pancancer['Inactive']=='BRCA1')&(sof_pancancer['SL_Candidate']=='PARP2'), ]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### 4.3. Functional examination inference module"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "4.3.1. CRISPR based functional examination inference module"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 212,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th>Inactive</th>\n",
       "      <th>InactiveDB</th>\n",
       "      <th>SL_Candidate</th>\n",
       "      <th>#InactiveSamples</th>\n",
       "      <th>#Samples</th>\n",
       "      <th>U1</th>\n",
       "      <th>PValue</th>\n",
       "      <th>FDR</th>\n",
       "      <th>Tissue</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Inactive</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th rowspan=\"11\" valign=\"top\">BRCA1</th>\n",
       "      <th>0</th>\n",
       "      <td>BRCA1</td>\n",
       "      <td>BRCA1</td>\n",
       "      <td>COASY</td>\n",
       "      <td>26</td>\n",
       "      <td>782</td>\n",
       "      <td>5016.0</td>\n",
       "      <td>0.000011</td>\n",
       "      <td>0.194523</td>\n",
       "      <td>['pancancer']</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>BRCA1</td>\n",
       "      <td>BRCA1</td>\n",
       "      <td>TYMS</td>\n",
       "      <td>26</td>\n",
       "      <td>782</td>\n",
       "      <td>5466.0</td>\n",
       "      <td>0.000059</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>['pancancer']</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>BRCA1</td>\n",
       "      <td>BRCA1</td>\n",
       "      <td>PARP1</td>\n",
       "      <td>26</td>\n",
       "      <td>771</td>\n",
       "      <td>5418.0</td>\n",
       "      <td>0.000066</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>['pancancer']</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>BRCA1</td>\n",
       "      <td>BRCA1</td>\n",
       "      <td>SKIV2L</td>\n",
       "      <td>26</td>\n",
       "      <td>782</td>\n",
       "      <td>5642.0</td>\n",
       "      <td>0.000109</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>['pancancer']</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>BRCA1</td>\n",
       "      <td>BRCA1</td>\n",
       "      <td>WBP4</td>\n",
       "      <td>26</td>\n",
       "      <td>782</td>\n",
       "      <td>5716.0</td>\n",
       "      <td>0.000141</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>['pancancer']</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1937</th>\n",
       "      <td>BRCA1</td>\n",
       "      <td>BRCA1</td>\n",
       "      <td>CALU</td>\n",
       "      <td>26</td>\n",
       "      <td>782</td>\n",
       "      <td>8375.0</td>\n",
       "      <td>0.099745</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>['pancancer']</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1939</th>\n",
       "      <td>BRCA1</td>\n",
       "      <td>BRCA1</td>\n",
       "      <td>ARID4A</td>\n",
       "      <td>26</td>\n",
       "      <td>771</td>\n",
       "      <td>8253.0</td>\n",
       "      <td>0.099780</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>['pancancer']</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1941</th>\n",
       "      <td>BRCA1</td>\n",
       "      <td>BRCA1</td>\n",
       "      <td>DTNBP1</td>\n",
       "      <td>26</td>\n",
       "      <td>782</td>\n",
       "      <td>8376.0</td>\n",
       "      <td>0.099900</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>['pancancer']</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1940</th>\n",
       "      <td>BRCA1</td>\n",
       "      <td>BRCA1</td>\n",
       "      <td>CEP295</td>\n",
       "      <td>26</td>\n",
       "      <td>782</td>\n",
       "      <td>8376.0</td>\n",
       "      <td>0.099900</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>['pancancer']</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1942</th>\n",
       "      <td>BRCA1</td>\n",
       "      <td>BRCA1</td>\n",
       "      <td>PRB2</td>\n",
       "      <td>26</td>\n",
       "      <td>782</td>\n",
       "      <td>8376.0</td>\n",
       "      <td>0.099900</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>['pancancer']</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1943 rows × 9 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "              Inactive InactiveDB SL_Candidate  #InactiveSamples  #Samples  \\\n",
       "Inactive                                                                     \n",
       "BRCA1    0       BRCA1      BRCA1        COASY                26       782   \n",
       "         1       BRCA1      BRCA1         TYMS                26       782   \n",
       "         2       BRCA1      BRCA1        PARP1                26       771   \n",
       "         3       BRCA1      BRCA1       SKIV2L                26       782   \n",
       "         4       BRCA1      BRCA1         WBP4                26       782   \n",
       "...                ...        ...          ...               ...       ...   \n",
       "         1937    BRCA1      BRCA1         CALU                26       782   \n",
       "         1939    BRCA1      BRCA1       ARID4A                26       771   \n",
       "         1941    BRCA1      BRCA1       DTNBP1                26       782   \n",
       "         1940    BRCA1      BRCA1       CEP295                26       782   \n",
       "         1942    BRCA1      BRCA1         PRB2                26       782   \n",
       "\n",
       "                   U1    PValue       FDR         Tissue  \n",
       "Inactive                                                  \n",
       "BRCA1    0     5016.0  0.000011  0.194523  ['pancancer']  \n",
       "         1     5466.0  0.000059  1.000000  ['pancancer']  \n",
       "         2     5418.0  0.000066  1.000000  ['pancancer']  \n",
       "         3     5642.0  0.000109  1.000000  ['pancancer']  \n",
       "         4     5716.0  0.000141  1.000000  ['pancancer']  \n",
       "...               ...       ...       ...            ...  \n",
       "         1937  8375.0  0.099745  1.000000  ['pancancer']  \n",
       "         1939  8253.0  0.099780  1.000000  ['pancancer']  \n",
       "         1941  8376.0  0.099900  1.000000  ['pancancer']  \n",
       "         1940  8376.0  0.099900  1.000000  ['pancancer']  \n",
       "         1942  8376.0  0.099900  1.000000  ['pancancer']  \n",
       "\n",
       "[1943 rows x 9 columns]"
      ]
     },
     "execution_count": 212,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import DAISY_operations2\n",
    "importlib.reload(DAISY_operations2)\n",
    "from DAISY_operations2 import *\n",
    "crispr_result = FunctionalExamination(client,'SL', \"CRISPR\", ['BRCA1'], percentile_threshold, cn_threshold, pval_correction,list(tumor_type.value), input_mutations )\n",
    "\n",
    "report=crispr_result.loc[(crispr_result['PValue'] < p_threshold),]\n",
    "                      \n",
    "if report.shape[0]>1:\n",
    "    crispr_report=report.groupby('Inactive').apply(lambda x: x.sort_values('PValue'))\n",
    "crispr_report\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 213,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Inactive</th>\n",
       "      <th>InactiveDB</th>\n",
       "      <th>SL_Candidate</th>\n",
       "      <th>#InactiveSamples</th>\n",
       "      <th>#Samples</th>\n",
       "      <th>U1</th>\n",
       "      <th>PValue</th>\n",
       "      <th>FDR</th>\n",
       "      <th>Tissue</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>BRCA1</td>\n",
       "      <td>BRCA1</td>\n",
       "      <td>PARP1</td>\n",
       "      <td>26</td>\n",
       "      <td>771</td>\n",
       "      <td>5418.0</td>\n",
       "      <td>0.000066</td>\n",
       "      <td>1.0</td>\n",
       "      <td>['pancancer']</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  Inactive InactiveDB SL_Candidate  #InactiveSamples  #Samples      U1  \\\n",
       "2    BRCA1      BRCA1        PARP1                26       771  5418.0   \n",
       "\n",
       "     PValue  FDR         Tissue  \n",
       "2  0.000066  1.0  ['pancancer']  "
      ]
     },
     "execution_count": 213,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "crispr_result.loc[(crispr_result['Inactive']==\"BRCA1\")& (crispr_result['SL_Candidate']==\"PARP1\"), ]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 214,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Inactive</th>\n",
       "      <th>InactiveDB</th>\n",
       "      <th>SL_Candidate</th>\n",
       "      <th>#InactiveSamples</th>\n",
       "      <th>#Samples</th>\n",
       "      <th>U1</th>\n",
       "      <th>PValue</th>\n",
       "      <th>FDR</th>\n",
       "      <th>Tissue</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>5430</th>\n",
       "      <td>BRCA1</td>\n",
       "      <td>BRCA1</td>\n",
       "      <td>PARP2</td>\n",
       "      <td>26</td>\n",
       "      <td>771</td>\n",
       "      <td>9080.0</td>\n",
       "      <td>0.293921</td>\n",
       "      <td>1.0</td>\n",
       "      <td>['pancancer']</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     Inactive InactiveDB SL_Candidate  #InactiveSamples  #Samples      U1  \\\n",
       "5430    BRCA1      BRCA1        PARP2                26       771  9080.0   \n",
       "\n",
       "        PValue  FDR         Tissue  \n",
       "5430  0.293921  1.0  ['pancancer']  "
      ]
     },
     "execution_count": 214,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "crispr_result.loc[(crispr_result['Inactive']==\"BRCA1\")& (crispr_result['SL_Candidate']==\"PARP2\"), ]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<br>\n",
    "4.3.2. shRNA based functional examination inference module"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 215,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th>Inactive</th>\n",
       "      <th>InactiveDB</th>\n",
       "      <th>SL_Candidate</th>\n",
       "      <th>#InactiveSamples</th>\n",
       "      <th>#Samples</th>\n",
       "      <th>U1</th>\n",
       "      <th>PValue</th>\n",
       "      <th>FDR</th>\n",
       "      <th>Tissue</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Inactive</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th rowspan=\"11\" valign=\"top\">BRCA1</th>\n",
       "      <th>0</th>\n",
       "      <td>BRCA1</td>\n",
       "      <td>BRCA1</td>\n",
       "      <td>TFAP2A</td>\n",
       "      <td>27</td>\n",
       "      <td>651</td>\n",
       "      <td>4337.0</td>\n",
       "      <td>0.000010</td>\n",
       "      <td>0.167602</td>\n",
       "      <td>['pancancer']</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>BRCA1</td>\n",
       "      <td>BRCA1</td>\n",
       "      <td>PIK3CA</td>\n",
       "      <td>27</td>\n",
       "      <td>653</td>\n",
       "      <td>4903.0</td>\n",
       "      <td>0.000109</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>['pancancer']</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>BRCA1</td>\n",
       "      <td>BRCA1</td>\n",
       "      <td>ERBB2</td>\n",
       "      <td>27</td>\n",
       "      <td>653</td>\n",
       "      <td>5043.0</td>\n",
       "      <td>0.000192</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>['pancancer']</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>BRCA1</td>\n",
       "      <td>BRCA1</td>\n",
       "      <td>BCLAF1</td>\n",
       "      <td>27</td>\n",
       "      <td>649</td>\n",
       "      <td>5130.0</td>\n",
       "      <td>0.000307</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>['pancancer']</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>BRCA1</td>\n",
       "      <td>BRCA1</td>\n",
       "      <td>CENPH</td>\n",
       "      <td>27</td>\n",
       "      <td>649</td>\n",
       "      <td>5130.0</td>\n",
       "      <td>0.000307</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>['pancancer']</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1467</th>\n",
       "      <td>BRCA1</td>\n",
       "      <td>BRCA1</td>\n",
       "      <td>XG</td>\n",
       "      <td>17</td>\n",
       "      <td>308</td>\n",
       "      <td>2015.0</td>\n",
       "      <td>0.099460</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>['pancancer']</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1468</th>\n",
       "      <td>BRCA1</td>\n",
       "      <td>BRCA1</td>\n",
       "      <td>RMI1</td>\n",
       "      <td>21</td>\n",
       "      <td>542</td>\n",
       "      <td>4567.0</td>\n",
       "      <td>0.099558</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>['pancancer']</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1469</th>\n",
       "      <td>BRCA1</td>\n",
       "      <td>BRCA1</td>\n",
       "      <td>ROBO4</td>\n",
       "      <td>24</td>\n",
       "      <td>506</td>\n",
       "      <td>4887.0</td>\n",
       "      <td>0.099734</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>['pancancer']</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1470</th>\n",
       "      <td>BRCA1</td>\n",
       "      <td>BRCA1</td>\n",
       "      <td>VPS52</td>\n",
       "      <td>17</td>\n",
       "      <td>308</td>\n",
       "      <td>2016.0</td>\n",
       "      <td>0.099951</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>['pancancer']</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1471</th>\n",
       "      <td>BRCA1</td>\n",
       "      <td>BRCA1</td>\n",
       "      <td>KRTAP10-12</td>\n",
       "      <td>24</td>\n",
       "      <td>506</td>\n",
       "      <td>4888.0</td>\n",
       "      <td>0.099985</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>['pancancer']</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1472 rows × 9 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "              Inactive InactiveDB SL_Candidate  #InactiveSamples  #Samples  \\\n",
       "Inactive                                                                     \n",
       "BRCA1    0       BRCA1      BRCA1       TFAP2A                27       651   \n",
       "         1       BRCA1      BRCA1       PIK3CA                27       653   \n",
       "         2       BRCA1      BRCA1        ERBB2                27       653   \n",
       "         3       BRCA1      BRCA1       BCLAF1                27       649   \n",
       "         4       BRCA1      BRCA1        CENPH                27       649   \n",
       "...                ...        ...          ...               ...       ...   \n",
       "         1467    BRCA1      BRCA1           XG                17       308   \n",
       "         1468    BRCA1      BRCA1         RMI1                21       542   \n",
       "         1469    BRCA1      BRCA1        ROBO4                24       506   \n",
       "         1470    BRCA1      BRCA1        VPS52                17       308   \n",
       "         1471    BRCA1      BRCA1   KRTAP10-12                24       506   \n",
       "\n",
       "                   U1    PValue       FDR         Tissue  \n",
       "Inactive                                                  \n",
       "BRCA1    0     4337.0  0.000010  0.167602  ['pancancer']  \n",
       "         1     4903.0  0.000109  1.000000  ['pancancer']  \n",
       "         2     5043.0  0.000192  1.000000  ['pancancer']  \n",
       "         3     5130.0  0.000307  1.000000  ['pancancer']  \n",
       "         4     5130.0  0.000307  1.000000  ['pancancer']  \n",
       "...               ...       ...       ...            ...  \n",
       "         1467  2015.0  0.099460  1.000000  ['pancancer']  \n",
       "         1468  4567.0  0.099558  1.000000  ['pancancer']  \n",
       "         1469  4887.0  0.099734  1.000000  ['pancancer']  \n",
       "         1470  2016.0  0.099951  1.000000  ['pancancer']  \n",
       "         1471  4888.0  0.099985  1.000000  ['pancancer']  \n",
       "\n",
       "[1472 rows x 9 columns]"
      ]
     },
     "execution_count": 215,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "shRNA_result = FunctionalExamination(client, 'SL', \"shRNA\", ['BRCA1'] , percentile_threshold, \\\n",
    "                                     cn_threshold, pval_correction, list(tumor_type.value),input_mutations)\n",
    "\n",
    "report=shRNA_result.loc[(shRNA_result['PValue'] < p_threshold),]\n",
    "                      \n",
    "if report.shape[0]>1:\n",
    "    shRNA_report=report.groupby('Inactive').apply(lambda x: x.sort_values('PValue'))\n",
    "shRNA_report\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 216,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Inactive</th>\n",
       "      <th>InactiveDB</th>\n",
       "      <th>SL_Candidate</th>\n",
       "      <th>#InactiveSamples</th>\n",
       "      <th>#Samples</th>\n",
       "      <th>U1</th>\n",
       "      <th>PValue</th>\n",
       "      <th>FDR</th>\n",
       "      <th>Tissue</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>5317</th>\n",
       "      <td>BRCA1</td>\n",
       "      <td>BRCA1</td>\n",
       "      <td>PARP2</td>\n",
       "      <td>15</td>\n",
       "      <td>394</td>\n",
       "      <td>2687.0</td>\n",
       "      <td>0.359623</td>\n",
       "      <td>1.0</td>\n",
       "      <td>['pancancer']</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     Inactive InactiveDB SL_Candidate  #InactiveSamples  #Samples      U1  \\\n",
       "5317    BRCA1      BRCA1        PARP2                15       394  2687.0   \n",
       "\n",
       "        PValue  FDR         Tissue  \n",
       "5317  0.359623  1.0  ['pancancer']  "
      ]
     },
     "execution_count": 216,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "shRNA_result.loc[(shRNA_result['Inactive']==\"BRCA1\")& (shRNA_result['SL_Candidate']==\"PARP2\"), ]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 217,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Inactive</th>\n",
       "      <th>InactiveDB</th>\n",
       "      <th>SL_Candidate</th>\n",
       "      <th>#InactiveSamples</th>\n",
       "      <th>#Samples</th>\n",
       "      <th>U1</th>\n",
       "      <th>PValue</th>\n",
       "      <th>FDR</th>\n",
       "      <th>Tissue</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>7575</th>\n",
       "      <td>BRCA1</td>\n",
       "      <td>BRCA1</td>\n",
       "      <td>PARP1</td>\n",
       "      <td>27</td>\n",
       "      <td>651</td>\n",
       "      <td>8429.0</td>\n",
       "      <td>0.502085</td>\n",
       "      <td>1.0</td>\n",
       "      <td>['pancancer']</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     Inactive InactiveDB SL_Candidate  #InactiveSamples  #Samples      U1  \\\n",
       "7575    BRCA1      BRCA1        PARP1                27       651  8429.0   \n",
       "\n",
       "        PValue  FDR         Tissue  \n",
       "7575  0.502085  1.0  ['pancancer']  "
      ]
     },
     "execution_count": 217,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "shRNA_result.loc[(shRNA_result['Inactive']==\"BRCA1\")& (shRNA_result['SL_Candidate']==\"PARP1\"), ]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 5. Integration of results"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "5.1. Integration of the pairwise Co-expression gene co-expression results on Pancancer and CCLE"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 218,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th>Inactive</th>\n",
       "      <th>SL_Candidate</th>\n",
       "      <th>FDR0</th>\n",
       "      <th>FDR1</th>\n",
       "      <th>AggregatedP</th>\n",
       "      <th>Tissue</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Inactive</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th rowspan=\"11\" valign=\"top\">BRCA1</th>\n",
       "      <th>0</th>\n",
       "      <td>BRCA1</td>\n",
       "      <td>ATAD5</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.215851e-226</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>['pancancer']</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>289</th>\n",
       "      <td>BRCA1</td>\n",
       "      <td>BUB1B</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.747697e-143</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>['pancancer']</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>288</th>\n",
       "      <td>BRCA1</td>\n",
       "      <td>TOP2A</td>\n",
       "      <td>0.0</td>\n",
       "      <td>6.025007e-129</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>['pancancer']</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>287</th>\n",
       "      <td>BRCA1</td>\n",
       "      <td>CDC6</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.129223e-134</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>['pancancer']</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>286</th>\n",
       "      <td>BRCA1</td>\n",
       "      <td>DTL</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.656038e-168</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>['pancancer']</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>706</th>\n",
       "      <td>BRCA1</td>\n",
       "      <td>RBM33</td>\n",
       "      <td>NaN</td>\n",
       "      <td>6.424646e-79</td>\n",
       "      <td>6.424646e-79</td>\n",
       "      <td>['pancancer']</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>707</th>\n",
       "      <td>BRCA1</td>\n",
       "      <td>SLC25A19</td>\n",
       "      <td>NaN</td>\n",
       "      <td>6.853632e-79</td>\n",
       "      <td>6.853632e-79</td>\n",
       "      <td>['pancancer']</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>708</th>\n",
       "      <td>BRCA1</td>\n",
       "      <td>PRRC2C</td>\n",
       "      <td>NaN</td>\n",
       "      <td>7.575563e-79</td>\n",
       "      <td>7.575563e-79</td>\n",
       "      <td>['pancancer']</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>709</th>\n",
       "      <td>BRCA1</td>\n",
       "      <td>HNRNPA3</td>\n",
       "      <td>NaN</td>\n",
       "      <td>8.613395e-79</td>\n",
       "      <td>8.613395e-79</td>\n",
       "      <td>['pancancer']</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>710</th>\n",
       "      <td>BRCA1</td>\n",
       "      <td>SMG1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.044682e-78</td>\n",
       "      <td>1.044682e-78</td>\n",
       "      <td>['pancancer']</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>711 rows × 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "             Inactive SL_Candidate  FDR0           FDR1   AggregatedP  \\\n",
       "Inactive                                                                \n",
       "BRCA1    0      BRCA1        ATAD5   0.0  1.215851e-226  0.000000e+00   \n",
       "         289    BRCA1        BUB1B   0.0  1.747697e-143  0.000000e+00   \n",
       "         288    BRCA1        TOP2A   0.0  6.025007e-129  0.000000e+00   \n",
       "         287    BRCA1         CDC6   0.0  2.129223e-134  0.000000e+00   \n",
       "         286    BRCA1          DTL   0.0  1.656038e-168  0.000000e+00   \n",
       "...               ...          ...   ...            ...           ...   \n",
       "         706    BRCA1        RBM33   NaN   6.424646e-79  6.424646e-79   \n",
       "         707    BRCA1     SLC25A19   NaN   6.853632e-79  6.853632e-79   \n",
       "         708    BRCA1       PRRC2C   NaN   7.575563e-79  7.575563e-79   \n",
       "         709    BRCA1      HNRNPA3   NaN   8.613395e-79  8.613395e-79   \n",
       "         710    BRCA1         SMG1   NaN   1.044682e-78  1.044682e-78   \n",
       "\n",
       "                     Tissue  \n",
       "Inactive                     \n",
       "BRCA1    0    ['pancancer']  \n",
       "         289  ['pancancer']  \n",
       "         288  ['pancancer']  \n",
       "         287  ['pancancer']  \n",
       "         286  ['pancancer']  \n",
       "...                     ...  \n",
       "         706  ['pancancer']  \n",
       "         707  ['pancancer']  \n",
       "         708  ['pancancer']  \n",
       "         709  ['pancancer']  \n",
       "         710  ['pancancer']  \n",
       "\n",
       "[711 rows x 6 columns]"
      ]
     },
     "execution_count": 218,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import DAISY_operations2\n",
    "importlib.reload(DAISY_operations2)\n",
    "from DAISY_operations2 import *\n",
    "coexpression_result = UnionResults([coexp_pancancer_report, coexp_CCLE_report],'SL', ['FDR', 'FDR'],  list(tumor_type.value))\n",
    "coexpression_result=coexpression_result.groupby('Inactive').apply(lambda x: x.sort_values('AggregatedP'))\n",
    "coexpression_result"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<br>\n",
    "5.2. Integration of Survival of Fittest results on Pancancer and CCLE"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 190,
   "metadata": {},
   "outputs": [],
   "source": [
    "sof_result = UnionResults([sof_ccle_report, sof_pancancer_report],  'SL', ['FDR', 'FDR'], list(tumor_type.value))\n",
    "sof_result=sof_result.groupby('Inactive').apply(lambda x: x.sort_values('AggregatedP'))\n",
    "#sof_result\n",
    "#no results from sof_result\n",
    "# don't run"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 219,
   "metadata": {},
   "outputs": [],
   "source": [
    "sof_result=sof_pancancer_report"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<br>\n",
    "5.3. Integration of shRNA and CRISPR based functional examination inference module."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 220,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th>Inactive</th>\n",
       "      <th>SL_Candidate</th>\n",
       "      <th>PValue0</th>\n",
       "      <th>PValue1</th>\n",
       "      <th>AggregatedP</th>\n",
       "      <th>Tissue</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Inactive</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th rowspan=\"11\" valign=\"top\">BRCA1</th>\n",
       "      <th>26</th>\n",
       "      <td>BRCA1</td>\n",
       "      <td>CENPH</td>\n",
       "      <td>0.001106</td>\n",
       "      <td>0.000307</td>\n",
       "      <td>0.000005</td>\n",
       "      <td>['pancancer']</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1943</th>\n",
       "      <td>BRCA1</td>\n",
       "      <td>TFAP2A</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.000010</td>\n",
       "      <td>0.000010</td>\n",
       "      <td>['pancancer']</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>BRCA1</td>\n",
       "      <td>COASY</td>\n",
       "      <td>0.000011</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.000011</td>\n",
       "      <td>['pancancer']</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>BRCA1</td>\n",
       "      <td>TYMS</td>\n",
       "      <td>0.000059</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.000059</td>\n",
       "      <td>['pancancer']</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>BRCA1</td>\n",
       "      <td>PARP1</td>\n",
       "      <td>0.000066</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.000066</td>\n",
       "      <td>['pancancer']</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1941</th>\n",
       "      <td>BRCA1</td>\n",
       "      <td>CEP295</td>\n",
       "      <td>0.099900</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.099900</td>\n",
       "      <td>['pancancer']</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1942</th>\n",
       "      <td>BRCA1</td>\n",
       "      <td>PRB2</td>\n",
       "      <td>0.099900</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.099900</td>\n",
       "      <td>['pancancer']</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1940</th>\n",
       "      <td>BRCA1</td>\n",
       "      <td>DTNBP1</td>\n",
       "      <td>0.099900</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.099900</td>\n",
       "      <td>['pancancer']</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3242</th>\n",
       "      <td>BRCA1</td>\n",
       "      <td>VPS52</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.099951</td>\n",
       "      <td>0.099951</td>\n",
       "      <td>['pancancer']</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3243</th>\n",
       "      <td>BRCA1</td>\n",
       "      <td>KRTAP10-12</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.099985</td>\n",
       "      <td>0.099985</td>\n",
       "      <td>['pancancer']</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3244 rows × 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "              Inactive SL_Candidate   PValue0   PValue1  AggregatedP  \\\n",
       "Inactive                                                               \n",
       "BRCA1    26      BRCA1        CENPH  0.001106  0.000307     0.000005   \n",
       "         1943    BRCA1       TFAP2A       NaN  0.000010     0.000010   \n",
       "         0       BRCA1        COASY  0.000011       NaN     0.000011   \n",
       "         1       BRCA1         TYMS  0.000059       NaN     0.000059   \n",
       "         2       BRCA1        PARP1  0.000066       NaN     0.000066   \n",
       "...                ...          ...       ...       ...          ...   \n",
       "         1941    BRCA1       CEP295  0.099900       NaN     0.099900   \n",
       "         1942    BRCA1         PRB2  0.099900       NaN     0.099900   \n",
       "         1940    BRCA1       DTNBP1  0.099900       NaN     0.099900   \n",
       "         3242    BRCA1        VPS52       NaN  0.099951     0.099951   \n",
       "         3243    BRCA1   KRTAP10-12       NaN  0.099985     0.099985   \n",
       "\n",
       "                      Tissue  \n",
       "Inactive                      \n",
       "BRCA1    26    ['pancancer']  \n",
       "         1943  ['pancancer']  \n",
       "         0     ['pancancer']  \n",
       "         1     ['pancancer']  \n",
       "         2     ['pancancer']  \n",
       "...                      ...  \n",
       "         1941  ['pancancer']  \n",
       "         1942  ['pancancer']  \n",
       "         1940  ['pancancer']  \n",
       "         3242  ['pancancer']  \n",
       "         3243  ['pancancer']  \n",
       "\n",
       "[3244 rows x 6 columns]"
      ]
     },
     "execution_count": 220,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "functional_screening_result = UnionResults([crispr_report, shRNA_report],'SL', ['PValue', 'PValue'], list(tumor_type.value))\n",
    "functional_screening_result=functional_screening_result.groupby('Inactive').apply(lambda x: x.sort_values('AggregatedP'))\n",
    "functional_screening_result"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<br>\n",
    "5.4. Merging the results from all three inference procedures"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 221,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th>Inactive</th>\n",
       "      <th>SL_Candidate</th>\n",
       "      <th>AggregatedP0</th>\n",
       "      <th>AggregatedP2</th>\n",
       "      <th>FinalP</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Inactive</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th rowspan=\"5\" valign=\"top\">BRCA1</th>\n",
       "      <th>0</th>\n",
       "      <td>BRCA1</td>\n",
       "      <td>KIF23</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>0.088368</td>\n",
       "      <td>0.000000e+00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>BRCA1</td>\n",
       "      <td>PRC1</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>0.078395</td>\n",
       "      <td>0.000000e+00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>BRCA1</td>\n",
       "      <td>BLM</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>0.095184</td>\n",
       "      <td>0.000000e+00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>BRCA1</td>\n",
       "      <td>HNRNPM</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>0.023475</td>\n",
       "      <td>0.000000e+00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>BRCA1</td>\n",
       "      <td>ANP32A</td>\n",
       "      <td>6.775424e-82</td>\n",
       "      <td>0.022137</td>\n",
       "      <td>2.875444e-81</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           Inactive SL_Candidate  AggregatedP0  AggregatedP2        FinalP\n",
       "Inactive                                                                  \n",
       "BRCA1    0    BRCA1        KIF23  0.000000e+00      0.088368  0.000000e+00\n",
       "         1    BRCA1         PRC1  0.000000e+00      0.078395  0.000000e+00\n",
       "         2    BRCA1          BLM  0.000000e+00      0.095184  0.000000e+00\n",
       "         3    BRCA1       HNRNPM  0.000000e+00      0.023475  0.000000e+00\n",
       "         4    BRCA1       ANP32A  6.775424e-82      0.022137  2.875444e-81"
      ]
     },
     "execution_count": 221,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "all_merged_results = MergeResults([coexpression_result, sof_result, functional_screening_result], 'SL', list(tumor_type.value))\n",
    "all_merged_results=all_merged_results.groupby('Inactive').apply(lambda x: x.sort_values('FinalP'))\n",
    "all_merged_results\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Results are saved in excel file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 241,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "WriteToExcel(\"DAISY_SL_results.xlsx\", [ all_merged_results],[ \"SL_results\"])\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "functional_screening_result"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
